diff -Nur gigolo-0.4.2.orig/gigolo.desktop.in gigolo-0.4.2/gigolo.desktop.in
--- gigolo-0.4.2.orig/gigolo.desktop.in	2014-01-09 12:58:20.000000000 -0700
+++ gigolo-0.4.2/gigolo.desktop.in	2015-01-30 17:15:12.499169337 -0700
@@ -5,7 +5,7 @@
 _GenericName=Gigolo
 _Comment=A simple frontend to easily connect to remote filesystems
 Exec=gigolo
-Icon=gtk-network
+Icon=folder-remote
 Terminal=false
 Categories=System;GTK;Filesystem;
 StartupNotify=true
