diff -Nuar a/src/lib/plugins/speeddial.cpp b/src/lib/plugins/speeddial.cpp
--- a/src/lib/plugins/speeddial.cpp	2018-02-27 16:56:58.000000000 +0300
+++ b/src/lib/plugins/speeddial.cpp	2018-04-08 00:13:44.000000000 +0300
@@ -63,10 +63,14 @@
     settings.endGroup();
 
     if (allPages.isEmpty()) {
-        allPages = "url:\"https:/kde.org\"|title:\"KDE Community Home\";"
+        allPages = "url:\"https://www.turhost.com\"|title:\"Turhost - Web Hosting\";"
+                   "url:\"https://www.pisilinux.org\"|title:\"Pisilinux - Hoşgeldiniz\";"
+                   "url:\"https://pisilinux.org/forum\"|title:\"Pisilinux Forum\";"
+                   "url:\"https://pisilinux.org/wiki\"|title:\"Pisilinux Wiki\";"
+                   "url:\"https://www.youtube.com/html5\"|title:\"YouTube-html5\";"
+                   "url:\"https://www.google.com\"|title:\"Google\";"
                    "url:\"https://phabricator.kde.org/source/falkon/\"|title:\"falkon Git\";"
-                   "url:\"https://bugs.kde.org/describecomponents.cgi?product=falkon\"|title:\"falkon Bug Tracker\";"
-                   "url:\"https://duckduckgo.com\"|title:\"DuckDuckGo\";";
+                   "url:\"https://github.com/pisilinux\"|title:\"Github\";";
 
     }
     changed(allPages);
