diff -Nuar a/settings/main.c b/settings/main.c
--- a/settings/main.c	2021-05-04 13:41:17.203543534 +0300
+++ b/settings/main.c	2021-05-04 13:53:22.336596835 +0300
@@ -419,11 +419,11 @@
         { N_("Home"), "user-home", "gnome-fs-desktop",
           DESKTOP_ICONS_SHOW_HOME, TRUE },
         { N_("File System"), "drive-harddisk", "gnome-dev-harddisk",
-          DESKTOP_ICONS_SHOW_FILESYSTEM, TRUE },
+          DESKTOP_ICONS_SHOW_FILESYSTEM, FALSE },
         { N_("Trash"), "user-trash", "gnome-fs-trash-empty",
           DESKTOP_ICONS_SHOW_TRASH, TRUE },
         { N_("Removable Devices"), "drive-removable-media", "gnome-dev-removable",
-          DESKTOP_ICONS_SHOW_REMOVABLE, TRUE },
+          DESKTOP_ICONS_SHOW_REMOVABLE, FALSE },
         { N_("Network Shares"), "gtk-network", "gnome-dev-network",
           DESKTOP_ICONS_SHOW_NETWORK_REMOVABLE, TRUE },
         { N_("Disks and Drives"), "drive-harddisk-usb", "gnome-dev-removable-usb",
