diff -Nuar a/src/nautilus-window.c b/src/nautilus-window.c
--- a/src/nautilus-window.c	2023-04-23 02:12:42.000000000 +0300
+++ b/src/nautilus-window.c	2023-04-26 11:26:44.870583418 +0300
@@ -1184,15 +1184,15 @@
     if (length == 1)
     {
         file_label = g_file_get_basename (files->data);
-        /* Translators: only one item has been deleted and %s is its name. */
+        /* Translators: only one item has been moved to trash and %s is its name. */
         label = g_markup_printf_escaped (_("“%s” moved to trash"), file_label);
         g_free (file_label);
     }
     else
     {
-        /* Translators: one or more items might have been deleted, and %d
+        /* Translators: one or more items might have been moved to trash, and %d
          * is the count. */
-        label = g_markup_printf_escaped (ngettext ("%d file deleted", "%d files deleted", length), length);
+        label = g_markup_printf_escaped (ngettext ("%d file moved to trash", "%d files moved to trash", length), length);
     }

     return label;
