diff --git a/src/gtk/nautilusgtksidebarrow.ui b/src/gtk/nautilusgtksidebarrow.ui
index 08fc3ccfac..43823fb2d4 100644
--- a/src/gtk/nautilusgtksidebarrow.ui
+++ b/src/gtk/nautilusgtksidebarrow.ui
@@ -57,7 +57,9 @@
                 <property name="halign">3</property>
                 <property name="valign">3</property>
                 <property name="margin-start">4px</property>
+                <property name="margin-end">4px</property>
                 <property name="visible">0</property>
+                <property name="opacity">0.7</property>
               </object>
             </child>
           </object>
