diff -Nuar a/nemo-share/src/nemo-share.c b/nemo-share/src/nemo-share.c
--- a/nemo-share/src/nemo-share.c	2024-11-28 10:57:23.097598174 +0300
+++ b/nemo-share/src/nemo-share.c	2024-11-28 10:57:55.770189710 +0300
@@ -32,8 +32,6 @@
 #include <libnemo-extension/nemo-property-page-provider.h>
 #include <libnemo-extension/nemo-name-and-desc-provider.h>
 
-#include <libcinnamon-desktop/gnome-installer.h>
-
 #include "nemo-share.h"
 
 #include <glib/gi18n-lib.h>
