diff -ru nemo-extensions-6.4.0.orig/nemo-seahorse/tool/seahorse-pgp-operation.c nemo-extensions-6.4.0/nemo-seahorse/tool/seahorse-pgp-operation.c
--- nemo-extensions-6.4.0.orig/nemo-seahorse/tool/seahorse-pgp-operation.c	2025-06-04 12:35:11.131984281 +0200
+++ nemo-extensions-6.4.0/nemo-seahorse/tool/seahorse-pgp-operation.c	2025-06-04 12:35:31.853668254 +0200
@@ -300,7 +300,6 @@
         break;
 
     case GPGME_EVENT_NEXT_KEY:
-    case GPGME_EVENT_NEXT_TRUSTITEM:
     default:
         /* Ignore unsupported event types */
         break;
