Author: Daniel Baumann <mail@daniel-baumann.ch>
Description: Setting no delay for fast booting machines (Closes: #752746).

Index: plymouth-0.8.8/src/plymouthd.defaults
===================================================================
--- plymouth-0.8.8.orig/src/plymouthd.conf
+++ plymouth-0.8.8/src/plymouthd.conf
@@ -1,3 +1,4 @@
 # Administrator customizations go in this file
 #[Daemon]
 #Theme=fade-in
+#ShowDelay=0
Index: plymouth-0.8.8/src/plymouthd.defaults
===================================================================
--- plymouth-0.8.8.orig/src/plymouthd.defaults
+++ plymouth-0.8.8/src/plymouthd.defaults
@@ -2,3 +2,4 @@
 # upgrades.
 [Daemon]
 Theme=spinner
+ShowDelay=0
