--- configy.py	2015-11-05 04:35:43.000000000 +0200
+++ pyqt5/config.py	2015-11-05 04:37:20.021862391 +0200
@@ -38,8 +38,8 @@
 
 # Run SIP to generate the code.  Note that we tell SIP where to find the qt
 # module's specification files using the -I flag.
-os.system(" ".join([config.sip_bin, "-c", ".", "-b", build_file, "-I",
-                    config.default_sip_dir+"/PyQt5", qt_sip_flags, "qtermwidget.sip"]))
+os.system(" ".join(["/usr/bin/py2sip", "-c", ".", "-b", build_file, "-I",
+                    config.default_sip_dir+"/Py2Qt5", qt_sip_flags, "qtermwidget.sip"]))
 		    #pyqt_sip_dir yok  config.pyqt_sip_dir
 		    
 # We are going to install the SIP specification file for this module and
