--- /dev/null	2022-11-16 05:56:56.793999937 +0300
+++ b/devtools/Site/site.config.m4	2022-11-16 07:57:02.406144281 +0300
@@ -0,0 +1,9 @@
+dnl Enable libmilter with a pool of workers
+APPENDDEF(`conf_libmilter_ENVDEF',`-D_FFR_WORKERS_POOL=1')
+APPENDDEF(`conf_libmilter_ENVDEF',`-DMIN_WORKERS=4')
+dnl Use poll instead of select
+APPENDDEF(`conf_libmilter_ENVDEF',`-DSM_CONF_POLL=1')
+dnl Enable IPv6
+APPENDDEF(`conf_libmilter_ENVDEF',`-DNETINET6=1')
+dnl Add -fPIC
+APPENDDEF(`conf_libmilter_ENVDEF',`-fPIC')
