diff -Nuar libostree-2018.8.orig/configure.ac libostree-2018.8/configure.ac
--- libostree-2018.8.orig/configure.ac	2018-08-22 16:53:53.000000000 +0300
+++ libostree-2018.8/configure.ac	2018-10-13 21:50:43.341668859 +0300
@@ -35,7 +35,6 @@
   -Werror=empty-body \
   -Werror=strict-prototypes \
   -Werror=missing-prototypes \
-  -Werror=implicit-function-declaration \
   "-Werror=format=2 -Werror=format-security -Werror=format-nonliteral" \
   -Werror=pointer-arith -Werror=init-self \
   -Werror=missing-declarations \
