diff -Nuar a/data/meson.build b/data/meson.build
--- a/data/meson.build	2022-11-03 14:58:44.501669486 +0300
+++ b/data/meson.build	2022-11-03 14:57:18.864740095 +0300
@@ -4,7 +4,6 @@
 
 appdata = 'org.gnome.gedit.appdata.xml'
 appdata_file = i18n.merge_file(
-  appdata,
   input: appdata + '.in',
   output: appdata,
   po_dir: '../po/',
@@ -23,7 +22,6 @@
 
 desktop_file = 'org.gnome.gedit.desktop'
 desktop_output_file = i18n.merge_file(
-  desktop_file,
   type: 'desktop',
   input: desktop_file + '.in',
   output: desktop_file,
