diff -Nuar a/runners/dictionary/dictionarymatchengine.cpp b/runners/dictionary/dictionarymatchengine.cpp
--- a/runners/dictionary/dictionarymatchengine.cpp	2017-10-24 13:51:25.000000000 +0300
+++ b/runners/dictionary/dictionarymatchengine.cpp	2017-11-04 17:44:10.006867811 +0300
@@ -73,7 +73,7 @@
         /* Because of QString's CoW semantics, we don't have to worry about
          * the overhead of assigning this to every item. */
         data->definition = definition;
-        data->waitCondition.notify_one();
+        data->waitCondition.wakeOne();
     }
     m_wordLock.unlock();
 }
