diff -Nuar themes_o/ambiance/wallpaper.cfg themes/ambiance/wallpaper.cfg
--- themes_o/ambiance/wallpaper.cfg	2017-10-10 18:45:01.000000000 +0300
+++ themes/ambiance/wallpaper.cfg	2018-03-28 12:16:32.476989000 +0300
@@ -1,2 +1,2 @@
 wallpapers/size=9
-wallpapers/1/file=Butterfly-Kenneth-Wimer.jpg
+wallpapers/1/file=felis_catus.png
diff -Nuar themes_o/frost/wallpaper.cfg themes/frost/wallpaper.cfg
--- themes_o/frost/wallpaper.cfg	2017-10-10 18:45:01.000000000 +0300
+++ themes/frost/wallpaper.cfg	2018-03-28 12:25:15.057980000 +0300
@@ -1,2 +1,2 @@
 wallpapers/size=1
-wallpapers/1/file=lxqt-origami-light.png
\ Dosya sonunda yenisatır yok.
+wallpapers/1/file=frost.png
diff -Nuar themes_o/kde-plasma/wallpaper.cfg themes/kde-plasma/wallpaper.cfg
--- themes_o/kde-plasma/wallpaper.cfg	2017-10-10 18:45:01.000000000 +0300
+++ themes/kde-plasma/wallpaper.cfg	2018-03-28 12:28:42.591976000 +0300
@@ -1,4 +1,4 @@
 /* Wallpaper by Nuno from the KDE Visual Design Group */
 
 wallpapers/size=9
-wallpapers/1/file=kde-plasma.png
+wallpapers/1/file=dolunay_ata.png
diff -Nuar themes_o/light/wallpaper.cfg themes/light/wallpaper.cfg
--- themes_o/light/wallpaper.cfg	2017-10-10 18:45:01.000000000 +0300
+++ themes/light/wallpaper.cfg	2018-03-28 12:33:35.314971000 +0300
@@ -1,2 +1,2 @@
 wallpapers/size=1
-wallpapers/1/file=simple_blue_widescreen.png
+wallpapers/1/file=light.png
