diff -Nuar a/themes/ambiance/wallpaper.cfg b/themes/ambiance/wallpaper.cfg
--- a/themes/ambiance/wallpaper.cfg	2022-04-16 16:26:52.000000000 +0300
+++ b/themes/ambiance/wallpaper.cfg	2022-04-19 13:57:17.377708398 +0300
@@ -1,2 +1,2 @@
 wallpapers/size=9
-wallpapers/1/file=../../wallpapers/Butterfly-Kenneth-Wimer.jpg
+wallpapers/1/file=../../wallpapers/Crocus-Ancyrensis.jpg
diff -Nuar a/themes/Arch-Colors/wallpaper.cfg b/themes/Arch-Colors/wallpaper.cfg
--- a/themes/Arch-Colors/wallpaper.cfg	2022-04-16 16:26:52.000000000 +0300
+++ b/themes/Arch-Colors/wallpaper.cfg	2022-04-19 13:56:56.081710035 +0300
@@ -1,2 +1,2 @@
 wallpapers/size=1
-wallpapers/1/file=../../wallpapers/plasma_arch.png
+wallpapers/1/file=../../wallpapers/Crocus-Ancyrensis.jpg
diff -Nuar a/themes/Clearlooks/wallpaper.cfg b/themes/Clearlooks/wallpaper.cfg
--- a/themes/Clearlooks/wallpaper.cfg	2022-04-16 16:26:52.000000000 +0300
+++ b/themes/Clearlooks/wallpaper.cfg	2022-04-19 14:00:11.465695015 +0300
@@ -1,2 +1,2 @@
 wallpapers/size=1
-wallpapers/1/file=../../wallpapers/waves-logo.png
+wallpapers/1/file=../../wallpapers/Crocus-Ancyrensis-2.jpg
diff -Nuar a/themes/dark/wallpaper.cfg b/themes/dark/wallpaper.cfg
--- a/themes/dark/wallpaper.cfg	2022-04-16 16:26:52.000000000 +0300
+++ b/themes/dark/wallpaper.cfg	2022-04-19 13:57:33.040707194 +0300
@@ -1,2 +1,2 @@
 wallpapers/size=1
-wallpapers/1/file=../../wallpapers/triangles.png
+wallpapers/1/file=../../wallpapers/Crocus-Ancyrensis.jpg
diff -Nuar a/themes/frost/wallpaper.cfg b/themes/frost/wallpaper.cfg
--- a/themes/frost/wallpaper.cfg	2022-04-16 16:26:52.000000000 +0300
+++ b/themes/frost/wallpaper.cfg	2022-04-19 13:57:53.464705624 +0300
@@ -1,2 +1,2 @@
 wallpapers/size=1
-wallpapers/1/file=../../wallpapers/origami-bright-logo.png
+wallpapers/1/file=../../wallpapers/Crocus-Ancyrensis.jpg
diff -Nuar a/themes/KDE-Plasma/wallpaper.cfg b/themes/KDE-Plasma/wallpaper.cfg
--- a/themes/KDE-Plasma/wallpaper.cfg	2022-04-16 16:26:52.000000000 +0300
+++ b/themes/KDE-Plasma/wallpaper.cfg	2022-04-19 14:00:42.744692610 +0300
@@ -1,4 +1,4 @@
 /* Wallpaper by Nuno from the KDE Visual Design Group */
 
 wallpapers/size=9
-wallpapers/1/file=../../wallpapers/kde-plasma.png
+wallpapers/1/file=../../wallpapers/Crocus-Ancyrensis-2.jpg
diff -Nuar a/themes/kvantum/wallpaper.cfg b/themes/kvantum/wallpaper.cfg
--- a/themes/kvantum/wallpaper.cfg	2022-04-16 16:26:52.000000000 +0300
+++ b/themes/kvantum/wallpaper.cfg	2022-04-19 13:58:17.568703771 +0300
@@ -1,2 +1,2 @@
 wallpapers/size=1
-wallpapers/1/file=../../wallpapers/plasma-logo-dark.png
+wallpapers/1/file=../../wallpapers/Crocus-Ancyrensis.jpg
diff -Nuar a/themes/Leech/wallpaper.cfg b/themes/Leech/wallpaper.cfg
--- a/themes/Leech/wallpaper.cfg	2022-04-16 16:26:52.000000000 +0300
+++ b/themes/Leech/wallpaper.cfg	2022-04-19 13:58:28.696702915 +0300
@@ -1,2 +1,2 @@
 wallpapers/size=1
-wallpapers/1/file=../../wallpapers/origami-dark.png
+wallpapers/1/file=../../wallpapers/Crocus-Ancyrensis.jpg
diff -Nuar a/themes/light/wallpaper.cfg b/themes/light/wallpaper.cfg
--- a/themes/light/wallpaper.cfg	2022-04-16 16:26:52.000000000 +0300
+++ b/themes/light/wallpaper.cfg	2022-04-19 14:00:57.272691494 +0300
@@ -1,2 +1,2 @@
 wallpapers/size=1
-wallpapers/1/file=../../wallpapers/triangles-logo.png
+wallpapers/1/file=../../wallpapers/Crocus-Ancyrensis-2.jpg
diff -Nuar a/themes/silver/wallpaper.cfg b/themes/silver/wallpaper.cfg
--- a/themes/silver/wallpaper.cfg	2022-04-16 16:26:52.000000000 +0300
+++ b/themes/silver/wallpaper.cfg	2022-04-19 13:58:45.368701634 +0300
@@ -1,2 +1,2 @@
 wallpapers/size=1
-wallpapers/1/file=../../wallpapers/beam.png
+wallpapers/1/file=../../wallpapers/Crocus-Ancyrensis.jpg
diff -Nuar a/themes/system/wallpaper.cfg b/themes/system/wallpaper.cfg
--- a/themes/system/wallpaper.cfg	2022-04-16 16:26:52.000000000 +0300
+++ b/themes/system/wallpaper.cfg	2022-04-19 13:59:21.728698839 +0300
@@ -1,2 +1,2 @@
 wallpapers/size=1
-wallpapers/1/file=../../wallpapers/plasma-logo-dark.png
+wallpapers/1/file=../../wallpapers/Crocus-Ancyrensis.jpg
diff -Nuar a/themes/Valendas/wallpaper.cfg b/themes/Valendas/wallpaper.cfg
--- a/themes/Valendas/wallpaper.cfg	2022-04-16 16:26:52.000000000 +0300
+++ b/themes/Valendas/wallpaper.cfg	2022-04-19 13:59:01.352700405 +0300
@@ -1,2 +1,2 @@
 wallpapers/size=1
-wallpapers/1/file=../../wallpapers/Valendas.png
+wallpapers/1/file=../../wallpapers/Crocus-Ancyrensis.jpg
