--- a/CMakeLists.txt      2014-09-09 02:40:08.000000000 +0300
+++ b/CMakeLists.txt  2014-09-22 17:47:14.215639453 +0300
@@ -6,13 +6,6 @@
 
 set(REQUIRED_QT_VERSION 5.2.0)
 
-# We don't build in-source
-if ("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_BINARY_DIR}")
-   message (
-      FATAL_ERROR
-      "kactivities require an out of source build. Please create a separate build directory and run 'cmake path_to_plasma [options]' there."
-   )
-endif ()
 
 # Extra CMake stuff
 find_package (ECM 1.2.0 REQUIRED NO_MODULE)
 
