diff --git a/gdk/x11/gdksurface-x11.c b/gdk/x11/gdksurface-x11.c
index a76e4824f2ed1c0e2c1f002257c3fcbda60c9f16..40f37523cd35301a8d3d3eb655d9cab7bdb7eee8 100644
--- a/gdk/x11/gdksurface-x11.c
+++ b/gdk/x11/gdksurface-x11.c
@@ -208,6 +208,8 @@ gdk_x11_surface_update_size (GdkX11Surface *self,
       cairo_surface_set_device_scale (self->cairo_surface, scale, scale);
     }
 
+  gdk_surface_invalidate_rect (surface, NULL);
+
   return TRUE;
 }
 
