--- linux/3rdparty/Kconfig.orig	2019-05-31 23:26:48.844304457 +0300
+++ linux/3rdparty/Kconfig	2019-05-31 23:30:41.568375751 +0300
@@ -3,5 +3,6 @@
 menu "External 3rdparty kernel additions"
 
 source "3rdparty/ndiswrapper/Kconfig"
+source "3rdparty/rtl8812au/Kconfig"
 
 endmenu
--- linux/3rdparty/Makefile.orig	2019-05-31 23:27:35.405319287 +0300
+++ linux/3rdparty/Makefile	2019-05-31 23:30:56.855708733 +0300
@@ -2,3 +2,4 @@
 
 obj- := 3rdparty.o  # Dummy rule to force built-in.o to be made
 obj-$(CONFIG_NDISWRAPPER) += ndiswrapper/
+obj-$(CONFIG_RTL8812AU)   += rtl8812au/
