--- linux-5.3.orig/3rdparty/Kconfig
+++ linux-5.3/3rdparty/Kconfig
@@ -3,6 +3,7 @@
 menu "External 3rdparty kernel additions"
 
 source "3rdparty/ndiswrapper/Kconfig"
+source "3rdparty/rtl8821ce/Kconfig"
 source "3rdparty/rtl8812au/Kconfig"
 source "3rdparty/rtl8723de/Kconfig"
 source "3rdparty/viahss/Kconfig"
--- linux-5.3.orig/3rdparty/Makefile
+++ linux-5.3/3rdparty/Makefile
@@ -2,6 +2,7 @@
 
 obj- := 3rdparty.o  # Dummy rule to force built-in.o to be made
 obj-$(CONFIG_NDISWRAPPER) += ndiswrapper/
+obj-$(CONFIG_RTL8821CE)   += rtl8821ce/
 obj-$(CONFIG_RTL8812AU)   += rtl8812au/
 obj-$(CONFIG_RTL8723DE)   += rtl8723de/
 obj-$(CONFIG_VIAHSS)      += viahss/
