--- linux/3rdparty/rtl8723de/include/wifi.h.orig	2019-01-19 16:29:26.302342688 +0200
+++ linux/3rdparty/rtl8723de/include/wifi.h	2019-01-19 16:32:25.543905828 +0200
@@ -1003,8 +1003,9 @@ typedef enum _HT_CAP_AMPDU_DENSITY {
  * According to IEEE802.11n spec size varies from 8K to 64K (in powers of 2)
  */
 #define IEEE80211_MIN_AMPDU_BUF 0x8
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 19, 0))
 #define IEEE80211_MAX_AMPDU_BUF 0x40
-
+#endif
 
 /* Spatial Multiplexing Power Save Modes */
 #define WLAN_HT_CAP_SM_PS_STATIC		0
