diff -up ./3rdparty/rtl8723de/Kconfig.tv ./3rdparty/rtl8723de/Kconfig
--- ./3rdparty/rtl8723de/Kconfig.tv	2020-10-13 21:02:42.452125426 +0200
+++ ./3rdparty/rtl8723de/Kconfig	2020-10-13 20:59:14.012371730 +0200
@@ -1,6 +1,6 @@
 config RTL8723DE
 	tristate "Realtek 8723DE PCI WiFi"
 	depends on PCI
-	---help---
+	help
 	  Help message of RTL8723DE
 
diff -up ./3rdparty/rtl8821ce/Kconfig.tv ./3rdparty/rtl8821ce/Kconfig
--- ./3rdparty/rtl8821ce/Kconfig.tv	2020-10-13 16:09:11.157856145 +0200
+++ ./3rdparty/rtl8821ce/Kconfig	2020-10-13 20:58:17.242118685 +0200
@@ -1,6 +1,6 @@
 config RTL8821CE
 	tristate "Realtek 8821C PCI WiFi"
 	depends on PCI
-	---help---
+	help
 	  Help message of RTL8821CE
 
diff -up ./3rdparty/viahss/Kconfig.tv ./3rdparty/viahss/Kconfig
--- ./3rdparty/viahss/Kconfig.tv	2020-10-13 21:02:44.844168369 +0200
+++ ./3rdparty/viahss/Kconfig	2020-10-13 20:59:19.875488718 +0200
@@ -1,7 +1,7 @@
 config VIAHSS
 	tristate "VIA High Speed Serial"
 	depends on SERIAL_CORE && m
-	---help---
+	help
 	  VIA High Speed Serial is a little kernel module (1 KB) which enables
 	  high speed serial port modes of VIA VT82C686A or VT82C686B
 	  southbridge-equipped motherboards. With this module, you can use the
