 tools/tracing/rtla/Makefile |    5 -----
 1 file changed, 5 deletions(-)

--- linux-6.1/tools/tracing/rtla/Makefile.orig
+++ linux-6.1/tools/tracing/rtla/Makefile
@@ -124,11 +124,6 @@ install: doc_install
 
 .PHONY: clean tarball
 clean: doc_clean
-	@test ! -f rtla || rm rtla
-	@test ! -f rtla-static || rm rtla-static
-	@test ! -f src/rtla.o || rm src/rtla.o
-	@test ! -f $(TARBALL) || rm -f $(TARBALL)
-	@rm -rf *~ $(OBJ) *.tar.$(CEXT)
 
 tarball: clean
 	rm -rf $(NAME)-$(VERSION) && mkdir $(NAME)-$(VERSION)
