diff -up linux-6.12/drivers/staging/Kconfig.setting linux-6.12/drivers/staging/Kconfig
--- linux-6.12/drivers/staging/Kconfig.setting	2025-12-09 17:31:43.591783063 +0100
+++ linux-6.12/drivers/staging/Kconfig	2025-12-09 18:37:20.268418187 +0100
@@ -30,6 +30,8 @@ source "drivers/staging/rtl8192e/Kconfig
 
 source "drivers/staging/rtl8723bs/Kconfig"
 
+source "drivers/staging/rtl8812au/Kconfig"
+
 source "drivers/staging/rts5208/Kconfig"
 
 source "drivers/staging/octeon/Kconfig"
diff -up linux-6.12/drivers/staging/Makefile.setting linux-6.12/drivers/staging/Makefile
--- linux-6.12/drivers/staging/Makefile.setting	2025-12-09 18:37:20.268418187 +0100
+++ linux-6.12/drivers/staging/Makefile	2025-12-09 18:38:48.903476477 +0100
@@ -5,6 +5,7 @@ obj-y				+= media/
 obj-$(CONFIG_FB_OLPC_DCON)	+= olpc_dcon/
 obj-$(CONFIG_RTL8192E)		+= rtl8192e/
 obj-$(CONFIG_RTL8723BS)		+= rtl8723bs/
+obj-$(CONFIG_RTL8812AU)		+= rtl8812au/
 obj-$(CONFIG_RTS5208)		+= rts5208/
 obj-$(CONFIG_OCTEON_ETHERNET)	+= octeon/
 obj-$(CONFIG_VT6655)		+= vt6655/
