--- a/kalarm/CMakeLists.txt~	2015-11-13 14:44:23.799000000 +0300
+++ b/kalarm/CMakeLists.txt	2015-11-13 14:45:28.256000000 +0300
@@ -179,6 +179,6 @@ install(FILES org.kde.kalarm.kalarm.xml
 
 add_executable(kalarm_helper rtcwakeaction.cpp kalarm_debug.cpp)
 target_link_libraries(kalarm_helper KF5::Auth KF5::KDELibs4Support)
-install(TARGETS kalarm_helper DESTINATION ${KDE_INSTALL_LIBEXECDIR})
+install(TARGETS kalarm_helper DESTINATION ${KAUTH_HELPER_INSTALL_DIR})
 kauth_install_helper_files(kalarm_helper org.kde.kalarmrtcwake root)
 kauth_install_actions(org.kde.kalarmrtcwake rtcwakeaction.actions)
