--- a/kleopatra/conf/CMakeLists.txt~	2015-12-21 16:21:25.000000000 +0300
+++ b/kleopatra/conf/CMakeLists.txt	2015-12-23 16:16:31.299000000 +0300
@@ -70,6 +70,7 @@ target_link_libraries(kcm_kleopatra
   KF5::Libkleo
   KF5::KCMUtils
   KF5::IconThemes
+  KF5::I18n
   Qt5::DBus
   ${_kcm_kleopatra_extra_libs}
   ${_kcm_kleopatra_libkleopatraclient_extra_LIBS}
--- a/kmail/CMakeLists.txt~	2015-12-21 16:21:25.000000000 +0300
+++ b/kmail/CMakeLists.txt	2015-12-23 16:28:13.774000000 +0300
@@ -268,7 +268,6 @@ add_library(kmailprivate ${kmailprivate_
 generate_export_header(kmailprivate BASE_NAME kmail)
 
 target_link_libraries(kmailprivate
-   PRIVATE
     KF5::Gravatar
     KF5::Mime
     KF5::AkonadiCore
