diff -up gimp-lqr-plugin-0.7.2/src/interface_I.c.me gimp-lqr-plugin-0.7.2/src/interface_I.c
--- gimp-lqr-plugin-0.7.2/src/interface_I.c.me	2020-03-05 13:29:16.392230114 +0100
+++ gimp-lqr-plugin-0.7.2/src/interface_I.c	2020-03-05 13:49:26.218667723 +0100
@@ -65,13 +65,8 @@ static void callback_alarm_triggered (Gt
 
 gint dialog_I_response = GTK_RESPONSE_OK;
 
-PlugInUIVals *ui_state;
-PlugInVals *state;
-PlugInDialogVals *dialog_state;
-gboolean features_are_sensitive;
 InterfaceIData interface_I_data;
 
-GtkWidget *dlg;
 GtkWidget *coordinates;
 
 gulong size_changed = 0;
diff -up gimp-lqr-plugin-0.7.2/src/interface_I.h.me gimp-lqr-plugin-0.7.2/src/interface_I.h
--- gimp-lqr-plugin-0.7.2/src/interface_I.h.me	2020-03-05 13:30:05.760958371 +0100
+++ gimp-lqr-plugin-0.7.2/src/interface_I.h	2020-03-05 13:52:27.564267742 +0100
@@ -46,4 +46,10 @@ dialog_I (PlugInImageVals * image_vals,
           PlugInColVals * col_vals,
           PlugInDialogVals * dialog_vals);
 
+extern PlugInUIVals *ui_state;
+extern PlugInVals *state;
+extern PlugInDialogVals *dialog_state;
+extern gboolean features_are_sensitive;
+extern GtkWidget *dlg;
+
 #endif /* __INTERFACE_I_H__ */
diff -up gimp-lqr-plugin-0.7.2/src/interface_aux.c.me gimp-lqr-plugin-0.7.2/src/interface_aux.c
--- gimp-lqr-plugin-0.7.2/src/interface_aux.c.me	2020-03-05 13:31:23.063424721 +0100
+++ gimp-lqr-plugin-0.7.2/src/interface_aux.c	2020-03-05 13:47:59.782119588 +0100
@@ -48,12 +48,6 @@ static void callback_dialog_aux_response
 
 gint dialog_aux_response = GTK_RESPONSE_OK;
 
-PlugInUIVals *ui_state;
-PlugInVals *state;
-PlugInDialogVals *dialog_state;
-
-GtkWidget *dlg;
-
 /***  Public functions  ***/
 
 gint
diff -up gimp-lqr-plugin-0.7.2/src/interface_aux.h.me gimp-lqr-plugin-0.7.2/src/interface_aux.h
--- gimp-lqr-plugin-0.7.2/src/interface_aux.h.me	2020-03-05 13:32:04.696294966 +0100
+++ gimp-lqr-plugin-0.7.2/src/interface_aux.h	2020-03-05 13:47:39.164860985 +0100
@@ -39,4 +39,9 @@ dialog_aux (PlugInImageVals * image_vals
 
 GimpRGB * colour_from_type (gint32 image_ID, AuxLayerType layer_type);
 
+extern PlugInUIVals *ui_state;
+extern PlugInVals *state;
+extern PlugInDialogVals *dialog_state;
+extern GtkWidget *dlg;
+
 #endif /* __INTERFACE_AUX_H__ */
