Always build pacrunners as separate modules, not libproxy builtins
--- libproxy-0.4.8/libproxy/cmake/modules.cmk.orig	2012-10-05 17:15:58.655470278 +0200
+++ libproxy-0.4.8/libproxy/cmake/modules.cmk	2012-10-05 17:47:10.435431265 +0200
@@ -19,10 +19,7 @@
 include(cmake/modules/pacrunner_webkit.cmk)
 
 # Build the pacrunner into libproxy unless we are building for multiple engines
-set(BIPR 1 CACHE BOOL "Build internal pacrunner? (Always false when building more than one PR")
-if((MOZJS_FOUND AND WEBKIT_FOUND) OR (MOZJS_FOUND AND NATUS_FOUND) OR (WEBKIT_FOUND AND NATUS_FOUND))
-  set(BIPR 0)
-endif()
+set(BIPR 0)
 
 #
 ## Module definition
