diff -Nuar a/nature/Makefile.am b/nature/Makefile.am
--- a/nature/Makefile.am	2020-08-20 17:53:37.000000000 +0300
+++ b/nature/Makefile.am	2021-04-11 21:00:35.447842274 +0300
@@ -1,6 +1,14 @@
 backgrounddir = $(datadir)/backgrounds/mate/nature
 
 background_DATA =\
+	1.jpg \
+	2.jpg \
+	3.jpg \
+	4.jpg \
+	5.jpg \
+	6.jpg \
+	7.jpg \
+	8.jpg \
 	Aqua.jpg \
 	Blinds.jpg \
 	Dune.jpg \
diff -Nuar a/nature/Makefile.in b/nature/Makefile.in
--- a/nature/Makefile.in	2020-08-20 17:55:17.000000000 +0300
+++ b/nature/Makefile.in	2021-04-11 21:00:50.655843392 +0300
@@ -270,6 +270,14 @@
 top_srcdir = @top_srcdir@
 backgrounddir = $(datadir)/backgrounds/mate/nature
 background_DATA = \
+	1.jpg \
+	2.jpg \
+	3.jpg \
+	4.jpg \
+	5.jpg \
+	6.jpg \
+	7.jpg \
+	8.jpg \
 	Aqua.jpg \
 	Blinds.jpg \
 	Dune.jpg \
diff -Nuar a/nature/mate-nature.xml.in b/nature/mate-nature.xml.in
--- a/nature/mate-nature.xml.in	1970-01-01 02:00:00.000000000 +0200
+++ b/nature/mate-nature.xml.in	2021-04-11 21:01:08.943844736 +0300
@@ -0,0 +1,120 @@
+<?xml version="1.0"?>
+<!DOCTYPE wallpapers SYSTEM "mate-wp-list.dtd">
+<wallpapers>
+	<wallpaper deleted="false">
+	<!-- Translators: This is the name of a desktop background image that shows an orange/red flower.
+		You might want to translate it into the equivalent words of your language  -->
+		<name>Fresh Flower</name>
+		<filename>/usr/share/backgrounds/mate/nature/FreshFlower.jpg</filename>
+		<options>zoom</options>
+		<artist>Jens Ansorg</artist>
+	</wallpaper>
+	<wallpaper deleted="false">
+		<!-- Translators: This is the name of a desktop background image that shows a green meadow.
+		  You might want to translate it into the equivalent words of your language  -->
+		<name>Green Meadow</name>
+		<filename>/usr/share/backgrounds/mate/nature/GreenMeadow.jpg</filename>
+		<options>zoom</options>
+		<artist>Timothy Schellhase, Lapo Calamandrei</artist>
+	</wallpaper>
+	<wallpaper deleted="false">
+		<name>Aqua</name>
+		<filename>/usr/share/backgrounds/mate/nature/Aqua.jpg</filename>
+		<options>zoom</options>
+		<artist>Charles Barbin</artist>
+	</wallpaper>
+	<wallpaper deleted="false">
+		<name>Dune</name>
+		<filename>/usr/share/backgrounds/mate/nature/Dune.jpg</filename>
+		<options>zoom</options>
+		<artist>Marco Cogoni, Hylke Bons</artist>
+	</wallpaper>
+	<wallpaper deleted="false">
+		<name>Storm</name>
+		<filename>/usr/share/backgrounds/mate/nature/Storm.jpg</filename>
+		<options>zoom</options>
+		<artist>Sergey Nazarov</artist>
+	</wallpaper>
+	<wallpaper deleted="false">
+		<name>Two Wings</name>
+		<filename>/usr/share/backgrounds/mate/nature/TwoWings.jpg</filename>
+		<options>zoom</options>
+		<artist>Charles Barbin, Hylke Bons</artist>
+	</wallpaper>
+	<wallpaper deleted="false">
+		<name>Lady Bird</name>
+		<filename>/usr/share/backgrounds/mate/nature/LadyBird.jpg</filename>
+		<options>zoom</options>
+		<artist>Charles Barbin</artist>
+	</wallpaper>
+	<wallpaper deleted="false">
+		<name>Blinds</name>
+		<filename>/usr/share/backgrounds/mate/nature/Blinds.jpg</filename>
+		<options>zoom</options>
+		<artist>Lennart Poettering, Andrea Cimitan</artist>
+	</wallpaper>
+	<wallpaper deleted="false">
+		<name>Rain Drops</name>
+		<filename>/usr/share/backgrounds/mate/nature/RainDrops.jpg</filename>
+		<options>zoom</options>
+		<artist>blackcombwhistler07</artist>
+	</wallpaper>
+	<wallpaper deleted="false">
+		<name>Yellow Flower</name>
+		<filename>/usr/share/backgrounds/mate/nature/YellowFlower.jpg</filename>
+		<options>zoom</options>
+		<artist>Charles Barbin</artist>
+	</wallpaper>
+	<wallpaper deleted="false">
+		<name>Garden</name>
+		<filename>/usr/share/backgrounds/mate/nature/Garden.jpg</filename>
+		<options>zoom</options>
+		<artist>Charles Barbin</artist>
+	</wallpaper>
+	<wallpaper deleted="false">
+		<name>Wood</name>
+		<filename>/usr/share/backgrounds/mate/nature/Wood.jpg</filename>
+		<options>zoom</options>
+		<artist>Jean-François Fortin Tam</artist>
+	</wallpaper>
+	+	<wallpaper deleted="false">
+		<_name>1</_name>
+		<filename>@datadir@/backgrounds/mate/nature/1.jpg</filename>
+		<options>zoom</options>
+	</wallpaper>
+	<wallpaper deleted="false">
+		<_name>2</_name>
+		<filename>@datadir@/backgrounds/mate/nature/2.jpg</filename>
+		<options>zoom</options>
+	</wallpaper>
+	<wallpaper deleted="false">
+		<_name>3</_name>
+		<filename>@datadir@/backgrounds/mate/nature/3.jpg</filename>
+		<options>zoom</options>
+	</wallpaper>
+	<wallpaper deleted="false">
+		<_name>4</_name>
+		<filename>@datadir@/backgrounds/mate/nature/4.jpg</filename>
+		<options>zoom</options>
+	</wallpaper>
+	<wallpaper deleted="false">
+		<_name>5</_name>
+		<filename>@datadir@/backgrounds/mate/nature/5.jpg</filename>
+		<options>zoom</options>
+	</wallpaper>
+	<wallpaper deleted="false">
+		<_name>6</_name>
+		<filename>@datadir@/backgrounds/mate/nature/6.jpg</filename>
+		<options>zoom</options>
+	</wallpaper>
+	<wallpaper deleted="false">
+		<_name>7</_name>
+		<filename>@datadir@/backgrounds/mate/nature/7.jpg</filename>
+		<options>zoom</options>
+	</wallpaper>
+	<wallpaper deleted="false">
+		<_name>8</_name>
+		<filename>@datadir@/backgrounds/mate/nature/8.jpg</filename>
+		<options>zoom</options>
+	</wallpaper>
+</wallpapers>
