--- a/libswscale/ppc/swscale_vsx.c
+++ b/libswscale/ppc/swscale_vsx.c
@@ -154,7 +154,7 @@
     }
 }
 
-static void yuv2plane1_nbps_vsx(const int16_t *src, uint16_t *dest, int dstW,
+static av_always_inline void yuv2plane1_nbps_vsx(const int16_t *src, uint16_t *dest, int dstW,
                            int big_endian, int output_bits)
 {
     const int dst_u = -(uintptr_t)dest & 7;
@@ -273,7 +273,7 @@
     }
 }
 
-static void yuv2plane1_16_vsx(const int32_t *src, uint16_t *dest, int dstW,
+static av_always_inline void yuv2plane1_16_vsx(const int32_t *src, uint16_t *dest, int dstW,
                            int big_endian, int output_bits)
 {
     const int dst_u = -(uintptr_t)dest & 7;

