diff -Nuar a/data/pcb.desktop.in b/data/pcb.desktop.in
--- a/data/pcb.desktop.in	2018-01-17 22:10:19.000000000 +0300
+++ b/data/pcb.desktop.in	2018-04-04 20:21:09.706912861 +0300
@@ -2,9 +2,10 @@
 Version=1.0
 _Name=gEDA PCB
 _GenericName=gEDA PCB Layout Editor
+GenericName[tr]=Baskı Devre Tasarlayıcısı
 _Comment=Create and edit printed circuit board layouts
 Type=Application
 Exec=pcb %f
 Icon=pcb
 MimeType=application/x-pcb-layout;application/x-pcb-footprint;
-Categories=Engineering;Electronics;Development;
+Categories=Application;Science;Engineering;Electronics;
