--- a/configure.ac	2019-06-02 14:13:16.000000000 +0300
+++ b/configure.ac	2020-05-26 18:34:07.592350821 +0300
@@ -141,7 +141,7 @@
 if $try_cdk; then
         CPPFLAGS_save="$CPPFLAGS"
         CPPFLAGS="-I$cdk_prefix/include $CPPFLAGS"
-        AC_CHECK_HEADER([cdk/cdk.h], [
+        AC_CHECK_HEADER([cdk.h], [
                 have_cdk=:
                 AC_PATH_PROG([CDK_CONFIG], [cdk5-config])
                 if test -n "${CDK_CONFIG}"; then
--- a/gphoto2/gphoto2-cmd-config.c	2015-08-01 14:20:27.000000000 +0300
+++ b/gphoto2/gphoto2-cmd-config.c	2020-05-26 18:52:32.905596310 +0300
@@ -29,7 +29,7 @@
 
 #include <gphoto2/gphoto2-widget.h>
 
-#include <cdk/cdk.h>
+#include <cdk.h>
 
 #ifndef MAX
 # define MAX(a, b) ((a) > (b) ? (a) : (b))
