--- swig/php/gdal_wrap.cpp~	2012-10-09 08:58:28.000000000 +0800
+++ swig/php/gdal_wrap.cpp	2012-10-12 12:19:02.201201302 +0800
@@ -932,7 +932,7 @@
   p = value->ptr;
   if (type==-1) return NULL;
 
-  type_name=zend_rsrc_list_get_rsrc_type(z->value.lval TSRMLS_CC);
+  type_name=(char *)zend_rsrc_list_get_rsrc_type(z->value.lval TSRMLS_CC);
 
   return SWIG_ZTS_ConvertResourceData(p, type_name, ty TSRMLS_CC);
 }
