diff -Nuar a/src/VBox/Additions/common/crOpenGL/Makefile.kmk b/src/VBox/Additions/common/crOpenGL/Makefile.kmk
--- a/src/VBox/Additions/common/crOpenGL/Makefile.kmk	2017-10-18 10:06:45.000000000 +0300
+++ b/src/VBox/Additions/common/crOpenGL/Makefile.kmk	2017-11-20 12:49:55.626419756 +0300
@@ -69,17 +69,12 @@
 if1of ($(KBUILD_TARGET), linux solaris freebsd)
  ifndef VBOX_USE_SYSTEM_GL_HEADERS
   VBoxOGL_INCS     += \
-	$(VBOX_PATH_X11_ROOT)/libXdamage-1.1 \
-	$(VBOX_PATH_X11_ROOT)/libXcomposite-0.4.0 \
-	$(VBOX_PATH_X11_ROOT)/libXext-1.3.1 \
-	$(VBOX_PATH_X11_ROOT)/libXfixes-4.0.3 \
-	$(VBOX_PATH_X11_ROOT)/damageproto-1.1.0 \
-	$(VBOX_PATH_X11_ROOT)/compositeproto-0.4 \
-	$(VBOX_PATH_X11_ROOT)/fixesproto-4.0 \
-	$(VBOX_PATH_X11_ROOT)/libx11-1.1.5-other \
-	$(VBOX_PATH_X11_ROOT)/xextproto-7.1.1 \
-	$(VBOX_PATH_X11_ROOT)/xproto-7.0.18 \
-	$(VBOX_GL_INCS)
+	/usr/include/x11 \
+	/usr/include/xorg \
+	/usr/include/pixman-1 \
+	$(VBOX_GL_INCS) \
+	/usr/include/drm \
+	/usr/include/libdrm
  endif
  VBoxOGL_DEFS     += VBOX_NO_NATIVEGL
 endif
@@ -219,10 +214,10 @@
   VBoxOGL_LIBS += Xcomposite Xdamage Xfixes Xext
  else
   VBoxOGL_LIBS += \
- 	$(PATH_STAGE_LIB)/libXcomposite.so \
- 	$(PATH_STAGE_LIB)/libXdamage.so \
- 	$(PATH_STAGE_LIB)/libXfixes.so \
- 	$(PATH_STAGE_LIB)/libXext.so
+ 	Xcomposite \
+ 	Xdamage \
+ 	Xfixes \
+ 	Xext
  endif
  ifdef VBoxOGL_FAKEDRI
   ifeq ($(KBUILD_TARGET), freebsd)
