--- VirtualBox-6.1.2/src/VBox/Frontends/VirtualBox/src/extradata/UIExtraDataManager.cpp.orig	2020-01-13 12:53:25.000000000 +0200
+++ VirtualBox-6.1.2/src/VBox/Frontends/VirtualBox/src/extradata/UIExtraDataManager.cpp	2020-01-15 22:34:18.352860666 +0200
@@ -2349,7 +2349,7 @@ QString UIExtraDataManager::preventBetaB
 bool UIExtraDataManager::applicationUpdateEnabled()
 {
     /* 'True' unless 'restriction' feature allowed: */
-    return !isFeatureAllowed(GUI_PreventApplicationUpdate);
+    return false;
 }
 
 QString UIExtraDataManager::applicationUpdateData()
