From ca745cdfc27b83b6bcef9f856d858a68c64429de Mon Sep 17 00:00:00 2001
From: Peter Jones <pjones@redhat.com>
Date: Tue, 8 Apr 2014 15:28:12 -0400
Subject: [PATCH] Add "install-all target" to top side of HAVE_FIRMWARE.

---
 Makefile | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/Makefile b/Makefile
index fb387c5..3429058 100644
--- a/Makefile
+++ b/Makefile
@@ -273,6 +273,16 @@ efi64:
 		FIRMWARE=EFI64 \
 		$(MAKECMDGOALS)
 
+install-all:
+
+install:
+
+netinstall:
+
+clean:
+
+all:
+
 else # ifeq($(HAVE_FIRMWARE),)
 
 all: all-local subdirs
-- 
1.9.3

