diff -ur cfitsio-4.2.0/configure cfitsio-4.2.0.rpath/configure
--- cfitsio-4.2.0/configure	2022-11-25 10:08:23.942637069 +0100
+++ cfitsio-4.2.0.rpath/configure	2022-11-25 10:10:13.213414398 +0100
@@ -5062,7 +5062,7 @@
         printf "%s\n" "#define _FILE_OFFSET_BITS 64" >>confdefs.h
 
         # LDFLAGS used by utilities:
-        LDFLAGS_BIN="$LDFLAGS_BIN -Wl,-rpath,\${CFITSIO_LIB}"
+        LDFLAGS_BIN="$LDFLAGS_BIN"
 	;;
     *mingw32*)
         { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for large file support" >&5
