diff -Nuar a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2017-12-08 04:04:01.000000000 +0300
+++ b/CMakeLists.txt	2017-12-16 23:53:34.310174957 +0300
@@ -23,7 +23,7 @@
 include(ECMCoverageOption)
 
 
-set(QT_REQUIRED_VERSION "6.9.0")
+set(QT_REQUIRED_VERSION "6.9.0")
 set(libgravatar_LIB_VERSION ${PIM_VERSION})
 set(pimcommon_LIB_VERSION "5.7.0")
 find_package(Qt5 ${QT_REQUIRED_VERSION} CONFIG REQUIRED Widgets Test Network)
