diff -Nuar a/gtk/__init__.py b/gtk/__init__.py
--- a/gtk/__init__.py	2011-04-01 13:37:25.000000000 +0300
+++ b/gtk/__init__.py	2022-08-24 10:43:09.899937881 +0300
@@ -37,7 +37,7 @@
 if 'gtk._gtk' in sys.modules:
     _gtk = sys.modules['gtk._gtk']
 else:
-    from gtk import _gtk
+    from gtk import gtk
 
 import gdk
 
@@ -124,7 +124,7 @@
 from gtk._gtk import *
 
 # Make PyGTK interactive
-set_interactive(1)
+gtk.set_interactive(1)
 
 # # For testing, so you can just turn off dynamicnamespace in gtk.override
 # if hasattr(_gtk, '_get_symbol_names'):

