From 5bf672b373c93a8ebd4040d934875969e697a470 Mon Sep 17 00:00:00 2001
From: Alexander Sokolov <sokoloff.a@gmail.com>
Date: Wed, 3 Aug 2022 17:43:29 +0300
Subject: [PATCH] Close issue #131: Tooltips of sidebar button.

---
 gui/controls.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/gui/controls.cpp b/gui/controls.cpp
index 4caaf6e..5d70140 100644
--- a/gui/controls.cpp
+++ b/gui/controls.cpp
@@ -53,7 +53,7 @@ ToolButton::ToolButton(const QIcon &icon, QWidget *parent) :
     QToolButton(parent)
 {
     setAutoRaise(true);
-    setStyleSheet("border: none;");
+    setStyleSheet("QToolButton{ border: none; }");
     setFixedWidth(ToolButton::sizeHint().width());
     setIcon(icon);
     setPopupMode(QToolButton::InstantPopup);
