--- setup_old.py	2017-02-01 15:02:25.088620379 +0300
+++ setup.py	2017-02-01 15:10:53.517611419 +0300
@@ -5,16 +5,8 @@
 import os
 import sys
 
-import pip
-
 from setuptools import setup, find_packages
 
-if 'docker-py' in [x.project_name for x in pip.get_installed_distributions()]:
-    print(
-        'ERROR: "docker-py" needs to be uninstalled before installing this'
-        ' package:\npip uninstall docker-py', file=sys.stderr
-    )
-    sys.exit(1)
 
 ROOT_DIR = os.path.dirname(__file__)
 SOURCE_DIR = os.path.join(ROOT_DIR)
