diff -Nuar a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2017-12-08 04:04:46.000000000 +0300
+++ b/CMakeLists.txt	2017-12-16 23:24:11.263302169 +0300
@@ -29,7 +29,7 @@
 set(pimcommon_LIB_VERSION ${PIM_VERSION})
 set(AKONADICONTACT_LIB_VERSION "5.7.0")
 set(libkdepim_LIB_VERSION "5.7.0")
-set(QT_REQUIRED_VERSION "6.6.3")
+set(QT_REQUIRED_VERSION "6.6.3")
 set(kmime_LIB_VERSION "5.7.0")
 set(kimap_LIB_VERSION "5.7.0")
 set(kcontacts_LIB_VERSION "5.7.0")
