diff -Nuar a/src/modules/initcpiocfg/main.py b/src/modules/initcpiocfg/main.py
--- a/src/modules/initcpiocfg/main.py	2024-07-04 00:05:30.000000000 +0300
+++ b/src/modules/initcpiocfg/main.py	2026-04-19 21:46:24.678005799 +0300
@@ -34,7 +34,7 @@
     @return True if plymouth exists in the target, False otherwise
     """
     # Used to only check existence of path /usr/bin/plymouth in target
-    return target_env_call(["sh", "-c", "which plymouth"]) == 0
+    return target_env_call(["sh", "-c", "which plymouth || test -f /bin/plymouth"]) == 0
 
 
 class cpuinfo(object):
@@ -244,8 +244,6 @@
 
     if uses_btrfs:
         modules.append("crc32c-intel" if cpuinfo().is_intel else "crc32c")
-    else:
-        hooks.append("fsck")
 
     # Modify according to the keys in the configuration
     hooks = [h for h in (hooks_prepend + hooks + hooks_append) if h not in hooks_remove]
