diff -Nuar a/src/modules/locale/LocalePage.cpp b/src/modules/locale/LocalePage.cpp
--- a/src/modules/locale/LocalePage.cpp	2017-10-23 23:22:25.000000000 +0300
+++ b/src/modules/locale/LocalePage.cpp	2017-11-05 16:41:49.935761099 +0300
@@ -320,7 +320,7 @@
                      << ". Assuming the supported languages are already built into "
                         "the locale archive.";
             QProcess localeA;
-            localeA.start( "locale", QStringList() << "-a" );
+            localeA.start( "cat", QStringList() << "/etc/locale.gen" );
             localeA.waitForFinished();
             ba = localeA.readAllStandardOutput();
         }
