diff -Nuar a/src/modules/users/SetHostNameJob.cpp b/src/modules/users/SetHostNameJob.cpp
--- a/src/modules/users/SetHostNameJob.cpp	2020-02-28 01:26:29.000000000 +0300
+++ b/src/modules/users/SetHostNameJob.cpp	2020-03-21 14:05:55.674849075 +0300
@@ -64,7 +64,7 @@
 setFileHostname( const QString& hostname )
 {
     return CalamaresUtils::System::instance()->createTargetFile(
-        QStringLiteral( "/etc/hostname" ), ( hostname + '\n' ).toUtf8(), WriteMode::Overwrite );
+        QStringLiteral( "/etc/env.d/01hostname" ), ( hostname + '\n' ).toUtf8(), WriteMode::Overwrite );
 }
 
 STATICTEST bool
