--- src/base/omx_base_component.c	2011-01-12 08:53:26.000000000 +0100
+++ src/base/omx_base_component.c-gil	2011-12-18 20:33:16.654326686 +0100
@@ -914,7 +914,7 @@
   if (ComponentParameterStructure == NULL) {
     return OMX_ErrorBadParameter;
   }
-  switch(nParamIndex) {
+  switch((int) nParamIndex) {
   case OMX_IndexParameterThreadsID:
 	    if ((err = checkHeader(ComponentParameterStructure, sizeof(OMX_PARAM_BELLAGIOTHREADS_ID))) != OMX_ErrorNone) {
 	      break;
