diff -Nuar a/configure.ac b/configure.ac
--- a/configure.ac	2020-01-14 22:24:39.000000000 +0300
+++ b/configure.ac	2020-01-14 22:17:15.664924409 +0300
@@ -5,7 +5,7 @@
 AC_PREREQ([2.59])
 
 AC_CONFIG_HEADERS([config.h])
-CFLAGS="${CFLAGS} -Wall"
+CFLAGS="${CFLAGS} -Wall -Werror -Wno-error=unused-but-set-variable -Wno-error=unused-but-set-parameter"
 
 ################################################################################
 # Set the shared versioning info, according to section 6.3 of the libtool info #
