diff -ru LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9.orig/cmake/dependencies.cmake LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9/cmake/dependencies.cmake
--- LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9.orig/cmake/dependencies.cmake	2025-10-16 21:52:17.599168371 +0200
+++ LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9/cmake/dependencies.cmake	2025-10-16 21:53:07.900119402 +0200
@@ -7,7 +7,6 @@
   filesystem
   iostreams
   regex
-  system
   thread
   REQUIRED
 )
diff -ru LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9.orig/src/contrib/CMakeLists.txt LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9/src/contrib/CMakeLists.txt
--- LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9.orig/src/contrib/CMakeLists.txt	2025-10-16 21:52:17.620913579 +0200
+++ LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9/src/contrib/CMakeLists.txt	2025-10-16 21:55:39.509895982 +0200
@@ -67,7 +67,6 @@
     Boost::filesystem
     Boost::iostreams
     Boost::regex
-    Boost::system
     Boost::thread
     ZLIB::ZLIB
     lucene++::lucene++)
diff -ru LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9.orig/src/core/CMakeLists.txt LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9/src/core/CMakeLists.txt
--- LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9.orig/src/core/CMakeLists.txt	2025-10-16 21:52:17.630332868 +0200
+++ LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9/src/core/CMakeLists.txt	2025-10-16 21:55:39.510349119 +0200
@@ -57,7 +57,6 @@
     Boost::filesystem
     Boost::iostreams
     Boost::regex
-    Boost::system
     Boost::thread
     ZLIB::ZLIB
 )
diff -ru LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9.orig/src/demo/deletefiles/CMakeLists.txt LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9/src/demo/deletefiles/CMakeLists.txt
--- LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9.orig/src/demo/deletefiles/CMakeLists.txt	2025-10-16 21:52:17.648223127 +0200
+++ LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9/src/demo/deletefiles/CMakeLists.txt	2025-10-16 21:55:39.510708458 +0200
@@ -38,7 +38,6 @@
     Boost::filesystem
     Boost::iostreams
     Boost::regex
-    Boost::system
     Boost::thread
     ZLIB::ZLIB
     lucene++::lucene++
diff -ru LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9.orig/src/demo/indexfiles/CMakeLists.txt LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9/src/demo/indexfiles/CMakeLists.txt
--- LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9.orig/src/demo/indexfiles/CMakeLists.txt	2025-10-16 21:52:17.648476446 +0200
+++ LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9/src/demo/indexfiles/CMakeLists.txt	2025-10-16 21:55:39.510875032 +0200
@@ -39,7 +39,6 @@
     Boost::filesystem
     Boost::iostreams
     Boost::regex
-    Boost::system
     Boost::thread
     ZLIB::ZLIB
     lucene++::lucene++
diff -ru LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9.orig/src/demo/searchfiles/CMakeLists.txt LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9/src/demo/searchfiles/CMakeLists.txt
--- LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9.orig/src/demo/searchfiles/CMakeLists.txt	2025-10-16 21:52:17.648728577 +0200
+++ LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9/src/demo/searchfiles/CMakeLists.txt	2025-10-16 21:55:39.511044539 +0200
@@ -38,7 +38,6 @@
     Boost::filesystem
     Boost::iostreams
     Boost::regex
-    Boost::system
     Boost::thread
     ZLIB::ZLIB
     lucene++::lucene++
diff -ru LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9.orig/src/test/CMakeLists.txt LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9/src/test/CMakeLists.txt
--- LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9.orig/src/test/CMakeLists.txt	2025-10-16 21:52:17.649059769 +0200
+++ LucenePlusPlus-b545dba48639ac5ecb691f49fb2748c847cf5be9/src/test/CMakeLists.txt	2025-10-16 21:55:39.511226059 +0200
@@ -59,7 +59,6 @@
     Boost::filesystem
     Boost::iostreams
     Boost::regex
-    Boost::system
     Boost::thread
     ZLIB::ZLIB
     gtest_main
