diff -Nuar "a/test cases/frameworks/15 llvm/meson.build" "b/test cases/frameworks/15 llvm/meson.build"
--- "a/test cases/frameworks/15 llvm/meson.build"	2017-10-08 22:22:05.000000000 +0300
+++ "b/test cases/frameworks/15 llvm/meson.build"	2017-12-07 18:31:23.786683687 +0300
@@ -17,5 +17,5 @@
   llvm_dep,
   dependency('zlib'),
   meson.get_compiler('c').find_library('dl', required : false),
-  dependency('tinfo'),
+  dependency('ncursesw'),
   ])
diff -Nuar "a/test cases/frameworks/4 qt/meson.build" "b/test cases/frameworks/4 qt/meson.build"
--- "a/test cases/frameworks/4 qt/meson.build"	2017-10-08 21:47:57.000000000 +0300
+++ "b/test cases/frameworks/4 qt/meson.build"	2017-12-07 18:33:49.091673203 +0300
@@ -77,8 +77,9 @@
       moc_headers : 'plugin/plugin.h',
       include_directories : plugin_includes
     )
-    plugin = library('plugin', 'plugin/plugin.cpp', pluginpreprocess,
+    plugin = library(qt + 'plugin', 'plugin/plugin.cpp', pluginpreprocess,
           include_directories : plugin_includes,
           dependencies : qtcore)
+    endif      
   endif
 endforeach
