diff -Nuar a/src/incidencechanger_p.cpp b/src/incidencechanger_p.cpp
--- a/src/incidencechanger_p.cpp	2017-12-04 13:20:36.000000000 +0300
+++ b/src/incidencechanger_p.cpp	2017-12-16 22:01:34.956659792 +0300
@@ -189,7 +189,9 @@
                                            << "isValid = "  << mDefaultCollection.isValid()
                                            << "has ACLs = " << hasRights(mDefaultCollection, ChangeTypeCreate);
         // else fallthrough, and ask the user.
+#if QT_VERSION >= QT_VERSION_CHECK(5,8,0)
         Q_FALLTHROUGH();
+#endif
         case DestinationPolicyAsk: {
             mPendingCreations << change;
             loadCollections(); // Now we wait, collections are being loaded async
