From ce58517e546f3f3433931609683b5b4e455bbb7f Mon Sep 17 00:00:00 2001
From: Stephan Hartmann <stha09@googlemail.com>
Date: Fri, 5 Nov 2021 15:48:05 +0000
Subject: [PATCH] GCC: fix undefined reference to ScrollView::SetContents()

---
 ui/views/controls/scroll_view.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/ui/views/controls/scroll_view.h b/ui/views/controls/scroll_view.h
index 3bd5a8bf..f850728c 100644
--- a/ui/views/controls/scroll_view.h
+++ b/ui/views/controls/scroll_view.h
@@ -386,6 +386,9 @@
   ScrollViewCallbackList on_contents_scroll_ended_;
 };
 
+// Required for WebAppUrlHandlerIntentPickerView
+template View* ScrollView::SetContents<View>(std::unique_ptr<View> a_view);
+
 BEGIN_VIEW_BUILDER(VIEWS_EXPORT, ScrollView, View)
 VIEW_BUILDER_VIEW_TYPE_PROPERTY(View, Contents)
 VIEW_BUILDER_PROPERTY(ui::LayerType, ContentsLayerType)
