diff -Nuar a/runtime/syntax/grub.vim b/runtime/syntax/grub.vim
--- a/runtime/syntax/grub.vim	2019-08-14 22:12:05.000000000 +0300
+++ b/runtime/syntax/grub.vim	2019-10-29 14:29:07.740399833 +0300
@@ -27,7 +27,7 @@
 
 syn cluster grubCommands      contains=grubCommand,grubTitleCommand
 
-syn keyword grubCommand       contained default fallback hiddenmenu timeout
+syn keyword grubCommand       contained default fallback hiddenmenu timeout splashimage background gfxmenu
 
 syn keyword grubTitleCommand  contained title nextgroup=grubTitle skipwhite
