--- src/lib/preferences/preferences.cpp.orig	2014-11-18 17:52:54.635782052 -0500
+++ src/lib/preferences/preferences.cpp	2014-11-18 17:52:55.129103922 -0500
@@ -140,7 +140,7 @@
     Settings settings;
     //GENERAL URLs
     settings.beginGroup("Web-URL-Settings");
-    m_homepage = settings.value("homepage", QUrl(QSL("qupzilla:start"))).toUrl();
+    m_homepage = settings.value("homepage", QUrl(QSL("qupzilla:speeddial"))).toUrl();
     m_newTabUrl = settings.value("newTabUrl", QUrl(QSL("qupzilla:speeddial"))).toUrl();
     ui->homepage->setText(m_homepage.toEncoded());
     ui->newTabUrl->setText(m_newTabUrl.toEncoded());