diff -ur QupZilla-2.2.5/src/lib/webengine/webpage.cpp QupZilla-2.2.5-qtbug-65523-workaround/src/lib/webengine/webpage.cpp
--- QupZilla-2.2.5/src/lib/webengine/webpage.cpp	2018-01-28 16:37:13.000000000 +0100
+++ src/lib/webengine/webpage.cpp	2018-02-01 04:34:19.159005227 +0100
@@ -102,7 +102,7 @@
     });
 
     // Workaround for broken load started/finished signals in QtWebEngine 5.10
-    if (qstrcmp(qVersion(), "5.10.0") == 0) {
+    {
         connect(this, &QWebEnginePage::loadProgress, this, [this](int progress) {
             if (progress == 100) {
                 emit loadFinished(true);
