diff -ur QupZilla-2.2.0/src/lib/preferences/preferences.cpp QupZilla-2.2.0-native-scrollbars/src/lib/preferences/preferences.cpp
--- QupZilla-2.2.0/src/lib/preferences/preferences.cpp	2017-10-08 16:05:56.000000000 +0200
+++ src/lib/preferences/preferences.cpp	2017-10-16 05:34:25.738624011 +0200
@@ -278,7 +278,7 @@
     ui->wheelScroll->setValue(settings.value("wheelScrollLines", qApp->wheelScrollLines()).toInt());
     ui->xssAuditing->setChecked(settings.value("XSSAuditing", false).toBool());
     ui->printEBackground->setChecked(settings.value("PrintElementBackground", true).toBool());
-    ui->useNativeScrollbars->setChecked(settings.value("UseNativeScrollbars", false).toBool());
+    ui->useNativeScrollbars->setChecked(settings.value("UseNativeScrollbars", true).toBool());
 
     foreach (int level, WebView::zoomLevels()) {
         ui->defaultZoomLevel->addItem(QString("%1%").arg(level));
diff -ur QupZilla-2.2.0/src/lib/webengine/webscrollbarmanager.cpp QupZilla-2.2.0-native-scrollbars/src/lib/webengine/webscrollbarmanager.cpp
--- QupZilla-2.2.0/src/lib/webengine/webscrollbarmanager.cpp	2017-10-08 16:05:56.000000000 +0200
+++ src/lib/webengine/webscrollbarmanager.cpp	2017-10-16 05:33:36.152339717 +0200
@@ -109,7 +109,7 @@
 
 void WebScrollBarManager::loadSettings()
 {
-    m_enabled = Settings().value(QSL("Web-Browser-Settings/UseNativeScrollbars"), false).toBool();
+    m_enabled = Settings().value(QSL("Web-Browser-Settings/UseNativeScrollbars"), true).toBool();
 
     if (!m_enabled) {
         for (WebView *view : m_scrollbars.keys()) {
