--- src/lib/plugins/speeddial.cpp	2016-02-29 11:18:45.000000000 -0500
+++ src/lib/plugins/speeddial.cpp	2016-03-04 17:25:39.726171052 -0500
@@ -58,15 +58,18 @@
     setBackgroundImage(settings.value("background", QString()).toString());
     m_backgroundImageSize = settings.value("backsize", "auto").toString();
     m_maxPagesInRow = settings.value("pagesrow", 4).toInt();
-    m_sizeOfSpeedDials = settings.value("sdsize", 231).toInt();
+    m_sizeOfSpeedDials = settings.value("sdsize", 225).toInt();
     m_sdcentered = settings.value("sdcenter", false).toBool();
     settings.endGroup();
 
     if (allPages.isEmpty()) {
-        allPages = "url:\"https://www.qupzilla.com\"|title:\"QupZilla\";"
-                   "url:\"http://blog.qupzilla.com\"|title:\"QupZilla Blog\";"
-                   "url:\"https://github.com/QupZilla/qupzilla\"|title:\"QupZilla GitHub\";"
-                   "url:\"https://duckduckgo.com\"|title:\"DuckDuckGo\";";
+        allPages = "url:\"http://www.pisilinux.org\"|title:\"Pisilinux - Hoşgeldiniz\";"
+                   "url:\"http://forum.pisilinux.org\"|title:\"Forum\";"
+                   "url:\"http://wiki.pisilinux.org\"|title:\"Pisilinux Wiki\";"
+                   "url:\"http://www.youtube.com/html5\"|title:\"YouTube-html5\";"
+                   "url:\"http://www.google.com\"|title:\"Google\";"
+                   "url:\"http://www.qupzilla.com\"|title:\"QupZilla\";"
+                   "url:\"https://github.com/pisilinux\"|title:\"Github\";";
     }
     changed(allPages);
 
