diff -Nuar a/libnemo-private/nemo-file-utilities.c b/libnemo-private/nemo-file-utilities.c
--- a/libnemo-private/nemo-file-utilities.c	2024-11-27 14:32:29.000000000 +0300
+++ b/libnemo-private/nemo-file-utilities.c	2024-11-28 11:08:10.118509299 +0300
@@ -1106,7 +1106,7 @@
 					      GSM_INTERFACE,
 					      "Inhibit",
 					      g_variant_new ("(susu)",
-							     "Nemo",
+							     "nemo",
 							     (guint) 0,
 							     message,
 							     (guint) (INHIBIT_LOGOUT | INHIBIT_SUSPEND)),
