diff -Nuar firefox-42.0.OSmanOS/config/baseconfig.mk firefox-42.0/config/baseconfig.mk
--- firefox-42.0.OSmanOS/config/baseconfig.mk	2015-10-30 00:17:52.000000000 +0200
+++ firefox-42.0/config/baseconfig.mk	2015-11-06 23:44:25.232724913 +0200
@@ -4,7 +4,7 @@
 # whether a normal build is happening or whether the check is running.
 includedir := $(includedir)/$(MOZ_APP_NAME)-$(MOZ_APP_VERSION)
 idldir = $(datadir)/idl/$(MOZ_APP_NAME)-$(MOZ_APP_VERSION)
-installdir = $(libdir)/$(MOZ_APP_NAME)-$(MOZ_APP_VERSION)
+installdir = $(libdir)/$(MOZ_APP_NAME)
 sdkdir = $(libdir)/$(MOZ_APP_NAME)-devel-$(MOZ_APP_VERSION)
 ifndef TOP_DIST
 TOP_DIST = dist
