diff -Nuar a/toolkit/moz.configure b/toolkit/moz.configure
--- a/toolkit/moz.configure	2017-11-16 13:38:11.571875852 +0300
+++ b/toolkit/moz.configure	2017-11-16 13:35:29.679887533 +0300
@@ -618,7 +618,7 @@
 @imports('os')
 @imports('subprocess')
 def llvm_config_paths(host):
-    llvm_supported_versions = ['6.0', '5.0', '4.0', '3.9']
+    llvm_supported_versions = ['6.0', '5.0', '4.0', '3.9', '3.8']
     llvm_config_progs = []
     for version in llvm_supported_versions:
         llvm_config_progs += [
@@ -662,9 +662,9 @@
 
 with only_when(building_stylo_bindgen):
     option('--with-libclang-path', nargs=1,
-           help='Absolute path to a directory containing Clang/LLVM libraries for Stylo (version 3.9.x or above)')
+           help='Absolute path to a directory containing Clang/LLVM libraries for Stylo (version 3.8.x or above)')
     option('--with-clang-path', nargs=1,
-           help='Absolute path to a Clang binary for Stylo bindgen (version 3.9.x or above)')
+           help='Absolute path to a Clang binary for Stylo bindgen (version 3.8.x or above)')
 
     def invoke_llvm_config(llvm_config, *options):
         '''Invoke llvm_config with the given options and return the first line of
@@ -675,7 +675,7 @@
     @imports(_from='textwrap', _import='dedent')
     def check_minimum_llvm_config_version(llvm_config):
         version = Version(invoke_llvm_config(llvm_config, '--version'))
-        min_version = Version('3.9.0')
+        min_version = Version('3.8.0')
         if version < min_version:
             die(dedent('''\
             llvm installation {} is incompatible with Stylo bindgen.
