diff -Nuar a/toolkit/moz.configure b/toolkit/moz.configure
--- a/toolkit/moz.configure	2017-09-14 23:16:01.000000000 +0300
+++ b/toolkit/moz.configure	2017-11-13 11:29:08.953620612 +0300
@@ -649,6 +649,8 @@
         'llvm-config40',
         'llvm-config-3.9',
         'llvm-config39',
+        'llvm-config-3.8',
+        'llvm-config38',
         'llvm-config',
     ]
 
@@ -699,7 +701,7 @@
     @imports(_from='textwrap', _import='dedent')
     def check_minimum_llvm_config_version(llvm_config):
         version = Version(invoke_llvm_config(llvm_config, '--version'))
-        min_version = Version('3.9.0')
+        min_version = Version('3.8.0')
         if version < min_version:
             die(dedent('''\
             llvm installation {} is incompatible with Stylo bindgen.
