diff -Nuar a/build/moz.configure/rust.configure b/build/moz.configure/rust.configure
--- a/build/moz.configure/rust.configure	2018-01-11 23:16:54.000000000 +0300
+++ b/build/moz.configure/rust.configure	2018-01-20 11:41:47.109584234 +0300
@@ -58,7 +58,7 @@
         You can install rust by running './mach bootstrap'
         or by directly running the installer from https://rustup.rs/
         '''))
-    rustc_min_version = Version('1.21.0')
+    rustc_min_version = Version('1.20.0')
     cargo_min_version = Version('0.{}'.format(rustc_min_version.minor + 1))
 
     version = rustc_info.version
