From b021f9835185ed262f203f83b9e6643d6374f9be Mon Sep 17 00:00:00 2001
From: Christophe Giboudeaux <christophe@krop.fr>
Date: Thu, 4 Feb 2021 22:42:13 +0100
Subject: [PATCH] Fix build failure with GCC10

---
 libexplain/output/stderr.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libexplain/output/stderr.h b/libexplain/output/stderr.h
index 5494880..7956cfc 100644
--- a/libexplain/output/stderr.h
+++ b/libexplain/output/stderr.h
@@ -22,7 +22,7 @@
 
 #include <libexplain/output.h>
 
-explain_output_t explain_output_static_stderr;
+extern explain_output_t explain_output_static_stderr;
 
 #endif /* LIBEXPLAIN_OUTPUT_STDERR_H */
 /* vim: set ts=8 sw=4 et : */
-- 
2.30.0

