--- Vulkan-ValidationLayers-1.1.127.orig/layers/gpu_validation.cpp  2019-11-05 22:57:02.000000000 +0300
+++ Vulkan-ValidationLayers-1.1.127/layers/gpu_validation.cpp      2019-11-08 19:58:57.729430476 +0300
@@ -26,9 +26,9 @@
 #define VMA_STATIC_VULKAN_FUNCTIONS 0
 #include "gpu_validation.h"
 #include "shader_validation.h"
-#include "spirv-tools/libspirv.h"
-#include "spirv-tools/optimizer.hpp"
-#include "spirv-tools/instrument.hpp"
+#include <spirv-tools/libspirv.h>
+#include <spirv-tools/optimizer.hpp>
+#include <spirv-tools/instrument.hpp>
 #include <SPIRV/spirv.hpp>
 #include <algorithm>
 #include <regex>
 
