From 8ddd2fb4909bf6ed1a3506723126432da8fcf0c4 Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Tue, 29 May 2012 20:44:18 +0100
Subject: [PATCH 3/8] configure: Allow user defined C compiler flags.

---
 configure | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/configure b/configure
index 1316b3c1e..53f45f85b 100755
--- a/configure
+++ b/configure
@@ -2050,6 +2050,10 @@ if $flat_float_array; then
   echo "#define FLAT_FLOAT_ARRAY" >> m.h
 fi
 
+# Allow user defined C Compiler flags
+bytecccompopts="$bytecccompopts $CFLAGS"
+nativecccompopts="$nativecccompopts $CFLAGS"
+
 # Finish generated files
 
 cclibs="$cclibs $mathlib"
-- 
2.17.1

