From 57b827bfdd4d0bed7ec7e824fc4477ed9111f9c9 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Nicolas=20L=C3=A9cureuil?= <neoclust@mageia.org>
Date: Sun, 24 May 2020 15:30:34 +0200
Subject: [PATCH 1/2] Use openbox by default

---
 config/session.conf | 1 +
 1 file changed, 1 insertion(+)

diff --git a/config/session.conf b/config/session.conf
index a9d06d6..7c44ac9 100644
--- a/config/session.conf
+++ b/config/session.conf
@@ -1,5 +1,6 @@
 [General]
 leave_confirmation=true
+window_manager=openbox
 
 [Environment]
 GTK_CSD=0
-- 
2.26.2

