diff -Nuar a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2017-12-08 04:03:55.000000000 +0300
+++ b/CMakeLists.txt	2017-12-16 22:59:16.160410049 +0300
@@ -30,7 +30,7 @@
 
 set(grantleetheme_LIB_VERSION ${PIM_VERSION})
 
-set(QT_REQUIRED_VERSION "6.9.0")
+set(QT_REQUIRED_VERSION "6.9.0")
 set(GRANTLEE_VERSION "5.1")
 
 find_package(Qt5 ${QT_REQUIRED_VERSION} CONFIG REQUIRED Widgets Network)
