diff -Nuar a/kcfg/kscreenlockersettings.kcfg b/kcfg/kscreenlockersettings.kcfg
--- a/kcfg/kscreenlockersettings.kcfg	2018-05-01 15:48:58.000000000 +0300
+++ b/kcfg/kscreenlockersettings.kcfg	2018-09-29 16:30:26.586903201 +0300
@@ -6,7 +6,7 @@
   <kcfgfile name="kscreenlockerrc" />
   <group name="Daemon">
     <entry key="Autolock" type="Bool">
-      <default>true</default>
+      <default>false</default>
       <label>Screen lock enabled</label>
       <whatsthis>Sets whether the screen will be locked after the specified time.</whatsthis>
     </entry>
