--- chromium-browser-44.0.2403.107/third_party/ffmpeg/chromium/scripts/generate_gyp.py~	2015-07-27 07:21:59.000000000 +0300
+++ chromium-browser-44.0.2403.107/third_party/ffmpeg/chromium/scripts/generate_gyp.py	2015-07-27 07:22:03.210572817 +0300
@@ -757,7 +757,8 @@
     elif include_file_path in IGNORED_INCLUDE_FILES:
       continue
     else:
-      die('Failed to find file: ' + include_file_path)
+      print('Failed to find file: ' + include_file_path)
+      continue
 
     # At this point we've found the file. Check if its in our ignore list which
     # means that the list should be updated to no longer mention this file.
@@ -776,7 +777,8 @@
   licensecheck_path = os.path.abspath(os.path.join(
       source_root, 'third_party', 'devscripts', 'licensecheck.pl'));
   if not os.path.exists(licensecheck_path):
-    die('Could not find licensecheck.pl: ' + str(licensecheck_path))
+    print('Could not find licensecheck.pl: ' + str(licensecheck_path))
+    return True
 
   check_process = subprocess.Popen([licensecheck_path,
                                     '-l', '100', 
