diff -Nuar a/step/data/CMakeLists.txt b/step/data/CMakeLists.txt
--- a/step/data/CMakeLists.txt	2022-12-12 13:37:26.472349327 +0300
+++ b/step/data/CMakeLists.txt	2022-12-12 13:39:06.563098021 +0300
@@ -37,7 +37,7 @@
         separate_arguments(_extractxml_args_copy)
         file(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/${_lang})
         add_custom_command(OUTPUT ${_xmlfile_tr}
-            COMMAND ${PYTHON_EXECUTABLE} ${EXTRACTXML_EXECUTABLE} --translate --mo-file=${_mo_file} --output-dir=${CMAKE_CURRENT_BINARY_DIR}/${_lang} ${_extractxml_args_copy} ${_xmlfile}
+            COMMAND ${EXTRACTXML_EXECUTABLE} --translate --mo-file=${_mo_file} --output-dir=${CMAKE_CURRENT_BINARY_DIR}/${_lang} ${_extractxml_args_copy} ${_xmlfile}
             WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
             DEPENDS ${_xmlfile} ${_mo_file}
         )
