diff -Nuar a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2017-12-02 10:27:41.000000000 +0300
+++ b/CMakeLists.txt	2017-12-16 13:36:51.970738394 +0300
@@ -26,7 +26,7 @@
 include(ECMQtDeclareLoggingCategory)
 include(ECMCoverageOption)
 
-set(QT_REQUIRED_VERSION "5.8.0")
+set(QT_REQUIRED_VERSION "5.7.1")
 
 find_package(Qt5 ${QT_REQUIRED_VERSION} CONFIG REQUIRED Core Gui XmlPatterns Test)
 
