diff -Nuar a/kate/main.cpp b/kate/main.cpp
--- a/kate/main.cpp	2021-04-14 18:06:06.000000000 +0300
+++ b/kate/main.cpp	2021-05-09 23:14:47.943847179 +0300
@@ -38,31 +38,8 @@
 #include "qtsingleapplication/qtsingleapplication.h"
 #endif

-#ifndef Q_OS_WIN
-#include <unistd.h>
-#endif
-#include <iostream>
-
 int main(int argc, char **argv)
 {
-#ifndef Q_OS_WIN
-    // Prohibit using sudo or kdesu (but allow using the root user directly)
-    if (getuid() == 0) {
-        if (!qEnvironmentVariableIsEmpty("SUDO_USER")) {
-            std::cout << "Executing Kate with sudo is not possible due to unfixable security vulnerabilities. "
-                         "It is also not necessary; simply use Kate normally, and you will be prompted for "
-                         "elevated privileges when saving documents if needed."
-                      << std::endl;
-            return EXIT_FAILURE;
-        } else if (!qEnvironmentVariableIsEmpty("KDESU_USER")) {
-            std::cout << "Executing Kate with kdesu is not possible due to unfixable security vulnerabilities. "
-                         "It is also not necessary; simply use Kate normally, and you will be prompted for "
-                         "elevated privileges when saving documents if needed."
-                      << std::endl;
-            return EXIT_FAILURE;
-        }
-    }
-#endif
     /**
      * init resources from our static lib
      */
