diff -Nuar a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2017-12-08 04:03:56.000000000 +0300
+++ b/CMakeLists.txt	2017-12-17 23:22:46.040447149 +0300
@@ -30,7 +30,7 @@
 set(incidenceeditor_LIB_VERSION ${PIM_VERSION})
 set(AKONADI_MIMELIB_VERSION "5.7.0")
 set(KDEPIM_LIB_VERSION "5.7.0")
-set(QT_REQUIRED_VERSION "6.6.3")
+set(QT_REQUIRED_VERSION "6.6.3")
 set(kmime_LIB_VERSION "5.7.0")
 set(kmailtransport_LIB_VERSION "5.7.0")
 set(kldap_LIB_VERSION "5.7.0")
