--- historymanager5.py	2018-01-05 16:34:47.539037429 +0300
+++ historymanager/historymanager5.py	2018-01-17 12:25:47.299975000 +0300
@@ -33,7 +33,7 @@
     app = QApplication(sys.argv)
     app.setOrganizationName("history-manager")
     app.setApplicationName("history-manager")
-    app.setApplicationVersion("0.2.8.0")
+    app.setApplicationVersion("0.2.8b")
 
     locale = QLocale.system().name()
     translator = QTranslator(app)
@@ -42,9 +42,9 @@
 
     # Create Main Widget and make some settings
     mainWindow = MainManager(None, app= app)
-    mainWindow.show()
     mainWindow.resize(640, 480)
     mainWindow.setWindowIcon(QIcon.fromTheme("view-history"))
+    mainWindow.show()
 
     # Create connection for lastWindowClosed signal to quit app
     app.lastWindowClosed.connect(app.quit)
@@ -54,4 +54,4 @@
 
 
 if __name__ == '__main__':
-    main()
\ Dosya sonunda yenisatır yok.
+    main()
