diff --git a/bin/dig/dig.docbook b/bin/dig/dig.docbook
index 1079421..f11abd1 100644
--- a/bin/dig/dig.docbook
+++ b/bin/dig/dig.docbook
@@ -1177,6 +1177,39 @@ dig +qr www.isc.org any -x 127.0.0.1 isc.org ns +noqr
     </para>
   </refsection>
 
+  <refsection><info><title>RETURN CODES</title></info>
+    <para>
+      <command>Dig</command> return codes are:
+      <variablelist>
+        <varlistentry>
+          <listitem>
+            <para>0: Everything went well, including things like NXDOMAIN</para>
+          </listitem>
+        </varlistentry>
+        <varlistentry>
+          <listitem>
+            <para>1: Usage error</para>
+          </listitem>
+        </varlistentry>
+        <varlistentry>
+          <listitem>
+            <para>8: Couldn't open batch file</para>
+          </listitem>
+        </varlistentry>
+        <varlistentry>
+          <listitem>
+            <para>9: No reply from server</para>
+          </listitem>
+        </varlistentry>
+        <varlistentry>
+          <listitem>
+            <para>10: Internal error</para>
+          </listitem>
+        </varlistentry>
+      </variablelist>
+    </para>
+  </refsection>
+
   <refsection><info><title>FILES</title></info>
 
     <para><filename>/etc/resolv.conf</filename>
