diff -Nuar a/panda.py b/panda.py
--- a/panda.py	2016-04-14 14:38:58.000000000 +0300
+++ b/panda.py	2026-05-17 12:04:20.214999865 +0300
@@ -183,7 +178,7 @@
         status, modified = self.update_grub_default_entries(arg)
         if status in ["os", "generic", "vendor"] and modified:
             self.update_grub_cfg()
-            self.set_libGL(self.driver_name if self.driver_name in ["nvidia-current", "nvidia96", "nvidia173", "nvidia304", "fglrx", "nvidia340"] and status == "vendor" else "mesa")
+            self.set_libGL(self.driver_name if self.driver_name in ["nvidia-current", "nvidia96", "nvidia173", "nvidia304", "fglrx", "nvidia340"] and status == "vendor" else "libglvnd")
             if self.driver_name in ["nvidia-current", "nvidia96", "nvidia173", "nvidia304", "nvidia340"] and status == "vendor":
                 open(nvidia_blacklist_file, "w").write("blacklist nouveau\n")
             elif self.driver_name in ["nvidia-current", "nvidia96", "nvidia173", "nvidia304", "nvidia340"] and os.path.isfile(nvidia_blacklist_file):
@@ -195,14 +190,14 @@
     def set_libGL(self, arg):
         '''alternatives --set libGL /usr/lib/arg/libGL.so.1.2.0'''
         try:
-            retcode = call("alternatives --set libGL /usr/lib/%s/libGL.so.1.2.0" % arg, shell=True)
+            retcode = call("alternatives --set libGL /usr/lib/%s/libGL.so.1.7.0" % arg, shell=True)
         except OSError as e:
-            print >>sys.stderr, "alternatives --set libGL /usr/lib/%s/libGL.so.1.2.0 failed:" % arg, e
-        if not arg in ["mesa", "nvidia-current", "fglrx"]: return
+            print >>sys.stderr, "alternatives --set libGL /usr/lib/%s/libGL.so.1.7.0 failed:" % arg, e
+        if not arg in ["libglvnd", "nvidia-current", "fglrx"]: return
         try:
-            retcode = call("alternatives --set libGL-32bit /usr/lib32/%s/libGL.so.1.2.0" % arg, shell=True)
+            retcode = call("alternatives --set libGL-32bit /usr/lib32/%s/libGL.so.1.7.0" % arg, shell=True)
         except OSError as e:
-            print >>sys.stderr, "alternatives --set libGL-32bit /usr/lib32/%s/libGL.so.1.2.0 failed:" % arg, e
+            print >>sys.stderr, "alternatives --set libGL-32bit /usr/lib32/%s/libGL.so.1.7.0 failed:" % arg, e
 
     ########################################
     # Functions essential for grub parsing #
