diff -Nuar a/yali/storage/storageset.py b/yali/storage/storageset.py
--- a/yali/storage/storageset.py	2022-05-24 12:50:17.000000000 +0300
+++ b/yali/storage/storageset.py	2022-08-20 12:57:14.912919918 +0300
@@ -75,7 +75,7 @@
 
     def __init__(self, chroot="/"):
         self.devices = {}
-        self.path = os.path.join(chroot, "etc/blkid/blkid.tab")
+        self.path = os.path.join(chroot, "run/blkid/blkid.tab")
 
     def parse(self):
         with open(self.path) as blkidTab:
