diff -Nuar libreoffice-5.3.0.3/configure.ac libreoffice/configure.ac
--- libreoffice-5.3.0.3/configure.ac	2017-01-27 01:22:46.000000000 +0300
+++ libreoffice/configure.ac	2017-03-03 11:25:53.765271189 +0300
@@ -7875,9 +7875,6 @@
 
 libo_CHECK_SYSTEM_MODULE([libfreehand],[FREEHAND],[libfreehand-0.1])
 
-libo_CHECK_SYSTEM_MODULE([libebook],[EBOOK],[libe-book-0.1])
-libo_PKG_VERSION([EBOOK], [libe-book-0.1], [0.1.2])
-
 libo_CHECK_SYSTEM_MODULE([libabw],[ABW],[libabw-0.1])
 
 libo_CHECK_SYSTEM_MODULE([libpagemaker],[PAGEMAKER],[libpagemaker-0.0])
diff -Nuar libreoffice-5.3.0.3/external/Module_external.mk libreoffice/external/Module_external.mk
--- libreoffice-5.3.0.3/external/Module_external.mk	2017-01-27 01:22:46.000000000 +0300
+++ libreoffice/external/Module_external.mk	2017-03-03 11:26:43.746274226 +0300
@@ -36,7 +36,6 @@
 	$(call gb_Helper_optional,CPPUNIT,cppunit) \
 	$(call gb_Helper_optional,CT2N,ct2n) \
 	$(call gb_Helper_optional,CURL,curl) \
-	$(call gb_Helper_optional,EBOOK,libebook) \
 	$(call gb_Helper_optional,EPM,epm) \
 	$(call gb_Helper_optional,ETONYEK,libetonyek) \
 	$(call gb_Helper_optional,EXPAT,expat) \
diff -Nuar libreoffice-5.3.0.3/filter/Configuration_filter.mk libreoffice/filter/Configuration_filter.mk
--- libreoffice-5.3.0.3/filter/Configuration_filter.mk	2017-01-27 01:22:46.000000000 +0300
+++ libreoffice/filter/Configuration_filter.mk	2017-03-03 11:28:16.996279891 +0300
@@ -345,10 +345,6 @@
 	writer_OOXML \
 	writer_OOXML_Template \
 	writer_layout_dump_xml \
-	writer_BroadBand_eBook \
-	writer_FictionBook_2 \
-	writer_PalmDoc \
-	writer_Plucker_eBook \
 	writer_ApplePages \
 	MWAW_Text_Document \
 	Palm_Text_Document \
@@ -391,10 +387,6 @@
 	OOXML_Text \
 	OOXML_Text_Template \
 	writer_layout_dump \
-	BroadBand_eBook \
-	FictionBook_2 \
-	PalmDoc \
-	Plucker_eBook \
 	ApplePages \
 	MWAW_Text_Document \
 	Palm_Text_Document \
diff -Nuar libreoffice-5.3.0.3/writerperfect/Library_wpftwriter.mk libreoffice/writerperfect/Library_wpftwriter.mk
--- libreoffice-5.3.0.3/writerperfect/Library_wpftwriter.mk	2017-01-27 01:22:46.000000000 +0300
+++ libreoffice/writerperfect/Library_wpftwriter.mk	2017-03-03 11:30:32.317288112 +0300
@@ -49,7 +49,6 @@
 $(eval $(call gb_Library_use_externals,wpftwriter,\
 	abw \
 	boost_headers \
-	ebook \
 	etonyek \
 	icu_headers \
 	icui18n \
@@ -67,7 +66,6 @@
 
 $(eval $(call gb_Library_add_exception_objects,wpftwriter,\
 	writerperfect/source/writer/AbiWordImportFilter \
-	writerperfect/source/writer/EBookImportFilter \
 	writerperfect/source/writer/MSWorksImportFilter \
 	writerperfect/source/writer/MWAWImportFilter \
 	writerperfect/source/writer/PagesImportFilter \
diff -Nuar libreoffice-5.3.0.3/writerperfect/qa/unit/WpftImportTestBase.hxx libreoffice/writerperfect/qa/unit/WpftImportTestBase.hxx
--- libreoffice-5.3.0.3/writerperfect/qa/unit/WpftImportTestBase.hxx	2017-01-27 01:22:46.000000000 +0300
+++ libreoffice/writerperfect/qa/unit/WpftImportTestBase.hxx	2017-03-03 11:31:44.484292497 +0300
@@ -29,9 +29,6 @@
         ((minor) > (req_minor) \
          || ((minor) == (req_minor) && ((micro) >= (req_micro)))))
 
-#define REQUIRE_EBOOK_VERSION(major, minor, micro) \
-    REQUIRE_VERSION(EBOOK_VERSION_MAJOR, EBOOK_VERSION_MINOR, EBOOK_VERSION_MICRO, major, minor, micro)
-
 #define REQUIRE_ETONYEK_VERSION(major, minor, micro) \
     REQUIRE_VERSION(ETONYEK_VERSION_MAJOR, ETONYEK_VERSION_MINOR, ETONYEK_VERSION_MICRO, major, minor, micro)
 
diff -Nuar libreoffice-5.3.0.3/writerperfect/qa/unit/WpftWriterFilterTest.cxx libreoffice/writerperfect/qa/unit/WpftWriterFilterTest.cxx
--- libreoffice-5.3.0.3/writerperfect/qa/unit/WpftWriterFilterTest.cxx	2017-01-27 01:22:46.000000000 +0300
+++ libreoffice/writerperfect/qa/unit/WpftWriterFilterTest.cxx	2017-03-03 11:33:22.738298466 +0300
@@ -31,10 +31,6 @@
 
 void WpftWriterFilterTest::test()
 {
-    const writerperfect::test::WpftOptionalMap_t aEBookOptional
-    {
-        {"FictionBook2.fb2.zip", REQUIRE_EBOOK_VERSION(0, 1, 1)},
-    };
     const writerperfect::test::WpftOptionalMap_t aEtonyekOptional
     {
         {"Pages_4.pages", REQUIRE_ETONYEK_VERSION(0, 1, 2)},
@@ -58,7 +54,6 @@
     };
 
     doTest("com.sun.star.comp.Writer.AbiWordImportFilter", "/writerperfect/qa/unit/data/writer/libabw/");
-    doTest("org.libreoffice.comp.Writer.EBookImportFilter", "/writerperfect/qa/unit/data/writer/libe-book/", aEBookOptional);
     doTest("com.sun.star.comp.Writer.MSWorksImportFilter", "/writerperfect/qa/unit/data/writer/libwps/", aWpsOptional);
     doTest("com.sun.star.comp.Writer.MWAWImportFilter", "/writerperfect/qa/unit/data/writer/libmwaw/", aMWAWOptional);
     doTest("org.libreoffice.comp.Writer.PagesImportFilter", "/writerperfect/qa/unit/data/writer/libetonyek/", aEtonyekOptional);
diff -Nuar libreoffice-5.3.0.3/writerperfect/source/writer/wpftwriter.component libreoffice/writerperfect/source/writer/wpftwriter.component
--- libreoffice-5.3.0.3/writerperfect/source/writer/wpftwriter.component	2017-01-27 01:22:46.000000000 +0300
+++ libreoffice/writerperfect/source/writer/wpftwriter.component	2017-03-03 11:34:21.210302018 +0300
@@ -38,11 +38,6 @@
     <service name="com.sun.star.document.ExtendedTypeDetection"/>
     <service name="com.sun.star.document.ImportFilter"/>
   </implementation>
-  <implementation name="org.libreoffice.comp.Writer.EBookImportFilter"
-      constructor="org_libreoffice_comp_Writer_EBookImportFilter_get_implementation">
-    <service name="com.sun.star.document.ExtendedTypeDetection"/>
-    <service name="com.sun.star.document.ImportFilter"/>
-  </implementation>
   <implementation name="org.libreoffice.comp.Writer.PagesImportFilter"
       constructor="org_libreoffice_comp_Writer_PagesImportFilter_get_implementation">
     <service name="com.sun.star.document.ExtendedTypeDetection"/>
