diff -rupN Pillow-8.3.1/docs/conf.py Pillow-8.3.1-new/docs/conf.py
--- Pillow-8.3.1/docs/conf.py	2021-07-07 10:45:16.614845303 +0200
+++ Pillow-8.3.1-new/docs/conf.py	2021-07-07 10:45:02.533856396 +0200
@@ -30,12 +30,10 @@ needs_sphinx = "2.4"
 # ones.
 extensions = [
     "sphinx_copybutton",
-    "sphinx_issues",
     "sphinx_removed_in",
     "sphinx.ext.autodoc",
     "sphinx.ext.intersphinx",
-    "sphinx.ext.viewcode",
-    "sphinxext.opengraph",
+    "sphinx.ext.viewcode"
 ]
 
 intersphinx_mapping = {"python": ("https://docs.python.org/3", None)}
