diff -rupN --no-dereference Pillow-8.3.1/docs/Makefile Pillow-8.3.1-new/docs/Makefile
--- Pillow-8.3.1/docs/Makefile	2021-07-06 16:05:28.000000000 +0200
+++ Pillow-8.3.1-new/docs/Makefile	2021-07-07 10:44:44.471870820 +0200
@@ -42,7 +42,7 @@ clean:
 	-rm -rf $(BUILDDIR)/*
 
 html:
-	$(SPHINXBUILD) -b html -W --keep-going $(ALLSPHINXOPTS) $(BUILDDIR)/html
+	$(SPHINXBUILD) -b html --keep-going $(ALLSPHINXOPTS) $(BUILDDIR)/html
 	@echo
 	@echo "Build finished. The HTML pages are in $(BUILDDIR)/html."
 
