--- a/cinnamon.session.in
+++ b/cinnamon.session.in
@@ -1,5 +1,5 @@
 [Cinnamon Session]
 Name=Cinnamon
-RequiredComponents=cinnamon;org.cinnamon.ScreenSaver;nemo-autostart;@REQUIRED@cinnamon-killer-daemon;
+RequiredComponents=cinnamon;org.cinnamon.ScreenSaver;polkit-cinnamon-authentication-agent-1;nemo-autostart;@REQUIRED@cinnamon-killer-daemon;
 DesktopName=X-Cinnamon
 

--- a/cinnamon2d.session.in
+++ b/cinnamon2d.session.in
@@ -1,5 +1,5 @@
 [Cinnamon Session]
 Name=Cinnamon (Software Rendering)
-RequiredComponents=cinnamon2d;org.cinnamon.ScreenSaver;nemo-autostart;@REQUIRED@cinnamon-killer-daemon;
+RequiredComponents=cinnamon2d;org.cinnamon.ScreenSaver;polkit-cinnamon-authentication-agent-1;nemo-autostart;@REQUIRED@cinnamon-killer-daemon;
 DesktopName=X-Cinnamon
 

