diff -Nuar a/files/usr/share/cinnamon/cinnamon-settings/modules/cs_notifications.py b/files/usr/share/cinnamon/cinnamon-settings/modules/cs_notifications.py
--- a/files/usr/share/cinnamon/cinnamon-settings/modules/cs_notifications.py	2022-06-21 18:01:39.000000000 +0300
+++ b/files/usr/share/cinnamon/cinnamon-settings/modules/cs_notifications.py	2022-07-17 17:59:13.743936377 +0300
@@ -1,7 +1,10 @@
 #!/usr/bin/python3
 
 import gi
-gi.require_version('Notify', '0.7')
+try:
+    gi.require_version("Notify", "0.7")
+except ValueError:
+    gi.require_version("Notify", "0.8")
 from gi.repository import Notify
 
 from SettingsWidgets import SidePage
