From 6b0eeb2c15aa278ffa577bcefebcff8f2c98aca9 Mon Sep 17 00:00:00 2001
From: Ilya Fedin <fedin-ilja2010@ya.ru>
Date: Thu, 1 Jun 2023 23:18:26 +0400
Subject: [PATCH] Ensure cppgir is not getting installed

---
 external/glib/CMakeLists.txt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/cmake/external/glib/CMakeLists.txt b/cmake/external/glib/CMakeLists.txt
index 66385e3b..20b62a70 100644
--- a/cmake/external/glib/CMakeLists.txt
+++ b/cmake/external/glib/CMakeLists.txt
@@ -11,7 +11,7 @@ function(add_cppgir) # isolate scope
     set(BUILD_TESTING OFF)
     set(BUILD_DOC OFF)
     set(CMAKE_DISABLE_FIND_PACKAGE_Qt5Core ON)
-    add_subdirectory(cppgir)
+    add_subdirectory(cppgir EXCLUDE_FROM_ALL)
 endfunction()
 add_cppgir()

