diff -Nuar a/Telegram/lib_lottie/lottie/lottie_emoji.h b/Telegram/lib_lottie/lottie/lottie_emoji.h
--- a/Telegram/lib_lottie/lottie/lottie_emoji.h	2022-07-25 12:46:55.000000000 +0300
+++ b/Telegram/lib_lottie/lottie/lottie_emoji.h	2022-08-06 15:19:52.852451704 +0300
@@ -9,6 +9,7 @@
 #include "ui/effects/frame_generator.h"

 #include <QtGui/QImage>
+#include <memory>

 namespace rlottie {
 class Animation;
diff -Nuar a/Telegram/SourceFiles/ffmpeg/ffmpeg_emoji.h b/Telegram/SourceFiles/ffmpeg/ffmpeg_emoji.h
--- a/Telegram/SourceFiles/ffmpeg/ffmpeg_emoji.h	2022-08-05 10:12:48.000000000 +0300
+++ b/Telegram/SourceFiles/ffmpeg/ffmpeg_emoji.h	2022-08-06 15:19:14.286454487 +0300
@@ -10,6 +10,7 @@
 #include "ui/effects/frame_generator.h"

 #include <QtGui/QImage>
+#include <memory>

 namespace FFmpeg {

