--- a/themes/Adwaita/gtk-2.0/gtkrc	2020-05-26 10:13:13.557213234 +0300
+++ b/themes/Adwaita/gtk-2.0/gtkrc	2020-05-26 12:31:51.312349746 +0300
@@ -34,6 +34,31 @@
 gtk-auto-mnemonics              = 1
 gtk-primary-button-warps-slider = 1
 
+# https://www.opendesktop.org/u/thefrankenstein
+style "fix_ugly_text"
+{
+  engine "murrine"
+    {
+      textstyle   = 0
+      text_shade  = 1.0
+    }
+}
+
+class "GtkLabel"                                           style "fix_ugly_text"
+class "GtkImage"                                           style "fix_ugly_text"
+class "GtkComboBox"                                        style "fix_ugly_text"
+class "GtkCellView"                                        style "fix_ugly_text"
+class "GtkMenuItem"                                        style "fix_ugly_text"
+widget_class "*.<GtkButton>.<GtkLabel>"                    style "fix_ugly_text"
+widget_class "*.<GtkComboBox>.<GtkCellView>"               style "fix_ugly_text"
+widget_class "*<GtkImage>*"                                style "fix_ugly_text"
+widget_class "*.<GtkLabel>"                                style:highest "fix_ugly_text"
+widget_class "*.<GtkMenuItem>.*"                           style:highest "fix_ugly_text"
+widget_class "*<GtkMenuBar>*<GtkLabel>*"                   style:highest "fix_ugly_text"
+widget_class "*<GtkToolbar>*<GtkButton>*<GtkLabel>*"       style:highest "fix_ugly_text"
+widget_class "*<GtkToolbar>*<GtkLabel>*"                   style:highest "fix_ugly_text"
+widget_class "*<GtkButton>*<GtkLabel>*"                    style:highest "fix_ugly_text"
+
 # And hand over the control to the theme files
 include "main.rc"
 include "apps.rc"

--- a/themes/Adwaita-dark/gtk-2.0/gtkrc	2020-05-26 10:20:15.758226186 +0300
+++ b/themes/Adwaita-dark/gtk-2.0/gtkrc	2020-05-26 12:32:06.927679341 +0300
@@ -34,6 +34,31 @@
 gtk-auto-mnemonics              = 1
 gtk-primary-button-warps-slider = 1
 
+# https://www.opendesktop.org/u/thefrankenstein
+style "fix_ugly_text"
+{
+  engine "murrine"
+    {
+      textstyle   = 0
+      text_shade  = 1.0
+    }
+}
+
+class "GtkLabel"                                           style "fix_ugly_text"
+class "GtkImage"                                           style "fix_ugly_text"
+class "GtkComboBox"                                        style "fix_ugly_text"
+class "GtkCellView"                                        style "fix_ugly_text"
+class "GtkMenuItem"                                        style "fix_ugly_text"
+widget_class "*.<GtkButton>.<GtkLabel>"                    style "fix_ugly_text"
+widget_class "*.<GtkComboBox>.<GtkCellView>"               style "fix_ugly_text"
+widget_class "*<GtkImage>*"                                style "fix_ugly_text"
+widget_class "*.<GtkLabel>"                                style:highest "fix_ugly_text"
+widget_class "*.<GtkMenuItem>.*"                           style:highest "fix_ugly_text"
+widget_class "*<GtkMenuBar>*<GtkLabel>*"                   style:highest "fix_ugly_text"
+widget_class "*<GtkToolbar>*<GtkButton>*<GtkLabel>*"       style:highest "fix_ugly_text"
+widget_class "*<GtkToolbar>*<GtkLabel>*"                   style:highest "fix_ugly_text"
+widget_class "*<GtkButton>*<GtkLabel>*"                    style:highest "fix_ugly_text"
+
 # And hand over the control to the theme files
 include "main.rc"
 include "apps.rc"
