diff -up qtspeech-opensource-src-5.9.2/config.tests/speechd/speechd.pro.includes qtspeech-opensource-src-5.9.2/config.tests/speechd/speechd.pro
--- qtspeech-opensource-src-5.9.2/config.tests/speechd/speechd.pro.includes	2017-09-08 08:11:16.000000000 -0500
+++ qtspeech-opensource-src-5.9.2/config.tests/speechd/speechd.pro	2018-01-25 11:04:10.109329474 -0600
@@ -1,3 +1,4 @@
 SOURCES = speechd.cpp
 
+INCLUDEPATH += /usr/include/speech-dispatcher
 LIBS += -lspeechd
diff -up qtspeech-opensource-src-5.9.2/src/plugins/tts/speechdispatcher/speechdispatcher.pro.includes qtspeech-opensource-src-5.9.2/src/plugins/tts/speechdispatcher/speechdispatcher.pro
--- qtspeech-opensource-src-5.9.2/src/plugins/tts/speechdispatcher/speechdispatcher.pro.includes	2017-09-08 08:11:16.000000000 -0500
+++ qtspeech-opensource-src-5.9.2/src/plugins/tts/speechdispatcher/speechdispatcher.pro	2018-01-25 11:06:07.772179149 -0600
@@ -9,6 +9,7 @@ QT += core texttospeech
 CONFIG += link_pkgconfig
 packagesExist(speech-dispatcher): PKGCONFIG = speech-dispatcher
 config_speechd: LIBS += -lspeechd
+config_speechd: INCLUDEPATH += /usr/include/speech-dispatcher
 
 HEADERS += \
     qtexttospeech_speechd.h \
