diff --git a/cmake/modules/DispatchCompilerWarnings.cmake b/cmake/modules/DispatchCompilerWarnings.cmake
index b106444..bdb3f28 100644
--- a/cmake/modules/DispatchCompilerWarnings.cmake
+++ b/cmake/modules/DispatchCompilerWarnings.cmake
@@ -6,7 +6,6 @@ elseif(WIN32)
   # so that we can use __popcnt64
   add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:-fms-extensions>)
 else()
-  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:-Werror>)
   add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:-Wall>)
   add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:-Wextra>)
