--- speex-1.2.0/configure.ac.orig	2016-12-08 00:20:14.000000000 +0100
+++ speex-1.2.0/configure.ac	2016-12-18 18:08:22.364596212 +0100
@@ -167,7 +167,7 @@
 fi])
 
 AC_ARG_ENABLE(binaries, [  --disable-binaries   Do not build the encoder and decoder programs, only the library])
-if test "$enableval" != no; then
+if test "$enable_binaries" != no; then
   PKG_CHECK_MODULES([OGG], [ogg],
     AM_CONDITIONAL([BUILD_BINARIES], true),
AM_CONDITIONAL([BUILD_BINARIES], false))