--- a/SConstruct	2020-08-04 18:40:06.000000000 +0000
+++ b/SConstruct	2020-12-12 18:29:38.378386573 +0000
@@ -1460,8 +1460,8 @@
             # follow FHS, put in /usr/local/libXX, not /usr/libXX
             # may be lib, lib32 or lib64
             python_libdir = polystr(python_libdir)
-            python_libdir = python_libdir.replace("/usr/lib",
-                                                  "/usr/local/lib")
+#            python_libdir = python_libdir.replace("/usr/lib",
+#                                                  "/usr/local/lib")
         python_module_dir = str(python_libdir) + os.sep
         # Many systems can have a problem with the Python path
         announce("Ensure your PYTHONPATH includes %s" % python_module_dir)
