--- a/SConscript	2021-09-22 00:53:44.000000000 +0300
+++ b/SConscript	2021-09-25 06:56:40.449906438 +0300
@@ -1492,8 +1492,8 @@
             # follow FHS, put in /usr/local/libXX, not /usr/libXX
             # may be lib, lib32 or lib64
             python_libdir = polystr(python_libdir)
-            python_libdir = python_libdir.replace("/usr/lib",
-                                                  "/usr/local/lib")
+            # python_libdir = python_libdir.replace("/usr/lib",
+            #                                       "/usr/local/lib")
         python_module_dir = str(python_libdir) + os.sep
         # Many systems can have a problem with the Python path
         if 'PYTHONPATH' in os.environ:
