diff -up qtwebkit-opensource-src-5.2.1/Tools/qmake/mkspecs/features/rpath.prf.no_rpath qtwebkit-opensource-src-5.2.1/Tools/qmake/mkspecs/features/rpath.prf
--- qtwebkit-opensource-src-5.2.1/Tools/qmake/mkspecs/features/rpath.prf.no_rpath	2014-02-01 14:37:41.000000000 -0600
+++ qtwebkit-opensource-src-5.2.1/Tools/qmake/mkspecs/features/rpath.prf	2014-05-02 18:47:26.789051711 -0500
@@ -6,7 +6,7 @@
 
 equals(ROOT_BUILD_DIR, $$dirname(DESTDIR)): RPATHDIR_RELATIVE_TO_DESTDIR = ../lib
 
-linux-*:!isEmpty(RPATHDIR_RELATIVE_TO_DESTDIR) {
+linux-rpath_me_harder_no_no_no:!isEmpty(RPATHDIR_RELATIVE_TO_DESTDIR) {
     # Do the rpath by hand since it's not possible to use ORIGIN in QMAKE_RPATHDIR
     # this expands to $ORIGIN (after qmake and make), it does NOT read a qmake var.
     QMAKE_RPATHDIR = \$\$ORIGIN/$${RPATHDIR_RELATIVE_TO_DESTDIR} $${QMAKE_RPATHDIR} 
