diff -up qtwebkit-opensource-src-5.2.0-rc1/Source/WebCore/plugins/PluginDatabase.cpp.pluginpath qtwebkit-opensource-src-5.2.0-rc1/Source/WebCore/plugins/PluginDatabase.cpp
--- qtwebkit-opensource-src-5.2.0-rc1/Source/WebCore/plugins/PluginDatabase.cpp.pluginpath	2013-11-26 19:01:47.000000000 -0600
+++ qtwebkit-opensource-src-5.2.0-rc1/Source/WebCore/plugins/PluginDatabase.cpp	2013-12-02 11:14:43.306114402 -0600
@@ -407,6 +407,7 @@ Vector<String> PluginDatabase::defaultPl
     paths.append("/usr/lib/firefox/plugins");
     paths.append("/usr/lib64/browser-plugins");
     paths.append("/usr/lib/browser-plugins");
+    paths.append("/usr/lib/mozilla/plugins-wrapped");
     paths.append("/usr/lib/mozilla/plugins");
     paths.append("/usr/local/netscape/plugins");
     paths.append("/opt/mozilla/plugins");
@@ -416,6 +417,7 @@ Vector<String> PluginDatabase::defaultPl
     paths.append("/usr/lib/netscape/plugins");
     paths.append("/usr/lib/netscape/plugins-libc5");
     paths.append("/usr/lib/netscape/plugins-libc6");
+    paths.append("/usr/lib64/mozilla/plugins-wrapped");
     paths.append("/usr/lib64/netscape/plugins");
     paths.append("/usr/lib64/mozilla/plugins");
     paths.append("/usr/lib/nsbrowser/plugins"); 
