diff -up chromium-46.0.2490.80/chrome/test/data/webui_test_resources.grd.542819fix chromium-46.0.2490.80/chrome/test/data/webui_test_resources.grd
--- chromium-46.0.2490.80/chrome/test/data/webui_test_resources.grd.542819fix   2015-10-14 10:47:17.568757249 -0400
+++ chromium-46.0.2490.80/chrome/test/data/webui_test_resources.grd     2015-10-14 10:48:01.196466805 -0400
@@ -8,7 +8,6 @@
   </outputs>
   <release seq="1">
     <includes>
-      <include name="IDR_WEBUI_TEST_I18N_PROCESS_CSS_TEST" file="webui/i18n_process_css_test.html" flattenhtml="true" allowexternalscript="true" type="BINDATA" />
     </includes>
   </release>
 </grit> 
