diff -Nuar a/kaptan5/kaptan.py b/kaptan5/kaptan.py
--- a/kaptan5/kaptan.py	2017-04-05 12:46:42.000000000 +0300
+++ b/kaptan5/kaptan.py	2018-02-06 01:23:49.000000000 +0300
@@ -36,8 +36,8 @@
         self.setWindowIcon(QIcon.fromTheme("kaptan-icon"))
         self.setMinimumSize(850, 600)
         self.setMaximumSize(950, 620)
-        x = (QDesktopWidget().screen().width() - self.width())/2
-        y = (QDesktopWidget().screen().height() - self.height())/2
+        x = (QDesktopWidget().screenGeometry().width() - self.width())/2
+        y = (QDesktopWidget().screenGeometry().height() - self.height())/2
         self.move(x, y)
         self.setPixmap(QWizard.LogoPixmap, QPixmap(":/data/images/kaptan.png"))
 
