diff -Nru violetland-0.5.orig/CMakeLists.txt violetland-0.5/CMakeLists.txt
--- violetland-0.5.orig/CMakeLists.txt	2016-01-04 22:15:21.000000000 +0200
+++ violetland-0.5/CMakeLists.txt	2019-07-03 17:58:35.395818499 +0300
@@ -13,7 +13,7 @@
 
 # Searching for libintl using bundled CMake script
 set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/lib/fcitx/cmake")
-include(FindLibintl)
+include(FindIntl)
 
 if(${MINGW})
 	set(LIBINTL_LIBRARIES intl)
diff -Nru violetland-0.5.orig/po/CMakeLists.txt violetland-0.5/po/CMakeLists.txt
--- violetland-0.5.orig/po/CMakeLists.txt	2016-01-04 22:15:21.000000000 +0200
+++ violetland-0.5/po/CMakeLists.txt	2019-07-03 18:00:14.454813019 +0300
@@ -2,9 +2,9 @@
 CMAKE_MINIMUM_REQUIRED(VERSION 2.6.0 FATAL_ERROR) 
 
 set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/../lib/CMake-Gettext")
-set(GettextTranslate_ALL true)
-set(GettextTranslate_GMO_BINARY true)
+# set(GettextTranslate_ALL true)
+# set(GettextTranslate_GMO_BINARY true)
 
-include(GettextTranslate)
-GettextTranslate()
+#include(GettextTranslate)
+#GettextTranslate()
 
