diff -Nuar a/src/mbox_p.cpp b/src/mbox_p.cpp
--- a/src/mbox_p.cpp	2017-12-04 13:20:23.000000000 +0300
+++ b/src/mbox_p.cpp	2017-12-16 16:10:23.271073753 +0300
@@ -157,7 +157,9 @@
                 *d++ = '>';
             }
             // fall through
+#if QT_VERSION >= QT_VERSION_CHECK(5,8,0)
             Q_FALLTHROUGH();
+#endif
         default:
             onlyAnglesAfterLF = false;
             break;
