diff -rupN --no-dereference podofo-0.9.8/CMakeLists.txt podofo-0.9.8-new/CMakeLists.txt
--- podofo-0.9.8/CMakeLists.txt	2022-05-03 14:18:23.000000000 +0200
+++ podofo-0.9.8-new/CMakeLists.txt	2022-05-03 15:21:08.930832253 +0200
@@ -407,8 +407,8 @@ ENDIF(NOT PODOFO_BUILD_LIB_ONLY)
 
 FIND_PACKAGE(OpenSSL)
 
-FIND_PACKAGE(FREETYPE REQUIRED)
-MESSAGE("Found freetype library at ${FREETYPE_LIBRARIES}, headers ${FREETYPE_INCLUDE_DIR}")
+FIND_PACKAGE(Freetype REQUIRED)
+MESSAGE("Found freetype library at ${FREETYPE_LIBRARIES}, headers ${FREETYPE_INCLUDE_DIRS}")
 
 FIND_PACKAGE(LIBSTLPORT)
 SET(stlport_libraries_if_use_stlport)
@@ -487,7 +487,7 @@ INCLUDE_DIRECTORIES(BEFORE # before tool
 
 INCLUDE_DIRECTORIES(
     ${PoDoFo_SOURCE_DIR}/src/podofo
-    ${FREETYPE_INCLUDE_DIR}
+    ${FREETYPE_INCLUDE_DIRS}
     ${ZLIB_INCLUDE_DIR}
     ${EXTRA_INCLUDES}
      )
