diff -rupN podofo-0.9.5/CMakeLists.txt podofo-0.9.5-new/CMakeLists.txt
--- podofo-0.9.5/CMakeLists.txt	2017-02-01 22:23:03.000000000 +0100
+++ podofo-0.9.5-new/CMakeLists.txt	2017-02-10 14:40:25.472066545 +0100
@@ -296,8 +296,6 @@ IF(CMAKE_COMPILER_IS_GNUCXX)
         SET(PODOFO_USE_VISIBILITY ${PODOFO_HAVE_GCC4})
     ENDIF(NOT DEFINED PODOFO_USE_VISIBILITY)
 
-    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++98")
-
     SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Woverloaded-virtual -Wswitch-enum -Wcast-qual -Wwrite-strings -Wredundant-decls -Wreorder")
 
     #
