diff -rupN podofo-0.9.4/CMakeLists.txt podofo-0.9.4-new/CMakeLists.txt
--- podofo-0.9.4/CMakeLists.txt	2016-06-08 22:55:08.000000000 +0200
+++ podofo-0.9.4-new/CMakeLists.txt	2016-09-17 22:27:38.342555473 +0200
@@ -391,8 +391,8 @@ ENDIF(NOT PODOFO_BUILD_LIB_ONLY)
 
 FIND_PACKAGE(OpenSSL)
 
-FIND_PACKAGE(FREETYPE REQUIRED)
-MESSAGE("Found freetype library at ${FREETYPE_LIBRARIES}, headers ${FREETYPE_INCLUDE_DIR}")
+FIND_PACKAGE(Freetype REQUIRED)
+MESSAGE("Found freetype library at ${FREETYPE_LIBRARIES}, headers ${FREETYPE_INCLUDE_DIRS}")
 
 FIND_PACKAGE(LIBSTLPORT)
 SET(stlport_libraries_if_use_stlport)
@@ -471,7 +471,7 @@ INCLUDE_DIRECTORIES(BEFORE # before tool
 
 INCLUDE_DIRECTORIES(
     ${PoDoFo_SOURCE_DIR}/src
-    ${FREETYPE_INCLUDE_DIR}
+    ${FREETYPE_INCLUDE_DIRS}
     ${ZLIB_INCLUDE_DIR}
     ${EXTRA_INCLUDES}
      )
