diff -Nuar a/src/konqueror.kcfg b/src/konqueror.kcfg
--- a/src/konqueror.kcfg	2017-10-09 07:28:51.000000000 +0300
+++ b/src/konqueror.kcfg	2017-11-08 11:54:55.848309054 +0300
@@ -463,12 +463,12 @@
 
   <group name="Reusing" >
     <entry key="AlwaysHavePreloaded" type="Bool">
-      <default>true</default>
+      <default>false</default>
       <label></label>
       <whatsthis></whatsthis>
     </entry>
     <entry key="PreloadOnStartup" type="Bool">
-      <default>false</default>
+      <default>true</default>
       <label></label>
       <whatsthis></whatsthis>
     </entry>
