diff --git a/daemon/BackendConfig.cmake b/daemon/BackendConfig.cmake
index 54094e0..c9e4dc5 100644
--- a/daemon/BackendConfig.cmake
+++ b/daemon/BackendConfig.cmake
@@ -18,6 +18,8 @@ if (UDEV_FOUND)
         backends/upower/udevqtdevice.cpp
     )
 
+    add_definitions(-DHAVE_UDEV)
+
     set_source_files_properties(
     ${CMAKE_CURRENT_SOURCE_DIR}/backends/upower/dbus/org.freedesktop.UPower.xml
     ${CMAKE_CURRENT_SOURCE_DIR}/backends/upower/dbus/org.freedesktop.UPower.Device.xml