From 69a8d66c4682410537ce8d1f2b7012fb49773005 Mon Sep 17 00:00:00 2001
From: Hrvoje Senjan <hrvoje.senjan@gmail.com>
Date: Wed, 3 Sep 2014 23:13:35 +0200
Subject: [PATCH 1/1] Use /run for ksysguardd's pid file

---
 ksysguardd/ksysguardd.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ksysguardd/ksysguardd.c b/ksysguardd/ksysguardd.c
index a5ecca1..883f8a1 100644
--- a/ksysguardd/ksysguardd.c
+++ b/ksysguardd/ksysguardd.c
@@ -62,7 +62,7 @@ static ClientInfo ClientList[ MAX_CLIENTS ];
 static int SocketPort = -1;
 static unsigned char BindToAllInterfaces = 0;
 static int CurrentSocket;
-static const char LockFile[] = "/var/run/ksysguardd.pid";
+static const char LockFile[] = "/run/ksysguardd.pid";
 static const char *ConfigFile = KSYSGUARDDRCFILE;
 
 void signalHandler( int sig );
-- 
2.1.0

