From 22bbf8efc0deeb62cd706f1c2fee97201eb7427a Mon Sep 17 00:00:00 2001
From: Tom Stellard <tstellar@redhat.com>
Date: Fri, 13 Dec 2019 00:39:17 +0000
Subject: [PATCH] Link against libclang-cpp.so


--- ispc-1.16.0/CMakeLists.txt.orig	2021-06-11 17:41:24.000000000 -0700
+++ ispc-1.16.0/CMakeLists.txt	2021-06-13 11:23:05.621934106 -0700
@@ -281,7 +281,7 @@
     list(APPEND ISPC_TARGETS wasm-i32x4)
 endif()
 
-set(CLANG_LIBRARY_LIST clangFrontend clangDriver clangSerialization clangParse clangSema clangAnalysis clangAST clangBasic clangEdit clangLex)
+set(CLANG_LIBRARY_LIST clang-cpp)
 set(LLVM_COMPONENTS engine ipo bitreader bitwriter instrumentation linker option frontendopenmp)
 
 if (X86_ENABLED)
