diff --git a/src/core/utils.cpp b/src/core/utils.cpp
index 5346bd4..3ba7122 100644
--- a/src/core/utils.cpp
+++ b/src/core/utils.cpp
@@ -44,8 +44,5 @@ QString Utils::bool2Str(bool val)
 
 QString Utils::ts2Str(quint64 ts)
 {
-    QDateTime dt;
-    dt.setTime_t(ts);
-    dt.setTimeSpec(Qt::UTC);
-    return dt.toString(Qt::ISODate);
+    return QDateTime::fromTime_t(ts, Qt::UTC, 0).toString(Qt::ISODate);
 }
--  
