diff -Nuar a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2017-12-08 04:03:49.000000000 +0300
+++ b/CMakeLists.txt	2017-12-17 00:25:23.527037197 +0300
@@ -28,7 +28,7 @@
 set(calendarsupport_LIB_VERSION ${PIM_VERSION})
 set(AKONADI_MIMELIB_VERSION "5.7.0")
 set(KDEPIM_LIB_VERSION "5.7.0")
-set(QT_REQUIRED_VERSION "6.9.0")
+set(QT_REQUIRED_VERSION "6.9.0")
 set(kmime_LIB_VERSION "5.7.0")
 set(CALENDARUTILS_LIB_VERSION "5.7.0")
 set(kcalendarcore_LIB_VERSION "5.7.0")
