From c47d47ab7f66a1973b1314ef531de2fb444c2fe6 Mon Sep 17 00:00:00 2001
From: Ilias Stamatis <ilstam@mailbox.org>
Date: Wed, 22 Dec 2021 14:43:22 +0000
Subject: [PATCH] QProgressBar.setValue() expects an int, not a float

---
 ffmulticonverter/progress.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ffmulticonverter/progress.py b/ffmulticonverter/progress.py
index 5d14efc..e196ef4 100755
--- a/ffmulticonverter/progress.py
+++ b/ffmulticonverter/progress.py
@@ -304,7 +304,7 @@ def convert_video(self, from_file, to_file, command):
                         time = utils.duration_in_seconds(time)
                     now_sec = int(float(time))
                     try:
-                        self.nowQPBar.setValue(100 * now_sec / total)
+                        self.nowQPBar.setValue(int(100 * now_sec / total))
                     except (UnboundLocalError, ZeroDivisionError):
                         pass
                 self.update_text_edit_signal.emit(myline)
