--- setup_old.py	2016-08-13 23:13:53.000000000 +0300
+++ setup.py	2017-03-15 20:53:58.626914961 +0300
@@ -19,7 +19,7 @@
         build.run(self)
 
         self.mkpath(self.build_base)
-        self.spawn(["rcc", "-binary", "data.qrc", "-o", "%s/data.rcc" % self.build_base])
+        self.spawn(["rcc-qt5", "-binary", "data.qrc", "-o", "%s/data.rcc" % self.build_base])
 
 
 class Install(install):
