--- a/src/dlg_ui_ex.cpp	2019-07-16 16:54:55.000000000 +0000
+++ b/src/dlg_ui_ex.cpp	2020-12-19 18:28:04.920908321 +0000
@@ -13,16 +13,16 @@
 #include "fluid/dlg_ui.hpp"
 #include "version.hpp"
 
-extern "C" 
-{
-    extern char _binary_LICENSE_res_start;
-    extern char _binary_LICENSE_res_end;
-    extern char _binary_LICENSE_res_size;
-
-    extern char _binary_KEYS_res_start;
-    extern char _binary_KEYS_res_end;
-    extern char _binary_KEYS_res_size;
-}
+//extern "C" 
+//{
+//    extern char _binary_LICENSE_res_start;
+//    extern char _binary_LICENSE_res_end;
+//    extern char _binary_LICENSE_res_size;
+
+//    extern char _binary_KEYS_res_start;
+//    extern char _binary_KEYS_res_end;
+//    extern char _binary_KEYS_res_size;
+//}
 
 static dlg_ui *ui;
 
@@ -188,8 +188,8 @@
     m_menuitem_gpsd_lockcursor->label(_("Lock to cursor"));
     // Help
     m_menuitem_help->label(_("Help"));
-    m_menuitem_help_about->label(_("About"));
-    m_menuitem_help_usage->label(_("Usage"));
+//    m_menuitem_help_about->label(_("About"));
+//    m_menuitem_help_usage->label(_("Usage"));
 
     // Populate the Basemap selector
     update_choice_map_ex();
@@ -492,38 +492,38 @@
     m_dlg_search->show();
 }
 
-void dlg_ui::about_ex()
-{
+//void dlg_ui::about_ex()
+//{
     // Show about dialog
-    if (!m_dlg_txtdisp)
-        m_dlg_txtdisp = new dlg_txtdisp;
-
-    std::string v(std::string(_("Version: ")) + std::string(FLORB_PROGSTR));
-    std::string l(&_binary_LICENSE_res_start, (size_t)&_binary_LICENSE_res_size);
-
-    m_dlg_txtdisp->title(_("About / License"));
-    m_dlg_txtdisp->clear();
-    m_dlg_txtdisp->append(v);
-    m_dlg_txtdisp->append("\n\n");
-    m_dlg_txtdisp->append(florb::utils::str_split(l, "\n\n"), "\n\n", true); 
-
-    m_dlg_txtdisp->show();
-}
-void dlg_ui::usage_ex()
-{
+//    if (!m_dlg_txtdisp)
+//        m_dlg_txtdisp = new dlg_txtdisp;
+//
+//    std::string v(std::string(_("Version: ")) + std::string(FLORB_PROGSTR));
+//    std::string l(&_binary_LICENSE_res_start, (size_t)&_binary_LICENSE_res_size);
+//
+//    m_dlg_txtdisp->title(_("About / License"));
+//    m_dlg_txtdisp->clear();
+//    m_dlg_txtdisp->append(v);
+//    m_dlg_txtdisp->append("\n\n");
+//    m_dlg_txtdisp->append(florb::utils::str_split(l, "\n\n"), "\n\n", true); 
+//
+//    m_dlg_txtdisp->show();
+//}
+//void dlg_ui::usage_ex()
+//{
     // Show about dialog
-    if (!m_dlg_txtdisp)
-        m_dlg_txtdisp = new dlg_txtdisp;
-
-    std::string u(&_binary_KEYS_res_start, (size_t)&_binary_KEYS_res_size);
-
-    m_dlg_txtdisp->title(_("Usage"));
-    m_dlg_txtdisp->clear();
-    m_dlg_txtdisp->append(florb::utils::str_split(u, "\n\n"), "\n\n", true);
-
-    m_dlg_txtdisp->show();
-}
-
+//    if (!m_dlg_txtdisp)
+//        m_dlg_txtdisp = new dlg_txtdisp;
+//
+//    std::string u(&_binary_KEYS_res_start, (size_t)&_binary_KEYS_res_size);
+//
+//    m_dlg_txtdisp->title(_("Usage"));
+//    m_dlg_txtdisp->clear();
+//    m_dlg_txtdisp->append(florb::utils::str_split(u, "\n\n"), "\n\n", true);
+//
+//    m_dlg_txtdisp->show();
+//}
+//
 
 void dlg_ui::cb_btn_loadtrack_ex(Fl_Widget *widget)
 {
@@ -709,13 +709,13 @@
     }
 
     // Help submenu
-    else if (mit == m_menuitem_help_about) { 
-        about_ex();
-    }
-    else if (mit == m_menuitem_help_usage) { 
-        usage_ex();
-    }
-
+//    else if (mit == m_menuitem_help_about) { 
+//        about_ex();
+//    }
+//    else if (mit == m_menuitem_help_usage) { 
+//        usage_ex();
+//    }
+//
     m_wgtmap->take_focus();
 }
 
