diff --git a/src/dlg_search_ex.cpp b/src/dlg_search_ex.cpp
index 6e00773..da5a692 100644
--- a/src/dlg_search_ex.cpp
+++ b/src/dlg_search_ex.cpp
@@ -66,9 +66,9 @@ void dlg_search::cb_btn_search_ex(Fl_Widget *widget)
 
     // Construct query string
     std::ostringstream oss;
-    oss << "http://nominatim.openstreetmap.org/search/"; 
+    oss << "https://nominatim.openstreetmap.org/search?q="; 
     oss << query;
-    oss << "?format=xml";
+    oss << "&format=xml";
 
     // Download the search results
     bool ret = m_downloader->queue(oss.str(), NULL);
