From 025f791b57c988d33249c5c33250229fa0e7e8f1 Mon Sep 17 00:00:00 2001
From: Peter Jones <pjones@redhat.com>
Date: Thu, 14 Jun 2018 12:15:03 -0400
Subject: [PATCH 08/34] Make path-helpers.c also import fix_coverity.h

Signed-off-by: Peter Jones <pjones@redhat.com>
---
 src/path-helpers.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/path-helpers.c b/src/path-helpers.c
index 1de00860d92..1b328071587 100644
--- a/src/path-helpers.c
+++ b/src/path-helpers.c
@@ -16,6 +16,7 @@
  * License along with this library; if not, see
  * <http://www.gnu.org/licenses/>.
  */
+#include "fix_coverity.h"
 
 #include "efivar.h"
 
-- 
2.17.1

