From 8f9c1406a2a50d0c67b1380ad6fddc2c266d39f6 Mon Sep 17 00:00:00 2001
From: Peter Jones <pjones@redhat.com>
Date: Thu, 21 Jun 2018 11:13:39 -0400
Subject: [PATCH 17/34] Pacify clang analyzer just a little.

Signed-off-by: Peter Jones <pjones@redhat.com>
---
 src/linux-acpi-root.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/src/linux-acpi-root.c b/src/linux-acpi-root.c
index c7d8276a642..e55af5fa385 100644
--- a/src/linux-acpi-root.c
+++ b/src/linux-acpi-root.c
@@ -181,7 +181,6 @@ dp_create_acpi_root(struct device *dev,
                         return new;
                 }
         }
-        off += new;
         sz += new;
 
         debug(DEBUG, "returning %zd", sz);
-- 
2.17.1

