diff -Nuar a/src-qt5/core/lumina-desktop/defaults/luminaDesktop.conf b/src-qt5/core/lumina-desktop/defaults/luminaDesktop.conf
--- a/src-qt5/core/lumina-desktop/defaults/luminaDesktop.conf	2020-01-31 23:25:37.000000000 +0300
+++ b/src-qt5/core/lumina-desktop/defaults/luminaDesktop.conf	2021-04-21 11:34:15.509962649 +0300
@@ -50,18 +50,18 @@
 mime_default_application/x-shellscript=lumina-textedit.desktop
 
 #THEME SETTINGS
-theme_themefile=DarkGlass #Name of the theme to use
+theme_themefile=Glass #Name of the theme to use
 theme_styles=scrollbar-simple, tooltip-simple, sliders-simple, traynotification-simple
-theme_colorfile=darker #Name of the color spec file to use for theming
-theme_iconset=material-design-dark #Name of the icon theme to use
-theme_font=Arial #Name of the font family to use
-theme_fontsize=10pt #Default size of the fonts to use on the desktop (can also use a percentage of the screen height (<number>%) )
+theme_colorfile=default #Name of the color spec file to use for theming
+theme_iconset=material-design-light #Name of the icon theme to use
+theme_font=Monospace #Name of the font family to use
+theme_fontsize=11pt #Default size of the fonts to use on the desktop (can also use a percentage of the screen height (<number>%) )
 
 #DESKTOP SETTINGS (used for the primary screen in multi-screen setups)
 desktop_visiblepanels=2 #[0 - 12] The number of panels visible by default
-#desktop_backgroundfiles= #list of absolute file paths for image files (disable for Lumina default)
+desktop_backgroundfiles= /usr/share/lumina-desktop/Crocus-Ancyrensis.jpg #list of absolute file paths for image files (disable for Lumina default)
 desktop_backgroundrotateminutes=5 #[positive integer] number of minutes between background rotations (if multiple files)
-desktop_plugins=rssreader #list of plugins to be shown on the desktop by default
+desktop_plugins= notepad #list of plugins to be shown on the desktop by default
 desktop_generate_icons=true #[true/false] Auto-generate launchers for ~/Desktop items
 
 #PANEL SETTINGS (preface with panel1.<setting> or panel2.<setting>, depending on the number of panels you have visible by default)
