diff -Nuar a/src-qt5/core/lumina-desktop/Lumina-DE.desktop b/src-qt5/core/lumina-desktop/Lumina-DE.desktop
--- a/src-qt5/core/lumina-desktop/Lumina-DE.desktop	2020-01-31 23:25:37.000000000 +0300
+++ b/src-qt5/core/lumina-desktop/Lumina-DE.desktop	2020-11-19 23:45:29.669184104 +0300
@@ -32,3 +32,4 @@
 Comment[pt_BR]=Um ambiente de trabalho leve para FreeBSD
 Comment[uk]=Легковісне оточення стільниці для FreeBSD
 Comment[vi]=Một máy tính để bàn nhẹ cho FreeBSD
+OnlyShowIn=Lumina;
diff -Nuar a/src-qt5/core/lumina-desktop-unified/lumina-desktop-unified.desktop b/src-qt5/core/lumina-desktop-unified/lumina-desktop-unified.desktop
--- a/src-qt5/core/lumina-desktop-unified/lumina-desktop-unified.desktop	2020-01-31 23:25:37.000000000 +0300
+++ b/src-qt5/core/lumina-desktop-unified/lumina-desktop-unified.desktop	2020-11-19 23:46:01.690186457 +0300
@@ -5,3 +5,4 @@
 Type=Application
 Name=Lumina 2
 Comment=Illuminating the desktop
+OnlyShowIn=Lumina;
diff -Nuar a/src-qt5/core-utils/lumina-search/lumina-search.desktop b/src-qt5/core-utils/lumina-search/lumina-search.desktop
--- a/src-qt5/core-utils/lumina-search/lumina-search.desktop	2020-01-31 23:25:37.000000000 +0300
+++ b/src-qt5/core-utils/lumina-search/lumina-search.desktop	2020-11-19 23:47:58.839195068 +0300
@@ -7,3 +7,4 @@
 Categories=Utility;
 Name=Lumina Search
 Comment=Search for files or applications
+OnlyShowIn=Lumina;
diff -Nuar a/src-qt5/core-utils/lumina-xconfig/lumina-xconfig.desktop b/src-qt5/core-utils/lumina-xconfig/lumina-xconfig.desktop
--- a/src-qt5/core-utils/lumina-xconfig/lumina-xconfig.desktop	2020-01-31 23:25:37.000000000 +0300
+++ b/src-qt5/core-utils/lumina-xconfig/lumina-xconfig.desktop	2020-11-19 23:48:18.017196478 +0300
@@ -7,3 +7,4 @@
 Categories=Settings;
 Name=Lumina Screen Configuration
 Comment=Setup monitors and screens
+OnlyShowIn=Lumina;
diff -Nuar a/src-qt5/desktop-utils/lumina-archiver/lumina-archiver.desktop b/src-qt5/desktop-utils/lumina-archiver/lumina-archiver.desktop
--- a/src-qt5/desktop-utils/lumina-archiver/lumina-archiver.desktop	2020-01-31 23:25:37.000000000 +0300
+++ b/src-qt5/desktop-utils/lumina-archiver/lumina-archiver.desktop	2020-11-19 23:48:43.363198341 +0300
@@ -8,3 +8,4 @@
 MimeType=application/x-xz-compressed-tar;application/x-lzma-compressed-tar;application/x-bzip-compressed-tar;application/x-tar;application/x-compressed-tar;application/zip;application/gzip;application/bzip;application/x-cd-image;application/x-raw-disk-image;
 Name=Lumina Archiver
 Comment=Archive creator and viewer
+OnlyShowIn=Lumina;
diff -Nuar a/src-qt5/desktop-utils/lumina-fileinfo/lumina-fileinfo.desktop b/src-qt5/desktop-utils/lumina-fileinfo/lumina-fileinfo.desktop
--- a/src-qt5/desktop-utils/lumina-fileinfo/lumina-fileinfo.desktop	2020-01-31 23:25:37.000000000 +0300
+++ b/src-qt5/desktop-utils/lumina-fileinfo/lumina-fileinfo.desktop	2020-11-19 23:48:59.949199560 +0300
@@ -8,3 +8,4 @@
 Categories=Utility;
 Name=Lumina File Information
 Comment=View or edit file information including application links
+OnlyShowIn=Lumina;
diff -Nuar a/src-qt5/desktop-utils/lumina-fm/lumina-fm.desktop b/src-qt5/desktop-utils/lumina-fm/lumina-fm.desktop
--- a/src-qt5/desktop-utils/lumina-fm/lumina-fm.desktop	2020-01-31 23:25:37.000000000 +0300
+++ b/src-qt5/desktop-utils/lumina-fm/lumina-fm.desktop	2020-11-19 23:49:17.370200841 +0300
@@ -16,3 +16,4 @@
 Comment[fr_CA]=Accéder et utiliser les fichiers
 Comment[ja]=最近使ったファイル
 Comment[pl]=Dostęp i używanie plików
+OnlyShowIn=Lumina;
diff -Nuar a/src-qt5/desktop-utils/lumina-mediaplayer/lumina-mediaplayer.desktop b/src-qt5/desktop-utils/lumina-mediaplayer/lumina-mediaplayer.desktop
--- a/src-qt5/desktop-utils/lumina-mediaplayer/lumina-mediaplayer.desktop	2020-01-31 23:25:37.000000000 +0300
+++ b/src-qt5/desktop-utils/lumina-mediaplayer/lumina-mediaplayer.desktop	2020-11-19 23:49:30.894201835 +0300
@@ -8,3 +8,4 @@
 MimeType=
 Name=Lumina Media Player
 Comment=Play Multimedia Content
+OnlyShowIn=Lumina;
diff -Nuar a/src-qt5/desktop-utils/lumina-photo/lumina-photo.desktop b/src-qt5/desktop-utils/lumina-photo/lumina-photo.desktop
--- a/src-qt5/desktop-utils/lumina-photo/lumina-photo.desktop	2020-01-31 23:25:37.000000000 +0300
+++ b/src-qt5/desktop-utils/lumina-photo/lumina-photo.desktop	2020-11-19 23:49:49.446203199 +0300
@@ -8,3 +8,4 @@
 MimeType=
 Name=Lumina Image Viewer
 Comment=View photo
+OnlyShowIn=Lumina;
diff -Nuar a/src-qt5/desktop-utils/lumina-screenshot/lumina-screenshot.desktop b/src-qt5/desktop-utils/lumina-screenshot/lumina-screenshot.desktop
--- a/src-qt5/desktop-utils/lumina-screenshot/lumina-screenshot.desktop	2020-01-31 23:25:37.000000000 +0300
+++ b/src-qt5/desktop-utils/lumina-screenshot/lumina-screenshot.desktop	2020-11-19 23:50:02.758204177 +0300
@@ -7,3 +7,4 @@
 Categories=Graphics;
 Name=Lumina Screenshot
 Comment=Take Screenshot
+OnlyShowIn=Lumina;
diff -Nuar a/src-qt5/desktop-utils/lumina-textedit/lumina-textedit.desktop b/src-qt5/desktop-utils/lumina-textedit/lumina-textedit.desktop
--- a/src-qt5/desktop-utils/lumina-textedit/lumina-textedit.desktop	2020-01-31 23:25:37.000000000 +0300
+++ b/src-qt5/desktop-utils/lumina-textedit/lumina-textedit.desktop	2020-11-19 23:50:19.106205379 +0300
@@ -8,3 +8,4 @@
 MimeType=text/plain;text/x-log;text/*;
 Name=Lumina Text Editor
 Comment=Plaintext editor with syntax highlighting
+OnlyShowIn=Lumina;
diff -Nuar a/src-qt5/experimental/lumina-screencast/lumina-screencast.desktop b/src-qt5/experimental/lumina-screencast/lumina-screencast.desktop
--- a/src-qt5/experimental/lumina-screencast/lumina-screencast.desktop	2020-01-31 23:25:37.000000000 +0300
+++ b/src-qt5/experimental/lumina-screencast/lumina-screencast.desktop	2020-11-19 23:50:50.821207710 +0300
@@ -7,3 +7,4 @@
 Categories=Utility;
 Name=Lumina Screencast
 Comment=Record Screen
+OnlyShowIn=Lumina;
diff -Nuar a/src-qt5/experimental/lumina-terminal/lumina-terminal.desktop b/src-qt5/experimental/lumina-terminal/lumina-terminal.desktop
--- a/src-qt5/experimental/lumina-terminal/lumina-terminal.desktop	2020-01-31 23:25:37.000000000 +0300
+++ b/src-qt5/experimental/lumina-terminal/lumina-terminal.desktop	2020-11-19 23:51:07.908208966 +0300
@@ -7,3 +7,4 @@
 Categories=Utility;
 Name=Lumina Terminal
 Comment=Dropdown 
+OnlyShowIn=Lumina;
