--- chromium-59.0.3071.86/third_party/swiftshader/src/Reactor/BUILD.gn.gcc-5	2017-06-05 21:04:23.000000000 +0200
+++ chromium-59.0.3071.86/third_party/swiftshader/src/Reactor/BUILD.gn	2017-06-10 18:13:09.946401226 +0200
@@ -92,7 +92,7 @@
         cflags += [ "-Wno-sign-compare" ]
       }
     }
-  } else if (is_linux || is_mac) {
+  } else if (is_clang && (is_linux || is_mac)) {
     cflags += [ "-Wno-macro-redefined" ]
   }
 
--- chromium-59.0.3071.86/third_party/swiftshader/third_party/subzero/src/IceTargetLowering.cpp.gcc-5	2017-06-05 21:04:24.000000000 +0200
+++ chromium-59.0.3071.86/third_party/swiftshader/third_party/subzero/src/IceTargetLowering.cpp	2017-06-10 19:47:02.404476512 +0200
@@ -48,6 +48,7 @@
 //       createTargetHeaderLowering(Ice::GlobalContext *);
 //   void staticInit(::Ice::GlobalContext *);
 // }
+#undef SUBZERO_TARGET
 #define SUBZERO_TARGET(X)                                                      \
   namespace X {                                                                \
   std::unique_ptr<::Ice::TargetLowering>                                       \
