--- chromium-62.0.3192.0/cc/paint/paint_op_reader.cc.enum	2017-08-22 21:04:38.000000000 +0200
+++ chromium-62.0.3192.0/cc/paint/paint_op_reader.cc	2017-08-30 19:19:41.082455143 +0200
@@ -54,7 +54,7 @@
 
 bool IsValidSkShaderTileMode(SkShader::TileMode mode) {
   // When Skia adds Decal, update this (skbug.com/7638)
-  return mode <= SkShader::kMirror_TileMode;
+  return int(mode) <= SkShader::kMirror_TileMode;
 }
 
 bool IsValidPaintShaderScalingBehavior(PaintShader::ScalingBehavior behavior) {
