diff -up chromium-53.0.2785.46/build/linux/unbundle/opus.gn.gn-system-opus chromium-53.0.2785.46/build/linux/unbundle/opus.gn
--- chromium-53.0.2785.46/build/linux/unbundle/opus.gn.gn-system-opus	2016-08-09 23:46:12.808470292 +0200
+++ chromium-53.0.2785.46/build/linux/unbundle/opus.gn	2016-08-09 23:53:16.660141755 +0200
@@ -12,6 +12,7 @@
 shim_headers("opus_shim") {
   root_path = "src/include"
   headers = [
+    "opus_custom.h",
     "opus.h",
     "opus_defines.h",
     "opus_multistream.h",
