--- chromium-69.0.3497.4/tools/gn/build/gen.py.gn-bootstrap	2018-07-21 09:17:17.000000000 +0200
+++ chromium-69.0.3497.4/tools/gn/build/gen.py	2018-07-21 17:08:45.465590164 +0200
@@ -367,10 +367,8 @@
         # Use the sid sysroot that UpdateLinuxSysroot() downloads.
         cflags.append('--sysroot=' + linux_sysroot)
         ldflags.append('--sysroot=' + linux_sysroot)
-      ldflags.extend([
-          '-static-libstdc++',
-          '-Wl,--as-needed',
-      ])
+        ldflags.append('-static-libstdc++')
+      ldflags.append('-Wl,--as-needed')
       libs.extend([
           # These are needed by libc++.
           '-ldl',
