diff -Nuar src_old/ui/gui/psppire-dialog-action-regression.c src/ui/gui/psppire-dialog-action-regression.c
--- src_old/ui/gui/psppire-dialog-action-regression.c	2019-03-27 17:03:55.000000000 +0300
+++ src/ui/gui/psppire-dialog-action-regression.c	2019-03-30 19:32:50.869378647 +0300
@@ -41,7 +41,8 @@
   RG (CI,    N_("Conf. Interval"), N_("Show the confidence interval for the regression coefficients"))   \
   RG (R,     N_("R"),              N_("Show the correlation between observed and predicted values")) \
   RG (ANOVA, N_("Anova"),          N_("Show the analysis of variance table"))  \
-  RG (BCOV,  N_("Bcov"),           N_("Show the variance coefficient matrix"))
+  RG (BCOV,  N_("Bcov"),           N_("Show the variance coefficient matrix"))  \
+  RG (TOL,   N_("Tol"),            N_("Show variance inflation factor and its reciprocal"))
 
 enum
   {
