diff --git a/cmake/FindInterbase.cmake b/cmake/FindInterbase.cmake
index 22f866d8..4ef16e75 100644
--- a/cmake/FindInterbase.cmake
+++ b/cmake/FindInterbase.cmake
@@ -19,12 +19,12 @@
 #     The Interbase client library

 find_path(Interbase_INCLUDE_DIR
-          NAMES ibase.h
+          NAMES firebird/ibase.h
           HINTS ${Interbase_INCLUDEDIR}
 )

 find_library(Interbase_LIBRARY
-             NAMES firebase_ms fbclient gds
+             NAMES firebase_ms fbclient gds fbclient
              HINTS ${Interbase_LIBDIR}
 )

diff --git a/src/plugins/sqldrivers/ibase/qsql_ibase_p.h b/src/plugins/sqldrivers/ibase/qsql_ibase_p.h
index 9109c2b4..7477d5ee 100644
--- a/src/plugins/sqldrivers/ibase/qsql_ibase_p.h
+++ b/src/plugins/sqldrivers/ibase/qsql_ibase_p.h
@@ -52,7 +52,7 @@
 //
 
 #include <QtSql/qsqldriver.h>
-#include <ibase.h>
+#include <firebird/ibase.h>
 
 #ifdef QT_PLUGIN
 #define Q_EXPORT_SQLDRIVER_IBASE
