diff --git a/examples/corelib/tools/contiguouscache/randomlistmodel.h b/examples/corelib/tools/contiguouscache/randomlistmodel.h
index 1fabb0d9..393ebaa3 100644
--- a/examples/corelib/tools/contiguouscache/randomlistmodel.h
+++ b/examples/corelib/tools/contiguouscache/randomlistmodel.h
@@ -50,6 +50,7 @@
 #ifndef RANDOMLISTMODEL_H
 #define RANDOMLISTMODEL_H
 
+#include <limits>
 #include <QContiguousCache>
 #include <QAbstractListModel>
 
diff --git a/src/corelib/text/qanystringview.h b/src/corelib/text/qanystringview.h
index a7606253..60747cf0 100644
--- a/src/corelib/text/qanystringview.h
+++ b/src/corelib/text/qanystringview.h
@@ -39,6 +39,7 @@
 #ifndef QANYSTRINGVIEW_H
 #define QANYSTRINGVIEW_H
 
+#include <limits>
 #include <QtCore/qstringview.h>
 #include <QtCore/qutf8stringview.h>
 
diff --git a/src/corelib/text/qbytearray.h b/src/corelib/text/qbytearray.h
index 9f646aaa..a5af793c 100644
--- a/src/corelib/text/qbytearray.h
+++ b/src/corelib/text/qbytearray.h
@@ -41,6 +41,7 @@
 #ifndef QBYTEARRAY_H
 #define QBYTEARRAY_H
 
+#include <limits>
 #include <QtCore/qrefcount.h>
 #include <QtCore/qnamespace.h>
 #include <QtCore/qarraydata.h>
