diff -Nuar a/src/libply/ply-logger.h b/src/libply/ply-logger.h
--- a/src/libply/ply-logger.h	2017-08-09 15:48:58.000000000 +0300
+++ b/src/libply/ply-logger.h	2018-03-07 18:31:03.323594465 +0300
@@ -91,8 +91,8 @@
                         ply_logger_flush (logger);                                             \
                         errno = _old_errno;                                                    \
                         ply_logger_inject (logger,                                             \
-                                           "[%s:%d] %45.45s:" format "\n",                   \
-                                           __FILE__, __LINE__, __func__, ## args);              \
+                                           "" format "\n",                   \
+                                           ##args);              \
                         ply_logger_flush (logger);                                             \
                         errno = _old_errno;                                                    \
                 }                                                                        \
