Largefile patch
https://github.com/martine/ninja/issues/829

--- ninja-1.5.1/configure.py~	2014-10-26 22:45:18.000000000 +0200
+++ ninja-1.5.1/configure.py	2014-10-26 22:45:26.284641172 +0200
@@ -142,6 +142,7 @@
               '-fno-exceptions',
               '-fvisibility=hidden', '-pipe',
               '-Wno-missing-field-initializers',
+              '-D_LARGEFILE_SOURCE', '-D_FILE_OFFSET_BITS=64',
               '-DNINJA_PYTHON="%s"' % options.with_python]
     if options.debug:
         cflags += ['-D_GLIBCXX_DEBUG', '-D_GLIBCXX_DEBUG_PEDANTIC'] 
