diff -Nuar a/src/main.cpp b/src/main.cpp
--- a/src/main.cpp	2022-04-13 00:17:49.000000000 +0300
+++ b/src/main.cpp	2022-05-08 10:17:21.036971482 +0300
@@ -42,21 +42,21 @@
 {
 #ifndef Q_OS_WIN
     // Prohibit using sudo or kdesu (but allow using the root user directly)
-    if (getuid() == 0) {
-        if (!qEnvironmentVariableIsEmpty("SUDO_USER")) {
-            std::cout << "Running Dolphin with sudo can cause bugs and expose you to security vulnerabilities. "
-                         "Instead use Dolphin normally and you will be prompted for elevated privileges when "
-                         "performing file operations that require them."
-                      << std::endl;
-            return EXIT_FAILURE;
-        } else if (!qEnvironmentVariableIsEmpty("KDESU_USER")) {
-            std::cout << "Running Dolphin with kdesu can cause bugs and expose you to security vulnerabilities. "
-                         "Instead use Dolphin normally and you will be prompted for elevated privileges when "
-                         "performing file operations that require them."
-                      << std::endl;
-            return EXIT_FAILURE;
-        }
-    }
+    //if (getuid() == 0) {
+       // if (!qEnvironmentVariableIsEmpty("SUDO_USER")) {
+         //   std::cout << "Running Dolphin with sudo can cause bugs and expose you to security vulnerabilities. "
+           //              "Instead use Dolphin normally and you will be prompted for elevated privileges when "
+             //            "performing file operations that require them."
+               //       << std::endl;
+            //return EXIT_FAILURE;
+        //} else if (!qEnvironmentVariableIsEmpty("KDESU_USER")) {
+          //  std::cout << "Running Dolphin with kdesu can cause bugs and expose you to security vulnerabilities. "
+            //             "Instead use Dolphin normally and you will be prompted for elevated privileges when "
+              //           "performing file operations that require them."
+                //      << std::endl;
+            //return EXIT_FAILURE;
+        //}
+    //}
 #endif
 
     /**
