diff -Nuar cups_old-2.2.6/config-scripts/cups-directories.m4 cups-2.2.6/config-scripts/cups-directories.m4
--- cups_old-2.2.6/config-scripts/cups-directories.m4	2017-12-14 06:56:50.708946000 +0300
+++ cups-2.2.6/config-scripts/cups-directories.m4	2017-12-14 09:23:45.490213940 +0300
@@ -101,8 +101,8 @@
 if test "$libdir" = "\${exec_prefix}/lib"; then
 	case "$host_os_name" in
 		linux*)
-			if test -d /usr/lib64 -a ! -d /usr/lib64/fakeroot; then
-				libdir="$exec_prefix/lib64"
+			if test -d /usr/lib -a ! -d /usr/lib/fakeroot; then
+				libdir="$exec_prefix/lib"
 			fi
 			;;
 	esac
diff -Nuar cups_old-2.2.6/configure cups-2.2.6/configure
--- cups_old-2.2.6/configure	2017-12-14 06:56:51.766946000 +0300
+++ cups-2.2.6/configure	2017-12-14 09:26:43.667224765 +0300
@@ -6187,8 +6187,8 @@
 if test "$libdir" = "\${exec_prefix}/lib"; then
 	case "$host_os_name" in
 		linux*)
-			if test -d /usr/lib64 -a ! -d /usr/lib64/fakeroot; then
-				libdir="$exec_prefix/lib64"
+			if test -d /usr/lib -a ! -d /usr/lib/fakeroot; then
+				libdir="$exec_prefix/lib"
 			fi
 			;;
 	esac
diff -Nuar cups_old-2.2.6/cups-config.in cups-2.2.6/cups-config.in
--- cups_old-2.2.6/cups-config.in	2017-11-01 17:57:53.000000000 +0300
+++ cups-2.2.6/cups-config.in	2017-12-14 09:22:37.641209818 +0300
@@ -48,7 +48,7 @@
 	CFLAGS="$CFLAGS -I$includedir"
     fi
 
-    if test $libdir != /usr/lib -a $libdir != /usr/lib32 -a $libdir != /usr/lib64; then
+    if test $libdir != /usr/lib -a $libdir != /usr/lib32 -a $libdir != /usr/lib; then
 	LDFLAGS="$LDFLAGS -L$libdir"
     fi
 fi
