diff -Nuar a/conf/pam.password b/conf/pam.password
--- a/conf/pam.password	2024-06-11 17:20:10.000000000 +0300
+++ b/conf/pam.password	2024-06-13 22:52:56.242279280 +0300
@@ -2,3 +2,4 @@
 # Use password-auth common PAM configuration for the daemon
 auth     include  password-auth
 account  include  password-auth
+session  include  password-auth
diff -Nuar a/conf/pam.system b/conf/pam.system
--- a/conf/pam.system	2024-06-11 17:20:10.000000000 +0300
+++ b/conf/pam.system	2024-06-13 22:52:14.370802746 +0300
@@ -1,3 +1,4 @@
 #%PAM-1.0
 auth     include  system-auth
 account  include  system-auth
+session  include  system-session
