diff -Nuar pyblock-0.53.orig/Makefile pyblock-0.53/Makefile
--- pyblock-0.53.orig/Makefile	2016-03-30 09:41:42.510926755 +0300
+++ pyblock-0.53/Makefile	2016-03-30 09:42:32.328924813 +0300
@@ -5,19 +5,19 @@
 DEBUGLIB := "/usr/lib/debug"
 CC := gcc
 PYVER := $(shell python -c "import sys; print sys.version[0:3]")
-CFLAGS := -g -I/usr/include/python$(PYVER) -Wall -Werror -fPIC $(OPTFLAGS) -fno-strict-aliasing
+CFLAGS := -g -I/usr/include/python$(PYVER) -Wall -fPIC $(OPTFLAGS) -fno-strict-aliasing
 LDFLAGS := -shared
 
 VERSION := $(shell awk '/Version:/ { print $$2 }' python-pyblock.spec)
 RELEASE := $(shell awk -F '[ %]' '/Release:/ { print $$2 }' python-pyblock.spec)
 USESELINUX = 1
 
-dm_LIBS = dmraid devmapper
+dm_LIBS = devmapper python2.7
 ifeq (1, $(USESELINUX))
 dm_LIBS += selinux
 CFLAGS += -DUSESELINUX=1
 endif
-dmraid_LIBS = dmraid devmapper
+dmraid_LIBS = dmraid python2.7
 
 PYFILES=__init__.py maps.py device.py
 LIBS = dmmodule.so dmraidmodule.so
