diff -Nuar a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2018-03-06 15:38:53.000000000 +0300
+++ b/CMakeLists.txt	2018-03-23 02:24:03.000651214 +0300
@@ -9,7 +9,6 @@
 find_package(KF5 REQUIRED COMPONENTS I18n KIO ConfigWidgets NewStuff Archive KCMUtils IconThemes)
 find_package(X11 REQUIRED)
 find_package(GTK3 REQUIRED)
-find_package(GSettingSchemas REQUIRED)
 
 include_directories(
     ${CMAKE_SOURCE_DIR}
