diff -Nuar a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2020-06-04 14:08:39.000000000 +0300
+++ b/CMakeLists.txt	2020-06-09 15:57:21.313347106 +0300
@@ -12,7 +12,6 @@
 find_package(KF5GuiAddons REQUIRED)
 find_package(KF5DBusAddons REQUIRED)
 find_package(PkgConfig REQUIRED)
-find_package(GSettingSchemas REQUIRED)
 find_package(XSettingsd)
 
 pkg_check_modules(GTK+3 REQUIRED IMPORTED_TARGET gtk+-3.0)
