--- a/configure.in	2020-09-13 18:27:24.186893066 +0300
+++ b/configure.in	2020-09-13 18:28:54.905398925 +0300
@@ -3,7 +3,8 @@
 AC_CONFIG_MACRO_DIR([m4])
 AC_CONFIG_SRCDIR([libmpcdec/mpc_reader.c])
 AM_CONFIG_HEADER([include/config.h])
-AM_INIT_AUTOMAKE
+AM_INIT_AUTOMAKE(subdir-objects)
+AM_MAINTAINER_MODE
 
 AC_LANG_C
 AC_PROG_CC
@@ -30,7 +31,7 @@
 AM_CONDITIONAL([MPC_CHAP], [test "x$enable_mpcchap" = xyes])
 
 
-CHECK_VISIBILITY
+AM_CONDITIONAL([HAVE_VISIBILITY], [true])
 
 AC_CONFIG_FILES([
 Makefile
