diff -Nuar a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2017-12-02 10:21:00.000000000 +0300
+++ b/CMakeLists.txt	2017-12-16 13:09:19.487857629 +0300
@@ -23,7 +23,7 @@
 if(POLICY CMP0053)
   cmake_policy(SET CMP0053 NEW)
 endif()
-set(QT_REQUIRED_VERSION "5.8.0")
+set(QT_REQUIRED_VERSION "5.7.1")
 
 set(KCALENDARCORE_LIB_VERSION ${PIM_VERSION})
 ecm_setup_version(PROJECT VARIABLE_PREFIX KCALCORE
