diff -up oce-upstream-V7_5_0/src/Standard/Standard_Type.hxx.omv~ oce-upstream-V7_5_0/src/Standard/Standard_Type.hxx
--- oce-upstream-V7_5_0/src/Standard/Standard_Type.hxx.omv~	2022-01-09 01:03:31.512273977 +0100
+++ oce-upstream-V7_5_0/src/Standard/Standard_Type.hxx	2022-01-09 01:03:38.866283692 +0100
@@ -26,7 +26,7 @@
 // DEFINE_STANDARD_RTTI* macro is actually a base class.
 #if ! defined(OCCT_CHECK_BASE_CLASS)
 
-#if (defined(__GNUC__) && ((__GNUC__ == 4 && __GNUC_MINOR__ >= 7) || (__GNUC__ > 4)))
+#if (defined(__GNUC__) && ((__GNUC__ == 4 && __GNUC_MINOR__ >= 7) || (__GNUC__ > 4))) && !defined(__clang__)
 
 // For GCC 4.7+, more strict check is possible -- ensuring that base class 
 // is direct base -- using non-standard C++ reflection functionality.
