diff -up opencascade-7.6.0/adm/templates/custom.build.sh.in.omv~ opencascade-7.6.0/adm/templates/custom.build.sh.in
--- opencascade-7.6.0/adm/templates/custom.build.sh.in.omv~	2022-01-12 04:05:12.227180642 +0100
+++ opencascade-7.6.0/adm/templates/custom.build.sh.in	2022-01-12 04:06:30.088406684 +0100
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-if [ "$1" == "@BIN_LETTER@" ]; then
+if true; then
   if [ "$2" == "@COMPILER_BITNESS@" ]; then
     # set environment variables used by OCCT
     export CSF_FPE=@BUILD_ENABLE_FPE_SIGNAL_HANDLER@
diff -up opencascade-7.6.0/adm/templates/env.sh.omv~ opencascade-7.6.0/adm/templates/env.sh
