diff -Nuar a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2017-12-08 04:04:03.000000000 +0300
+++ b/CMakeLists.txt	2017-12-16 22:15:57.334597567 +0300
@@ -25,7 +25,7 @@
 
 set(libkdepim_LIB_VERSION ${PIM_VERSION})
 set(AKONADICONTACT_LIB_VERSION "5.7.0")
-set(QT_REQUIRED_VERSION "6.6.3")
+set(QT_REQUIRED_VERSION "6.6.3")
 set(kldap_LIB_VERSION "5.7.0")
 set(kcontacts_LIB_VERSION "5.7.0")
 set(AKONADISEARCH_LIB_VERSION "5.7.0")
diff -Nuar a/src/libkdepimakonadi/addressline/addresslineedit/addresseelineedit_p.cpp b/src/libkdepimakonadi/addressline/addresslineedit/addresseelineedit_p.cpp
--- a/src/libkdepimakonadi/addressline/addresslineedit/addresseelineedit_p.cpp	2017-12-04 13:20:38.000000000 +0300
+++ b/src/libkdepimakonadi/addressline/addresslineedit/addresseelineedit_p.cpp	2017-12-16 22:23:46.137563740 +0300
@@ -498,7 +498,9 @@
             break;
         }
         //else: fall-through to the CompletionPopup case
+#if QT_VERSION >= QT_VERSION_CHECK(5,8,0)
         Q_FALLTHROUGH();
+#endif
 
     case kcompletion::CompletionPopup:
     {
