diff -Nuar a/data/lightdm.conf b/data/lightdm.conf
--- a/data/lightdm.conf	2021-04-29 15:59:46.714211010 +0300
+++ b/data/lightdm.conf	2021-04-29 16:01:37.594219160 +0300
@@ -99,7 +99,7 @@
 #xdmcp-manager=
 #xdmcp-port=177
 #xdmcp-key=
-#greeter-session=example-gtk-gnome
+greeter-session=pisi-lightdm-greeter
 #greeter-hide-users=false
 #greeter-allow-guest=true
 #greeter-show-manual-login=false
@@ -108,7 +108,7 @@
 #allow-user-switching=true
 #allow-guest=true
 #guest-session=
-#session-wrapper=lightdm-session
+session-wrapper=/usr/bin/wrapper
 #greeter-wrapper=
 #guest-wrapper=
 #display-setup-script=
diff -Nuar a/data/users.conf b/data/users.conf
--- a/data/users.conf	2021-04-29 16:00:22.690213654 +0300
+++ b/data/users.conf	2021-04-29 16:01:45.322219728 +0300
@@ -9,6 +9,6 @@
 # hidden-shells = Shells that indicate a user cannot login
 #
 [UserList]
-minimum-uid=500
+minimum-uid=1000
 hidden-users=nobody nobody4 noaccess
 hidden-shells=/bin/false /usr/sbin/nologin /sbin/nologin
