--- lightdm-1.15.0.orig/data/lightdm.conf	2015-05-14 07:12:11.000000000 +0300
+++ lightdm-1.15.0/data/lightdm.conf	2016-09-07 06:27:09.913001757 +0300
@@ -106,7 +106,7 @@
 #allow-user-switching=true
 #allow-guest=true
 #guest-session=
-#session-wrapper=lightdm-session
+session-wrapper=/usr/bin/wrapper
 #greeter-wrapper=
 #guest-wrapper=
 #display-setup-script=
--- lightdm-1.15.0.orig/data/users.conf	2014-05-19 23:10:03.000000000 +0300
+++ lightdm-1.15.0/data/users.conf	2016-09-07 06:27:25.985001874 +0300
@@ -9,6 +9,6 @@
 # hidden-shells = Shells that indicate a user cannot login
 #
 [UserList]
-minimum-uid=500
+minimum-uid=1000
 hidden-users=nobody nobody4 noaccess
 hidden-shells=/bin/false /usr/sbin/nologin
