diff -Nuar a/data/pam/lightdm-autologin b/data/pam/lightdm-autologin
--- a/data/pam/lightdm-autologin	2018-04-10 06:07:52.000000000 +0300
+++ b/data/pam/lightdm-autologin	2021-05-02 15:11:22.125900705 +0300
@@ -17,4 +17,4 @@
 
 # Setup session
 session   required pam_unix.so
-session   optional pam_systemd.so
+# session   optional pam_systemd.so
diff -Nuar a/data/pam/lightdm-greeter b/data/pam/lightdm-greeter
--- a/data/pam/lightdm-greeter	2018-04-10 06:07:52.000000000 +0300
+++ b/data/pam/lightdm-greeter	2021-05-02 15:11:33.473899886 +0300
@@ -14,4 +14,4 @@
 
 # Setup session
 session   required pam_unix.so
-session   optional pam_systemd.so
+# session   optional pam_systemd.so
