diff -Nuar a/data/lightdm.conf b/data/lightdm.conf
--- a/data/lightdm.conf	2021-05-01 12:49:09.402828071 +0300
+++ b/data/lightdm.conf	2021-05-01 12:51:47.210825962 +0300
@@ -99,7 +99,7 @@
 #xdmcp-manager=
 #xdmcp-port=177
 #xdmcp-key=
-#greeter-session=example-gtk-gnome
+greeter-session=pisi-lightdm-greeter
 #greeter-hide-users=false
 #greeter-allow-guest=true
 #greeter-show-manual-login=false
@@ -108,7 +108,7 @@
 #allow-user-switching=true
 #allow-guest=true
 #guest-session=
-#session-wrapper=lightdm-session
+#session-wrapper=/usr/bin/wrapper
 #greeter-wrapper=
 #guest-wrapper=
 #display-setup-script=
@@ -117,8 +117,8 @@
 #session-setup-script=
 #session-cleanup-script=
-#autologin-guest=false
+autologin-guest=false
-#autologin-user=
-#autologin-user-timeout=0
+autologin-user=pisi
+autologin-user-timeout=0
 #autologin-in-background=false
 #autologin-session=
 #exit-on-failure=false
diff -Nuar a/data/users.conf b/data/users.conf
--- a/data/users.conf	2021-05-01 12:49:38.754827679 +0300
+++ b/data/users.conf	2021-05-01 12:49:59.619827400 +0300
@@ -9,6 +9,6 @@
 # hidden-shells = Shells that indicate a user cannot login
 #
 [UserList]
-minimum-uid=500
+minimum-uid=1000
 hidden-users=nobody nobody4 noaccess
 hidden-shells=/bin/false /usr/sbin/nologin /sbin/nologin
