diff --git a/startkde/startkde.cmake b/startkde/startkde.cmake
index 91c8a3c..0c658bd 100644
--- a/startkde/startkde.cmake
+++ b/startkde/startkde.cmake
@@ -20,13 +20,13 @@ unset DYLD_FORCE_FLAT_NAMESPACE
 bindir=`echo "$0" | sed -n 's,^\(/.*\)/[^/][^/]*$,\1,p'`
 if [ -n "$bindir" ]; then
   qbindir=`qtpaths --binaries-dir`
-  qdbus=$qbindir/qdbus
+  qdbus=$qbindir/qdbus-qt5
   case $PATH in
     $bindir|$bindir:*|*:$bindir|*:$bindir:*) ;;
     *) PATH=$bindir:$PATH; export PATH;;
   esac
 else
-  qdbus=qdbus
+  qdbus=qdbus-qt5
 fi
 
 # Check if a KDE session already is running and whether it's possible to connect to X
@@ -300,9 +300,9 @@ fi
 if $qdbus >/dev/null 2>/dev/null; then
     : # ok
 else
-    echo 'startkde: Could not start D-Bus. Can you call qdbus?'  1>&2
+    echo 'startkde: Could not start D-Bus. Can you call qdbus-qt5?'  1>&2
     test -n "$ksplash_pid" && kill "$ksplash_pid" 2>/dev/null
-    xmessage -geometry 500x100 "Could not start D-Bus. Can you call qdbus?"
+    xmessage -geometry 500x100 "Could not start D-Bus. Can you call qdbus-qt5?"
     exit 1
 fi
 
