Index: chromium-73.0.3683.27/components/version_ui_strings.grdp
===================================================================
--- chromium-73.0.3683.27.orig/components/version_ui_strings.grdp
+++ chromium-73.0.3683.27/components/version_ui_strings.grdp
@@ -4,7 +4,7 @@
     About Version
   </message>
   <message name="IDS_VERSION_UI_OFFICIAL" desc="official build on the about:version page">
-    Official Build
+    PisiLinux Build
   </message>
   <message name="IDS_VERSION_UI_UNOFFICIAL" desc="unofficial build on the about:version page">
     Developer Build