diff -Nru chromium-77.0.3865.93.orig/third_party/blink/renderer/modules/webgl/khr_parallel_shader_compile.cc chromium-77.0.3865.93/third_party/blink/renderer/modules/webgl/khr_parallel_shader_compile.cc
--- chromium-77.0.3865.93.orig/third_party/blink/renderer/modules/webgl/khr_parallel_shader_compile.cc  2019-09-20 01:18:30.000000000 +0300
+++ chromium-77.0.3865.93/third_party/blink/renderer/modules/webgl/khr_parallel_shader_compile.cc       2019-09-23 03:44:32.589602845 +0300
@@ -24,6 +24,7 @@
  */
 
 #include "third_party/blink/renderer/modules/webgl/khr_parallel_shader_compile.h"
+#include "third_party/blink/renderer/platform/graphics/canvas_2d_layer_bridge.h"
 
 #include <thread>
 
diff -Nru chromium-77.0.3865.93.orig/third_party/blink/renderer/modules/webgl/oes_element_index_uint.cc chromium-77.0.3865.93/third_party/blink/renderer/modules/webgl/oes_element_index_uint.cc
--- chromium-77.0.3865.93.orig/third_party/blink/renderer/modules/webgl/oes_element_index_uint.cc       2019-09-20 01:18:30.000000000 +0300
+++ chromium-77.0.3865.93/third_party/blink/renderer/modules/webgl/oes_element_index_uint.cc    2019-09-22 21:07:19.163921146 +0300
@@ -24,6 +24,7 @@
  */
 
 #include "third_party/blink/renderer/modules/webgl/oes_element_index_uint.h"
+#include "third_party/blink/renderer/platform/graphics/canvas_2d_layer_bridge.h"
 
 namespace blink {
 
diff -Nru chromium-77.0.3865.93.orig/third_party/blink/renderer/modules/vr/navigator_vr.cc chromium-77.0.3865.93/third_party/blink/renderer/modules/vr/navigator_vr.cc
--- chromium-77.0.3865.93.orig/third_party/blink/renderer/modules/vr/navigator_vr.cc    2019-09-20 01:18:30.000000000 +0300
+++ chromium-77.0.3865.93/third_party/blink/renderer/modules/vr/navigator_vr.cc 2019-09-22 16:29:52.676841915 +0300
@@ -18,6 +18,7 @@
 #include "third_party/blink/renderer/modules/xr/navigator_xr.h"
 #include "third_party/blink/renderer/platform/heap/heap.h"
 #include "third_party/blink/renderer/platform/instrumentation/use_counter.h"
+#include "third_party/blink/renderer/platform/graphics/canvas_2d_layer_bridge.h"
 
 namespace blink {
 
diff -Nru chromium-77.0.3865.93.orig/third_party/blink/renderer/modules/vr/vr_controller.cc chromium-77.0.3865.93/third_party/blink/renderer/modules/vr/vr_controller.cc
--- chromium-77.0.3865.93.orig/third_party/blink/renderer/modules/vr/vr_controller.cc   2019-09-20 01:18:30.000000000 +0300
+++ chromium-77.0.3865.93/third_party/blink/renderer/modules/vr/vr_controller.cc        2019-09-21 20:59:54.174954333 +0300
@@ -12,6 +12,7 @@
 #include "third_party/blink/renderer/core/frame/local_frame.h"
 #include "third_party/blink/renderer/modules/vr/navigator_vr.h"
 #include "third_party/blink/renderer/platform/wtf/assertions.h"
+#include "third_party/blink/renderer/platform/graphics/canvas_2d_layer_bridge.h"
 
 namespace blink {
 
 
