diff --git a/CMakeLists.txt b/CMakeLists.txt
index a8e7de035e..9fe5440174 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -127,12 +127,6 @@ set_package_properties(Threads PROPERTIES
     TYPE REQUIRED
 )
 
-find_package(KWayland ${PROJECT_DEP_VERSION} CONFIG)
-set_package_properties(KWayland PROPERTIES
-    PURPOSE "Required to build wayland platform plugin and tests"
-    TYPE REQUIRED
-)
-
 # optional frameworks
 find_package(PlasmaActivities ${PROJECT_DEP_VERSION} CONFIG)
 set_package_properties(PlasmaActivities PROPERTIES
@@ -341,9 +335,6 @@ else()
     set(HAVE_GBM_BO_CREATE_WITH_MODIFIERS2 0)
 endif()
 
-pkg_check_modules(Libxcvt IMPORTED_TARGET libxcvt>=0.1.1 REQUIRED)
-add_feature_info(Libxcvt Libxcvt_FOUND "Required for generating modes in the drm backend")
-
 add_feature_info("XInput" X11_Xi_FOUND "Required for poll-free mouse cursor updates")
 set(HAVE_X11_XINPUT ${X11_Xinput_FOUND})
 
@@ -509,6 +500,12 @@ add_subdirectory(kconf_update)
 add_subdirectory(src)
 
 if (BUILD_TESTING)
+    find_package(KWayland ${PROJECT_DEP_VERSION} CONFIG)
+    set_package_properties(KWayland PROPERTIES
+        PURPOSE "Required to build wayland platform plugin and tests"
+        TYPE REQUIRED
+    )
+
     add_subdirectory(autotests)
     add_subdirectory(tests)
 endif()
