Index: src/mpvprocess.cpp
===================================================================
--- src/mpvprocess.cpp	(revision 9231)
+++ src/mpvprocess.cpp	(revision 9232)
@@ -122,7 +122,7 @@
 
 void MPVProcess::initializeRX() {
 #ifdef CUSTOM_STATUS
-	rx_av.setPattern("^STATUS: ([0-9\\.-]+) / ([0-9\\.-]+) P: (yes|no) B: (yes|no) I: (yes|no) VB: ([0-9\\.-]+) AB: ([0-9\\.-]+)");
+	rx_av.setPattern("STATUS: ([0-9\\.-]+) / ([0-9\\.-]+) P: (yes|no) B: (yes|no) I: (yes|no) VB: ([0-9\\.-]+) AB: ([0-9\\.-]+)");
 #else
 	rx_av.setPattern("^(\\((.*)\\) |)(AV|V|A): ([0-9]+):([0-9]+):([0-9]+) / ([0-9]+):([0-9]+):([0-9]+)"); //AV: 00:02:15 / 00:09:56
 #endif
@@ -197,7 +197,7 @@
 	static double last_sec = -1;
 
 	// Parse A: V: line
-	//qDebug("MPVProcess::parseLine: %s", line.toUtf8().data());
+	//qDebug() << "MPVProcess::parseLine:" << line;
 	if (rx_av.indexIn(line) > -1) {
 		#ifdef CUSTOM_STATUS
 		double sec = rx_av.cap(1).toDouble();
Index: src/mpvoptions.cpp
===================================================================
--- src/mpvoptions.cpp	(revision 9233)
+++ src/mpvoptions.cpp	(revision 9234)
@@ -63,7 +63,8 @@
 	arg << "--term-playing-msg="
 			"MPV_VERSION=${=mpv-version:}\n"
 			"INFO_VIDEO_WIDTH=${=width}\nINFO_VIDEO_HEIGHT=${=height}\n"
-			"INFO_VIDEO_ASPECT=${=video-aspect}\n"
+//			"INFO_VIDEO_ASPECT=${=video-aspect}\n" // old
+			"INFO_VIDEO_ASPECT=${=video-params/aspect}\n"
 //			"INFO_VIDEO_DSIZE=${=dwidth}x${=dheight}\n"
 			"INFO_VIDEO_FPS=${=container-fps:${=fps}}\n"
 //			"INFO_VIDEO_BITRATE=${=video-bitrate}\n"
