diff -Nuar a/src/loginjob.cpp b/src/loginjob.cpp
--- a/src/loginjob.cpp	2017-12-04 13:20:25.000000000 +0300
+++ b/src/loginjob.cpp	2017-12-16 15:57:30.307129526 +0300
@@ -425,7 +425,9 @@
         case LoginJobPrivate::Authenticate:
             sasl_dispose(&d->conn);   //SASL authentication done
         // Fall through
+#if QT_VERSION >= QT_VERSION_CHECK(5,8,0)
             Q_FALLTHROUGH();
+#endif
         case LoginJobPrivate::Login:
             d->saveServerGreeting(response);
             emitResult(); //got an OK, command done
