--- a/src/servicemanager/widgets.py	2020-04-04 13:04:20.000000000 +0300
+++ b/src/servicemanager/widgets.py	2021-11-27 14:35:17.755051834 +0300
@@ -95,7 +95,7 @@
         self.ui.buttonStop.setEnabled(self.running)
         self.ui.buttonReload.setEnabled(self.running)
 
-        self.ui.labelStatus.setPixmap(KIcon(icon).pixmap(32, 32))
+        self.ui.labelStatus.setPixmap(QIcon(icon).pixmap(32, 32))
         self.showStatus()
         self.runningAtStart = False
         if state in ('on', 'stopped'):
