diff -Nuar a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2017-12-08 04:01:29.000000000 +0300
+++ b/CMakeLists.txt	2017-12-16 13:20:44.507808201 +0300
@@ -4,7 +4,7 @@
 project(kcontacts5 VERSION ${PIM_VERSION})
 
 # ECM setup
-set(QT_REQUIRED_VERSION "5.8.0")
+set(QT_REQUIRED_VERSION "5.7.1")
 set(KF5_VERSION "5.39.0")
 
 find_package(ECM ${KF5_VERSION} CONFIG REQUIRED)
diff -Nuar a/src/vcardparser/vcardparser.cpp b/src/vcardparser/vcardparser.cpp
--- a/src/vcardparser/vcardparser.cpp	2017-12-02 10:20:32.000000000 +0300
+++ b/src/vcardparser/vcardparser.cpp	2017-12-16 13:28:28.299774737 +0300
@@ -183,7 +183,9 @@
                 start = pos + 1;
             }
             // fall-through intended
+#if QT_VERSION >= QT_VERSION_CHECK(5,8,0)
             Q_FALLTHROUGH();
+#endif
         case StateAfterParamValue:
             if (ch == ';') {
                 state = StateParamKey;
