--- chromium-60.0.3112.40/chrome/browser/extensions/component_extensions_whitelist/whitelist.cc.no-external-components	2017-06-23 23:37:14.038911768 +0200
+++ chromium-60.0.3112.40/chrome/browser/extensions/component_extensions_whitelist/whitelist.cc	2017-06-23 23:49:00.077985242 +0200
@@ -26,8 +26,10 @@
 
 bool IsComponentExtensionWhitelisted(const std::string& extension_id) {
   const char* const kAllowed[] = {
+#if 0
     extension_misc::kInAppPaymentsSupportAppId,
     extension_misc::kMediaRouterStableExtensionId,
+#endif
     extension_misc::kPdfExtensionId,
 #if defined(OS_CHROMEOS)
     extension_misc::kAssessmentAssistantExtensionId,
