diff -Nuar a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2017-12-08 04:04:08.000000000 +0300
+++ b/CMakeLists.txt	2017-12-16 22:30:39.116533942 +0300
@@ -24,7 +24,7 @@
 
 
 set(libkleo_LIB_VERSION ${PIM_VERSION})
-set(QT_REQUIRED_VERSION "6.9.0")
+set(QT_REQUIRED_VERSION "6.9.0")
 set(KDEPIMTEXTEDIT_VERSION "5.7.0")
 
 find_package(Qt5 ${QT_REQUIRED_VERSION} CONFIG REQUIRED Widgets)
diff -Nuar a/src/kleo/dn.cpp b/src/kleo/dn.cpp
--- a/src/kleo/dn.cpp	2017-12-04 13:20:17.000000000 +0300
+++ b/src/kleo/dn.cpp	2017-12-16 22:35:27.941513102 +0300
@@ -298,7 +298,9 @@
         case ';':
             result += QLatin1Char('\\');
             // fall through
+#if QT_VERSION >= QT_VERSION_CHECK(5,8,0)
             Q_FALLTHROUGH();
+#endif
         default:
             result += ch;
         }
diff -Nuar a/src/utils/formatting.cpp b/src/utils/formatting.cpp
--- a/src/utils/formatting.cpp	2017-12-04 13:20:17.000000000 +0300
+++ b/src/utils/formatting.cpp	2017-12-16 22:40:08.089492888 +0300
@@ -593,7 +593,9 @@
                 return i18nc("good/valid signature", "good");
             }
         }
+#if QT_VERSION >= QT_VERSION_CHECK(5,8,0)
         Q_FALLTHROUGH();
+#endif
         // fall through:
     case UserID::Signature::GeneralError:
         return i18n("invalid");
