diff -Nur openvr-1.16.8/src/vrcommon/strtools_public.cpp new/src/vrcommon/strtools_public.cpp
--- openvr-1.16.8/src/vrcommon/strtools_public.cpp	2021-03-28 20:07:17.901245911 +0200
+++ new/src/vrcommon/strtools_public.cpp	2021-03-28 20:08:25.491270144 +0200
@@ -1,6 +1,7 @@
 //========= Copyright Valve Corporation ============//
 #include "vrcommon/strtools_public.h"
 //#include "vrcommon/assert.h"
+#include <stdarg.h>
 #include <string.h>
 #include <stdio.h>
 #include <stdlib.h>
