diff -Nuar a/pds/environments.py b/pds/environments.py
--- a/pds/environments.py	2017-08-30 19:34:22.000000000 +0300
+++ b/pds/environments.py	2018-10-02 14:50:09.075085646 +0300
@@ -30,7 +30,7 @@
 # kde5 için sınıfı düzenlenecek ve kde5 için bilgiler istenecek!!!
 class Kde5(DefaultDe):
     Name                 = 'kde'
-    SessionTypes         = ('/usr/share/xsessions/plasma-mediacenter')
+    SessionTypes         = ('/usr/share/xsessions/plasma')
     Version              = '5'
     VersionKey           = 'KDE_SESSION_VERSION'
     ConfigPath           = ('$HOME/.config/', '$HOME/.config5/')
diff -Nuar a/pds/environments.py b/pds/environments.py
--- a/pds/environments.py	2017-08-30 19:34:22.000000000 +0300
+++ b/pds/environments.py	2018-10-21 22:44:52.389522233 +0300
@@ -33,7 +33,7 @@
     SessionTypes         = ('/usr/share/xsessions/plasma-mediacenter')
     Version              = '5'
     VersionKey           = 'KDE_SESSION_VERSION'
-    ConfigPath           = ('$HOME/.config/', '$HOME/.config5/')
+    ConfigPath           = ('$HOME/.config/', '$HOME/.config/')
     ConfigFile           = 'kdeglobals'
     ConfigType           = 'ini'
     #ConfigBin            = 'kde4-config'
