--- pds-2.0.0.orig/pds/environments.py	2016-03-02 11:27:31.000000000 +0200
+++ pds-2.0.0/pds/environments.py	2016-06-03 15:42:26.860036236 +0300
@@ -33,7 +33,7 @@
     SessionTypes         = ('/usr/share/xsessions/plasma-mediacenter')
     Version              = '5'
     VersionKey           = 'KDE_SESSION_VERSION'
-    ConfigPath           = ('$HOME/.local5/', '$HOME/.kde4/')
+    ConfigPath           = ('$HOME/.local/', '$HOME/.kde/')
     ConfigFile           = 'share/config/kdeglobals'
     ConfigType           = 'ini'
     #ConfigBin            = 'kde4-config'
