diff -Nuar a/pds/__init__.py b/pds/__init__.py
--- a/pds/__init__.py	2021-05-21 12:07:38.000000000 +0300
+++ b/pds/__init__.py	2021-06-08 23:04:07.126974303 +0300
@@ -158,7 +158,7 @@
                     elif v[1] == 'DESKTOP':
                         env = v[0].lower()
                         break
-            elif env == 'default' or env == 'gnome':
+            elif env == 'default' or env == 'Gnome':
                 session = readfile('/etc/default/desktop', DefaultDe.Name)
                 env     = session.split('=')[1].strip()
             for de in Pds.SupportedDesktops:
