diff -Nuar a/pds/environments.py b/pds/environments.py
--- a/pds/environments.py	2022-06-02 13:10:01.000000000 +0300
+++ b/pds/environments.py	2023-02-05 14:09:42.383765629 +0300
@@ -41,7 +41,7 @@
     IconKey              = 'Icons/Theme'
 
 class Xfce(DefaultDe):
-    Name                 = 'xfce'
+    Name                 = 'xfce4'
     Version              = '4'
     ConfigPath           = '$HOME/.config/xfce4/'
     ConfigFile           = 'xfconf/xfce-perchannel-xml/xsettings.xml'
