diff --git a/tests/fonts/DejaVuSansMono.ttf b/tests/fonts/DejaVuSansMono.ttf
new file mode 100644
index 0000000000000000000000000000000000000000..d723948fed5dc9d2e63f57af9f80f2a472aff898
Binary files /dev/null and b/tests/fonts/DejaVuSansMono.ttf differ
diff --git a/tests/fonts/fonts.conf b/tests/fonts/fonts.conf
index 5ff7022b727b37ef2d52efd2738682baa8af9280..4308770bc153dfae16639b2c6d8b4eb6822b0db2 100644
--- a/tests/fonts/fonts.conf
+++ b/tests/fonts/fonts.conf
@@ -62,4 +62,10 @@
       <family>Noto Color Emoji</family>
     </prefer>
   </alias>
+
+  <alias>
+    <family>DejaVu Sans Mono</family>
+    <default><family>monospace</family></default>
+  </alias>
+
 </fontconfig>
diff --git a/tests/layouts/valid-12.layout b/tests/layouts/valid-12.layout
index 099813cefb41cf5f93040323d98342c4c06a0970..3c09f51f9b22cbde88bebd139296cf34b25052dd 100644
--- a/tests/layouts/valid-12.layout
+++ b/tests/layouts/valid-12.layout
@@ -52,8 +52,8 @@
     "is-wrapped" : true,
     "is-ellipsized" : false,
     "unknown-glyphs" : 1,
-    "width" : 202752,
-    "height" : 202752,
+    "width" : 179200,
+    "height" : 179200,
     "log-attrs" : [
       {
         "char-break" : true,
@@ -258,7 +258,7 @@
             "glyphs" : [
               {
                 "glyph" : 268435455,
-                "width" : 51200,
+                "width" : 43008,
                 "is-cluster-start" : true,
                 "log-cluster" : 0
               }
@@ -330,7 +330,7 @@
             "glyphs" : [
               {
                 "glyph" : 268435455,
-                "width" : 51200,
+                "width" : 43008,
                 "is-cluster-start" : true,
                 "log-cluster" : 0
               }
@@ -402,7 +402,7 @@
             "glyphs" : [
               {
                 "glyph" : 268435455,
-                "width" : 51200,
+                "width" : 44032,
                 "is-cluster-start" : true,
                 "log-cluster" : 0
               }
@@ -554,7 +554,7 @@
             "glyphs" : [
               {
                 "glyph" : 268435455,
-                "width" : 51200,
+                "width" : 43008,
                 "is-cluster-start" : true,
                 "log-cluster" : 0
               }
@@ -626,7 +626,7 @@
             "glyphs" : [
               {
                 "glyph" : 268435455,
-                "width" : 51200,
+                "width" : 46080,
                 "is-cluster-start" : true,
                 "log-cluster" : 0
               }
@@ -698,7 +698,7 @@
             "glyphs" : [
               {
                 "glyph" : 268435455,
-                "width" : 51200,
+                "width" : 43008,
                 "is-cluster-start" : true,
                 "log-cluster" : 0
               }
diff --git a/tests/layouts/valid-13.layout b/tests/layouts/valid-13.layout
index bd0f04fa11763fdcd57077c92c5c69fced592960..cb2ac92a2af2e46e0c4277234cbb5f3b46e80275 100644
--- a/tests/layouts/valid-13.layout
+++ b/tests/layouts/valid-13.layout
@@ -53,8 +53,8 @@
     "is-wrapped" : false,
     "is-ellipsized" : false,
     "unknown-glyphs" : 1,
-    "width" : 386048,
-    "height" : 386048,
+    "width" : 315392,
+    "height" : 315392,
     "log-attrs" : [
       {
         "char-break" : true,
@@ -271,7 +271,7 @@
             "glyphs" : [
               {
                 "glyph" : 268435455,
-                "width" : 51200,
+                "width" : 43008,
                 "is-cluster-start" : true,
                 "log-cluster" : 0
               }
@@ -355,7 +355,7 @@
             "glyphs" : [
               {
                 "glyph" : 268435455,
-                "width" : 51200,
+                "width" : 43008,
                 "is-cluster-start" : true,
                 "log-cluster" : 0
               }
@@ -439,7 +439,7 @@
             "glyphs" : [
               {
                 "glyph" : 268435455,
-                "width" : 51200,
+                "width" : 44032,
                 "is-cluster-start" : true,
                 "log-cluster" : 0
               }
@@ -607,7 +607,7 @@
             "glyphs" : [
               {
                 "glyph" : 268435455,
-                "width" : 51200,
+                "width" : 17408,
                 "is-cluster-start" : true,
                 "log-cluster" : 0
               }
@@ -691,7 +691,7 @@
             "glyphs" : [
               {
                 "glyph" : 268435455,
-                "width" : 51200,
+                "width" : 46080,
                 "is-cluster-start" : true,
                 "log-cluster" : 0
               }
@@ -775,7 +775,7 @@
             "glyphs" : [
               {
                 "glyph" : 268435455,
-                "width" : 51200,
+                "width" : 43008,
                 "is-cluster-start" : true,
                 "log-cluster" : 0
               }
