diff -Nuar a/src/qroundprogressbar.cpp b/src/qroundprogressbar.cpp
--- a/src/qroundprogressbar.cpp	2018-03-10 14:17:40.000000000 +0300
+++ b/src/qroundprogressbar.cpp	2021-02-01 11:19:45.693950956 +0300
@@ -20,6 +20,7 @@
 #include "qroundprogressbar.h"
 
 #include <QtGui/QPainter>
+#include <QPainterPath>
 
 
 QRoundProgressBar::QRoundProgressBar(QWidget *parent) :
diff -Nuar a/src/qroundprogressbar.h b/src/qroundprogressbar.h
--- a/src/qroundprogressbar.h	2018-03-10 14:17:40.000000000 +0300
+++ b/src/qroundprogressbar.h	2021-02-01 11:18:48.379955092 +0300
@@ -20,6 +20,7 @@
 #define QROUNDPROGRESSBAR_H
 
 #include <QWidget>
+#include <QPainterPath>
 
 /**
  * @brief The QRoundProgressBar class represents a circular progress bar and maintains its API
