diff -Nuar a/plugins/libzipplugin/libzipplugin.cpp b/plugins/libzipplugin/libzipplugin.cpp
--- a/plugins/libzipplugin/libzipplugin.cpp	2018-04-09 23:42:03.000000000 +0300
+++ b/plugins/libzipplugin/libzipplugin.cpp	2018-04-25 15:30:09.038240448 +0300
@@ -41,6 +41,7 @@
 
 #include <utime.h>
 #include <zlib.h>
+#include <memory>
 
 K_PLUGIN_FACTORY_WITH_JSON(LibZipPluginFactory, "kerfuffle_libzip.json", registerPlugin<LibzipPlugin>();)
 
