diff -Nru vlc-3.0.7.1.orig/modules/lua/vlc.h vlc-3.0.7.1/modules/lua/vlc.h
--- vlc-3.0.7.1.orig/modules/lua/vlc.h	2019-06-28 02:26:12.000000000 +0300
+++ vlc-3.0.7.1/modules/lua/vlc.h	2019-06-28 02:44:42.666003381 +0300
@@ -40,10 +40,12 @@
 
 #define LUA_COMPAT_APIINTCASTS
 #define LUA_COMPAT_MODULE
+extern "C" { 
 #include <lua.h>        /* Low level lua C API */
 #include <lauxlib.h>    /* Higher level C API */
 #include <lualib.h>     /* Lua libs */
-
+}
+	
 #if LUA_VERSION_NUM >= 502
 # define lua_equal(L,idx1,idx2)    lua_compare(L,(idx1),(idx2),LUA_OPEQ)
 # define lua_objlen(L,idx)         lua_rawlen(L,idx)
