Author: Ludovico Cavedon <cavedon@debian.org>
Description: Fix compilation with GCC 4.6
Forwarded: yes

Index: htmlcxx/html/tree.h
===================================================================
--- htmlcxx.orig/html/tree.h
+++ htmlcxx/html/tree.h
@@ -116,7 +116,7 @@ class tree {
             typedef T*                              pointer;
             typedef T&                              reference;
             typedef size_t                          size_type;
-            typedef ptrdiff_t                       difference_type;
+            typedef std::ptrdiff_t                  difference_type;
             typedef std::bidirectional_iterator_tag iterator_category;
 
             iterator_base();
