diff -Naur flang-11.0.0rc2.src/CMakeLists.txt flang-11.0.0rc2.src.cmake/CMakeLists.txt
--- flang-11.0.0rc2.src/CMakeLists.txt	2020-08-20 16:24:59.000000000 +0000
+++ flang-11.0.0rc2.src.cmake/CMakeLists.txt	2020-09-03 13:44:24.226776260 +0000
@@ -199,8 +199,6 @@
   # Always build tco tool
   set(LLVM_BUILD_TOOLS ON)
   message(STATUS "Linking driver with FIR and LLVM")
-  llvm_map_components_to_libnames(LLVM_COMMON_LIBS support)
-  message(STATUS "LLVM libraries: ${LLVM_COMMON_LIBS}")
 endif()
 
 # Add Flang-centric modules to cmake path.
