diff -Naur flang-11.0.0rc2.src.orig/lib/Parser/token-sequence.cpp flang-11.0.0rc2.src/lib/Parser/token-sequence.cpp
--- flang-11.0.0rc2.src.orig/lib/Parser/token-sequence.cpp	2020-09-03 12:01:15.278223312 +0000
+++ flang-11.0.0rc2.src/lib/Parser/token-sequence.cpp	2020-09-03 12:51:40.082870848 +0000
@@ -134,7 +134,8 @@
   std::size_t atToken{0};
   for (std::size_t j{0}; j < chars;) {
     std::size_t nextStart{atToken + 1 < tokens ? start_[++atToken] : chars};
-    char *p{&char_[j]}, *limit{&char_[nextStart]};
+    char *p{&char_[j]};
+    char const* limit{char_.data() + nextStart};
     j = nextStart;
     if (IsDecimalDigit(*p)) {
       while (p < limit && IsDecimalDigit(*p)) {
