SPDX-License-Identifier: GPL-2.0
aufs6.18 standalone patch

diff --git a/fs/dcache.c b/fs/dcache.c
index 93bfb29d0b991..d32683b6cb404 100644
--- a/fs/dcache.c
+++ b/fs/dcache.c
@@ -1387,6 +1387,7 @@ void d_walk(struct dentry *parent, void *data,
 	seq = 1;
 	goto again;
 }
+EXPORT_SYMBOL_GPL(d_walk);
 
 struct check_mount {
 	struct vfsmount *mnt;
diff --git a/fs/exec.c b/fs/exec.c
index 4298e7e08d5d7..26a62cfc04376 100644
--- a/fs/exec.c
+++ b/fs/exec.c
@@ -120,6 +120,7 @@ bool path_noexec(const struct path *path)
 	return (path->mnt->mnt_flags & MNT_NOEXEC) ||
 	       (path->mnt->mnt_sb->s_iflags & SB_I_NOEXEC);
 }
+EXPORT_SYMBOL_GPL(path_noexec);
 
 #ifdef CONFIG_MMU
 /*
diff --git a/fs/fcntl.c b/fs/fcntl.c
index d248048c75540..0d3731f14d9ec 100644
--- a/fs/fcntl.c
+++ b/fs/fcntl.c
@@ -89,6 +89,7 @@ int setfl(int fd, struct file * filp, unsigned int arg)
  out:
 	return error;
 }
+EXPORT_SYMBOL_GPL(setfl);
 
 /*
  * Allocate an file->f_owner struct if it doesn't exist, handling racing
diff --git a/fs/file_table.c b/fs/file_table.c
index cd4a3db4659ac..27cd4c4d5d4bb 100644
--- a/fs/file_table.c
+++ b/fs/file_table.c
@@ -256,6 +256,7 @@ struct file *alloc_empty_file(int flags, const struct cred *cred)
 	}
 	return ERR_PTR(-ENFILE);
 }
+EXPORT_SYMBOL_GPL(alloc_empty_file);
 
 /*
  * Variant of alloc_empty_file() that doesn't check and modify nr_files.
diff --git a/fs/namespace.c b/fs/namespace.c
index 4272349650b14..87a6839796e31 100644
--- a/fs/namespace.c
+++ b/fs/namespace.c
@@ -2268,6 +2268,7 @@ const struct path *collect_paths(const struct path *path,
 		path_get(p);
 	return res;
 }
+EXPORT_SYMBOL_GPL(collect_paths);
 
 void drop_collected_paths(const struct path *paths, const struct path *prealloc)
 {
@@ -2276,6 +2277,7 @@ void drop_collected_paths(const struct path *paths, const struct path *prealloc)
 	if (paths != prealloc)
 		kfree(paths);
 }
+EXPORT_SYMBOL_GPL(drop_collected_paths);
 
 static struct mnt_namespace *alloc_mnt_ns(struct user_namespace *, bool);
 
@@ -6099,6 +6101,7 @@ bool our_mnt(struct vfsmount *mnt)
 {
 	return check_mnt(real_mount(mnt));
 }
+EXPORT_SYMBOL_GPL(our_mnt);
 
 bool current_chrooted(void)
 {
diff --git a/fs/notify/fsnotify.c b/fs/notify/fsnotify.c
index 46bfc543f9467..49be3944ece01 100644
--- a/fs/notify/fsnotify.c
+++ b/fs/notify/fsnotify.c
@@ -225,6 +225,7 @@ int fsnotify_pre_content(const struct path *path, const loff_t *ppos,
 	return fsnotify_parent(path->dentry, FS_PRE_ACCESS, &range,
 			       FSNOTIFY_EVENT_FILE_RANGE);
 }
+EXPORT_SYMBOL_GPL(fsnotify_pre_content);
 
 /*
  * Notify this dentry's parent about a child's events with child name info
diff --git a/fs/notify/group.c b/fs/notify/group.c
index 18446b7b0d495..09138e0b8a6a5 100644
--- a/fs/notify/group.c
+++ b/fs/notify/group.c
@@ -100,6 +100,7 @@ void fsnotify_get_group(struct fsnotify_group *group)
 {
 	refcount_inc(&group->refcnt);
 }
+EXPORT_SYMBOL_GPL(fsnotify_get_group);
 
 /*
  * Drop a reference to a group.  Free it if it's through.
diff --git a/fs/open.c b/fs/open.c
index 3d64372ecc675..f264518f60d7c 100644
--- a/fs/open.c
+++ b/fs/open.c
@@ -69,6 +69,7 @@ int do_truncate(struct mnt_idmap *idmap, struct dentry *dentry,
 	inode_unlock(dentry->d_inode);
 	return ret;
 }
+EXPORT_SYMBOL_GPL(do_truncate);
 
 int vfs_truncate(const struct path *path, loff_t length)
 {
diff --git a/fs/read_write.c b/fs/read_write.c
index 833bae068770a..1e1f8c6349c57 100644
--- a/fs/read_write.c
+++ b/fs/read_write.c
@@ -579,6 +579,7 @@ ssize_t vfs_read(struct file *file, char __user *buf, size_t count, loff_t *pos)
 	inc_syscr(current);
 	return ret;
 }
+EXPORT_SYMBOL_GPL(vfs_read);
 
 static ssize_t new_sync_write(struct file *filp, const char __user *buf, size_t len, loff_t *ppos)
 {
@@ -694,6 +695,7 @@ ssize_t vfs_write(struct file *file, const char __user *buf, size_t count, loff_
 	file_end_write(file);
 	return ret;
 }
+EXPORT_SYMBOL_GPL(vfs_write);
 
 /* file_ppos returns &file->f_pos or NULL if file is stream */
 static inline loff_t *file_ppos(struct file *file)
diff --git a/fs/splice.c b/fs/splice.c
index 19c0ba712acc1..657ad35bdf9ff 100644
--- a/fs/splice.c
+++ b/fs/splice.c
@@ -937,6 +937,7 @@ ssize_t do_splice_from(struct pipe_inode_info *pipe, struct file *out,
 		return warn_unsupported(out, "write");
 	return out->f_op->splice_write(pipe, out, ppos, len, flags);
 }
+EXPORT_SYMBOL_GPL(do_splice_from);
 
 /*
  * Indicate to the caller that there was a premature EOF when reading from the
diff --git a/fs/xattr.c b/fs/xattr.c
index 8851a5ef34f5a..01880c373f068 100644
--- a/fs/xattr.c
+++ b/fs/xattr.c
@@ -405,6 +405,7 @@ vfs_getxattr_alloc(struct mnt_idmap *idmap, struct dentry *dentry,
 	*xattr_value = value;
 	return error;
 }
+EXPORT_SYMBOL_GPL(vfs_getxattr_alloc);
 
 ssize_t
 __vfs_getxattr(struct dentry *dentry, struct inode *inode, const char *name,
diff --git a/kernel/locking/lockdep.c b/kernel/locking/lockdep.c
index b720d966c7b92..02c8ce18d6e0d 100644
--- a/kernel/locking/lockdep.c
+++ b/kernel/locking/lockdep.c
@@ -245,6 +245,7 @@ inline struct lock_class *lockdep_hlock_class(struct held_lock *hlock)
 	 */
 	return lock_classes + class_idx;
 }
+EXPORT_SYMBOL_GPL(lockdep_hlock_class);
 #define hlock_class(hlock) lockdep_hlock_class(hlock)
 
 #ifdef CONFIG_LOCK_STAT
diff --git a/kernel/task_work.c b/kernel/task_work.c
index d1efec571a4a4..eed18cc26c370 100644
--- a/kernel/task_work.c
+++ b/kernel/task_work.c
@@ -230,3 +230,4 @@ void task_work_run(void)
 		} while (work);
 	}
 }
+EXPORT_SYMBOL_GPL(task_work_run);
diff --git a/security/security.c b/security/security.c
index 4d3c03a4524c5..29540ce958dad 100644
--- a/security/security.c
+++ b/security/security.c
@@ -1952,6 +1952,7 @@ void security_path_post_mknod(struct mnt_idmap *idmap, struct dentry *dentry)
 		return;
 	call_void_hook(path_post_mknod, idmap, dentry);
 }
+EXPORT_SYMBOL_GPL(security_path_post_mknod);
 
 /**
  * security_path_mkdir() - Check if creating a new directory is allowed
@@ -1987,6 +1988,7 @@ int security_path_rmdir(const struct path *dir, struct dentry *dentry)
 		return 0;
 	return call_int_hook(path_rmdir, dir, dentry);
 }
+EXPORT_SYMBOL_GPL(security_path_rmdir);
 
 /**
  * security_path_unlink() - Check if removing a hard link is allowed
@@ -2022,6 +2024,7 @@ int security_path_symlink(const struct path *dir, struct dentry *dentry,
 		return 0;
 	return call_int_hook(path_symlink, dir, dentry, old_name);
 }
+EXPORT_SYMBOL_GPL(security_path_symlink);
 
 /**
  * security_path_link - Check if creating a hard link is allowed
@@ -2040,6 +2043,7 @@ int security_path_link(struct dentry *old_dentry, const struct path *new_dir,
 		return 0;
 	return call_int_hook(path_link, old_dentry, new_dir, new_dentry);
 }
+EXPORT_SYMBOL_GPL(security_path_link);
 
 /**
  * security_path_rename() - Check if renaming a file is allowed
@@ -2101,6 +2105,7 @@ int security_path_chmod(const struct path *path, umode_t mode)
 		return 0;
 	return call_int_hook(path_chmod, path, mode);
 }
+EXPORT_SYMBOL_GPL(security_path_chmod);
 
 /**
  * security_path_chown() - Check if changing the file's owner/group is allowed
@@ -2118,6 +2123,7 @@ int security_path_chown(const struct path *path, kuid_t uid, kgid_t gid)
 		return 0;
 	return call_int_hook(path_chown, path, uid, gid);
 }
+EXPORT_SYMBOL_GPL(security_path_chown);
 
 /**
  * security_path_chroot() - Check if changing the root directory is allowed
@@ -2362,6 +2368,7 @@ int security_inode_permission(struct inode *inode, int mask)
 		return 0;
 	return call_int_hook(inode_permission, inode, mask);
 }
+EXPORT_SYMBOL_GPL(security_inode_permission);
 
 /**
  * security_inode_setattr() - Check if setting file attributes is allowed
@@ -2912,6 +2919,7 @@ int security_file_permission(struct file *file, int mask)
 {
 	return call_int_hook(file_permission, file, mask);
 }
+EXPORT_SYMBOL_GPL(security_file_permission);
 
 /**
  * security_file_alloc() - Allocate and init a file's LSM blob
@@ -3214,6 +3222,7 @@ int security_file_truncate(struct file *file)
 {
 	return call_int_hook(file_truncate, file);
 }
+EXPORT_SYMBOL_GPL(security_file_truncate);
 
 /**
  * security_task_alloc() - Allocate a task's LSM blob
