SPDX-License-Identifier: GPL-2.0
aufs6.12 kbuild patch

diff --git a/fs/Kconfig b/fs/Kconfig
index aae170fc27952..d5bb584649f50 100644
--- a/fs/Kconfig
+++ b/fs/Kconfig
@@ -341,6 +341,7 @@ source "fs/sysv/Kconfig"
 source "fs/ufs/Kconfig"
 source "fs/erofs/Kconfig"
 source "fs/vboxsf/Kconfig"
+source "fs/aufs/Kconfig"
 
 endif # MISC_FILESYSTEMS
 
diff --git a/fs/Makefile b/fs/Makefile
index 61679fd587b7e..b2539712b80d0 100644
--- a/fs/Makefile
+++ b/fs/Makefile
@@ -130,3 +130,4 @@ obj-$(CONFIG_EROFS_FS)		+= erofs/
 obj-$(CONFIG_VBOXSF_FS)		+= vboxsf/
 obj-$(CONFIG_ZONEFS_FS)		+= zonefs/
 obj-$(CONFIG_BPF_LSM)		+= bpf_fs_kfuncs.o
+obj-$(CONFIG_AUFS_FS)           += aufs/
