diff -Nuar a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2017-12-08 04:01:55.000000000 +0300
+++ b/CMakeLists.txt	2017-12-16 22:11:14.110618003 +0300
@@ -26,7 +26,7 @@
 set(CALENDARCORE_LIB_VERSION "5.7.0")
 set(CALENDARUTILS_LIB_VERSION "5.7.0")
 set(kcontacts_LIB_VERSION "5.7.0")
-set(QT_REQUIRED_VERSION "6.6.3")
+set(QT_REQUIRED_VERSION "6.6.3")
 ecm_setup_version(PROJECT VARIABLE_PREFIX ktnef
                         VERSION_HEADER "${CMAKE_CURRENT_BINARY_DIR}/ktnef_version.h"
                         PACKAGE_VERSION_FILE "${CMAKE_CURRENT_BINARY_DIR}/KF5TnefConfigVersion.cmake"
