diff --git a/src/basketdialog.py b/src/basketdialog.py
index c929f08..3ce2cca 100755
--- a/src/basketdialog.py
+++ b/src/basketdialog.py
@@ -70,8 +70,8 @@ class BasketDialog(PAbstractBox, Ui_BasketDialog):
         try:
             self.filterExtras()
         except Exception, e:
-            messageBox = QtWidgets.QMessageBox(_translate("Package Manager","Pisi Error"), unicode(e),
-                    QtWidgets.QMessageBox.Critical, QtWidgets.QMessageBox.Ok, 0, 0)
+            messageBox = QtWidgets.QMessageBox(QMessageBox.Critical, _translate("Package Manager","Pisi Error"), unicode(e),                  
+                         QtWidgets.QMessageBox.Ok)
             QTimer.singleShot(0, restoreCursor)
             messageBox.exec_()
             return
