diff -rupN cairo-dock-plug-ins-3.4.1.orig/CMakeLists.txt cairo-dock-plug-ins-3.4.1/CMakeLists.txt
--- cairo-dock-plug-ins-3.4.1.orig/CMakeLists.txt	2015-03-12 09:43:57.171067365 +0100
+++ cairo-dock-plug-ins-3.4.1/CMakeLists.txt	2015-03-12 09:46:20.717120675 +0100
@@ -573,11 +573,11 @@ if (enable-mono-interface)
 	if (NOT GMCS_EXECUTABLE OR NOT EXISTS ${GMCS_EXECUTABLE})
 		message (STATUS "Could not find Mono compiler gmcs, won't build Mono interface.")
 	else()
-		pkg_check_modules (MONO_PACKAGE glib-sharp-2.0 ndesk-dbus-1.0 ndesk-dbus-glib-1.0)
+		pkg_check_modules (MONO_PACKAGE dbus-sharp-2.0 dbus-sharp-glib-2.0 glib-sharp-3.0)
 		if (NOT MONO_PACKAGE_FOUND)
-			message (STATUS "Could not find glib-sharp-2.0, ndesk-dbus-1.0 or ndesk-dbus-glib-1.0; won't be built Mono interface.")
-			message (WARNING "These modules are required to compile DBus applet with Mono interface: glib-sharp-2.0, ndesk-dbus-1.0 and ndesk-dbus-glib-1.0")
-			set (MODULES_MISSING "${MODULES_MISSING} glib-sharp-2.0 ndesk-dbus-1.0 ndesk-dbus-glib-1.0")
+			message (STATUS "Could not find dbus-sharp-2.0, dbus-sharp-glib-2.0 or glib-sharp-3.0; won't be built Mono interface.")
+			message (WARNING "These modules are required to compile DBus applet with Mono interface: dbus-sharp-2.0, dbus-sharp-glib-2.0 and glib-sharp-3.0")
+			set (MODULES_MISSING "${MODULES_MISSING} dbus-sharp-2.0 dbus-sharp-glib-2.0 glib-sharp-3.0")
 		else()
 			set (MONO_FOUND TRUE)
 			set (with_mono yes)

diff -rupN cairo-dock-plug-ins-3.4.1.orig/Dbus/interfaces/mono/CMakeLists.txt cairo-dock-plug-ins-3.4.1/Dbus/interfaces/mono/CMakeLists.txt
--- cairo-dock-plug-ins-3.4.1.orig/Dbus/interfaces/mono/CMakeLists.txt	2015-03-12 09:52:56.491934965 +0100
+++ cairo-dock-plug-ins-3.4.1/Dbus/interfaces/mono/CMakeLists.txt	2015-03-12 09:54:07.418777441 +0100
@@ -3,7 +3,7 @@

 execute_process(COMMAND ${GMCS_EXECUTABLE}
 	-target:library
-	-pkg:glib-sharp-2.0 -pkg:ndesk-dbus-1.0 -pkg:ndesk-dbus-glib-1.0
+	-pkg:dbus-sharp-2.0 -pkg:dbus-sharp-glib-2.0 -pkg:glib-sharp-3.0
 	-out:${CMAKE_CURRENT_BINARY_DIR}/CDApplet.dll
 	${CMAKE_CURRENT_SOURCE_DIR}/CDApplet.cs ${CMAKE_CURRENT_SOURCE_DIR}/ISubApplet.cs ${CMAKE_CURRENT_SOURCE_DIR}/IApplet.cs)
 ### find how to register to GAC ...
