From d084645fdc9bf805b39b0fb3ff7ec315661884be Mon Sep 17 00:00:00 2001
From: Gary Lin <glin@suse.com>
Date: Wed, 9 Mar 2016 12:19:26 +0800
Subject: [PATCH] Remove efivar-static since it requires static libs

Signed-off-by: Gary Lin <glin@suse.com>
---
 src/Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/Makefile b/src/Makefile
index 5fc7887..e1c1ccd 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -8,7 +8,7 @@ include $(TOPDIR)/Make.defaults
 
 LIBTARGETS=libefivar.so libefiboot.so
 STATICLIBTARGETS=libefivar.a libefiboot.a
-BINTARGETS=efivar efivar-static
+BINTARGETS=efivar
 PCTARGETS=efivar.pc efiboot.pc
 TARGETS=$(LIBTARGETS) $(STATICLIBTARGETS) $(BINTARGETS) $(PCTARGETS)
 
-- 
2.7.2
