diff -Nrup a/src/enginio_client/enginioclient_p.h b/src/enginio_client/enginioclient_p.h
--- a/src/enginio_client/enginioclient_p.h	2016-11-28 16:36:23.000000000 +0100
+++ b/src/enginio_client/enginioclient_p.h	2020-01-30 09:16:08.088599338 +0100
@@ -58,6 +58,7 @@
 #include <QtCore/qset.h>
 #include <QtCore/qlogging.h>
 #include <QtCore/qdebug.h>
+#include <QLinkedList>
 
 #include <QtCore/private/qobject_p.h>
 
