Index: gypsy-0.9/configure.ac
===================================================================
--- gypsy-0.9/configure.ac
+++ gypsy-0.9/configure.ac	2016-01-07 01:12:43.372398664 +0000
@@ -58,7 +58,7 @@
 AS_AC_EXPAND(LIBEXECDIR, $libexecdir)
 AS_AC_EXPAND(DAEMON_USER, $with_user)
 
-CFLAGS="$CFLAGS -g -Wall -Werror -Wno-format"
+CFLAGS="$CFLAGS -g -Wall -Wno-format"
 
 GTK_DOC_CHECK([1.6])
 # TODO: add --with-xsltproc argument to override PATH
