diff -up focusblur-3.2.6/src/focusblurparam.h.me focusblur-3.2.6/src/focusblurparam.h
--- focusblur-3.2.6/src/focusblurparam.h.me	2020-03-05 12:27:23.815302199 +0100
+++ focusblur-3.2.6/src/focusblurparam.h	2020-03-05 12:27:58.880383839 +0100
@@ -81,7 +81,7 @@ struct _FblurParam
 
 /*---- Variable ----*/
 
-const FblurStoreParam fblur_init_param;
+extern const FblurStoreParam fblur_init_param;
 
 
 /*---- Functions ----*/
diff -up focusblur-3.2.6/src/focusblurrc.h.me focusblur-3.2.6/src/focusblurrc.h
--- focusblur-3.2.6/src/focusblurrc.h.me	2020-03-05 12:28:17.097529345 +0100
+++ focusblur-3.2.6/src/focusblurrc.h	2020-03-05 12:28:36.424697791 +0100
@@ -40,7 +40,7 @@ struct _FblurPreferences
 
 /*---- Variables ----*/
 
-const FblurPreferences fblur_init_preferences;
+extern const FblurPreferences fblur_init_preferences;
 
 
 /*---- Functions ----*/
