diff -uNr focusblur-3.2.6.orig/libgimp/gettext.h focusblur-3.2.6/libgimp/gettext.h
--- focusblur-3.2.6.orig/libgimp/gettext.h	2007-04-19 14:06:49.000000000 +0200
+++ focusblur-3.2.6/libgimp/gettext.h	2016-08-13 17:45:13.116158292 +0200
@@ -3,7 +3,7 @@
 
 #ifdef ENABLE_NLS
 
-#include <glib/gi18n.h>
+#include <glib.h>
 
 #define INIT_I18N()     G_STMT_START{                           \
     bindtextdomain (GETTEXT_PACKAGE, gimp_locale_directory ()); \
diff -uNr focusblur-3.2.6.orig/src/aaa.h focusblur-3.2.6/src/aaa.h
--- focusblur-3.2.6.orig/src/aaa.h	2008-10-13 17:13:10.000000000 +0200
+++ focusblur-3.2.6/src/aaa.h	2016-08-13 17:43:13.182532655 +0200
@@ -19,8 +19,7 @@
 #ifndef __AAA_H__
 #define __AAA_H__
 
-#include <glib/gmacros.h>
-#include <glib/gtypes.h>
+#include <glib.h>
 
 
 G_BEGIN_DECLS
diff -uNr focusblur-3.2.6.orig/src/brush.h focusblur-3.2.6/src/brush.h
--- focusblur-3.2.6.orig/src/brush.h	2009-01-23 14:06:54.000000000 +0100
+++ focusblur-3.2.6/src/brush.h	2016-08-13 17:44:26.891760912 +0200
@@ -22,7 +22,7 @@
 #ifndef __FOCUSBLUR_BRUSH_H__
 #define __FOCUSBLUR_BRUSH_H__
 
-#include <glib/gtypes.h>
+#include <glib.h>
 #include "focusblurtypes.h"
 
 G_BEGIN_DECLS
diff -uNr focusblur-3.2.6.orig/src/depthmap.h focusblur-3.2.6/src/depthmap.h
--- focusblur-3.2.6.orig/src/depthmap.h	2009-01-23 14:06:54.000000000 +0100
+++ focusblur-3.2.6/src/depthmap.h	2016-08-13 17:44:53.749572833 +0200
@@ -22,7 +22,7 @@
 #ifndef __FOCUSBLUR_DEPTHMAP_H__
 #define __FOCUSBLUR_DEPTHMAP_H__
 
-#include <glib/gtypes.h>
+#include <glib.h>
 
 #include "focusblurtypes.h"
 #include "focusblurenums.h"
diff -uNr focusblur-3.2.6.orig/src/diffusion.h focusblur-3.2.6/src/diffusion.h
--- focusblur-3.2.6.orig/src/diffusion.h	2009-01-23 14:06:54.000000000 +0100
+++ focusblur-3.2.6/src/diffusion.h	2016-08-13 17:44:15.291410229 +0200
@@ -23,7 +23,7 @@
 #define __FOCUSBLUR_DIFFUSION_H__
 
 
-#include <glib/gtypes.h>
+#include <glib.h>
 
 #include "focusblur.h"
 #include "focusblurtypes.h"
diff -uNr focusblur-3.2.6.orig/src/fftblurbuffer.h focusblur-3.2.6/src/fftblurbuffer.h
--- focusblur-3.2.6.orig/src/fftblurbuffer.h	2009-01-23 14:06:54.000000000 +0100
+++ focusblur-3.2.6/src/fftblurbuffer.h	2016-08-13 17:44:42.194223511 +0200
@@ -28,8 +28,7 @@
 #endif
 #include <fftw3.h>
 
-#include <glib/gmacros.h>
-#include <glib/gtypes.h>
+#include <glib.h>
 #include <gtk/gtkstyle.h>
 #include <libgimp/gimptypes.h>
 #include <libgimpwidgets/gimpwidgetstypes.h>
diff -uNr focusblur-3.2.6.orig/src/fftblur.h focusblur-3.2.6/src/fftblur.h
--- focusblur-3.2.6.orig/src/fftblur.h	2009-01-23 14:06:54.000000000 +0100
+++ focusblur-3.2.6/src/fftblur.h	2016-08-13 17:44:58.889728222 +0200
@@ -23,8 +23,7 @@
 #define __FOCUSBLUR_FFTBLUR_H__
 
 
-#include <glib/gmacros.h>
-#include <glib/gtypes.h>
+#include <glib.h>
 #include <libgimpwidgets/gimpwidgetstypes.h>
 
 #include "focusblurparam.h"
diff -uNr focusblur-3.2.6.orig/src/fftblurproc.h focusblur-3.2.6/src/fftblurproc.h
--- focusblur-3.2.6.orig/src/fftblurproc.h	2009-01-23 14:06:54.000000000 +0100
+++ focusblur-3.2.6/src/fftblurproc.h	2016-08-13 17:44:49.022429930 +0200
@@ -23,8 +23,7 @@
 #define __FOCUSBLUR_FFTBLUR_PROC_H__
 
 
-#include <glib/gmacros.h>
-#include <glib/gtypes.h>
+#include <glib.h>
 
 #include "focusblurtypes.h"
 
diff -uNr focusblur-3.2.6.orig/src/focusblur.h focusblur-3.2.6/src/focusblur.h
--- focusblur-3.2.6.orig/src/focusblur.h	2009-01-23 14:06:54.000000000 +0100
+++ focusblur-3.2.6/src/focusblur.h	2016-08-13 17:44:38.196102646 +0200
@@ -22,7 +22,7 @@
 #ifndef __FOCUSBLUR_H__
 #define __FOCUSBLUR_H__
 
-#include <glib/gmacros.h>
+#include <glib.h>
 
 G_BEGIN_DECLS
 
diff -uNr focusblur-3.2.6.orig/src/focusblurparam.h focusblur-3.2.6/src/focusblurparam.h
--- focusblur-3.2.6.orig/src/focusblurparam.h	2009-01-23 14:06:54.000000000 +0100
+++ focusblur-3.2.6/src/focusblurparam.h	2016-08-13 17:43:41.657393461 +0200
@@ -22,8 +22,7 @@
 #ifndef __FOCUSBLUR_PARAM_H__
 #define __FOCUSBLUR_PARAM_H__
 
-#include <glib/gmacros.h>
-#include <glib/gtypes.h>
+#include <glib.h>
 #include <gtk/gtkstyle.h>
 #include <libgimp/gimptypes.h>
 
diff -uNr focusblur-3.2.6.orig/src/focusblurrc.c focusblur-3.2.6/src/focusblurrc.c
--- focusblur-3.2.6.orig/src/focusblurrc.c	2009-07-31 03:54:51.000000000 +0200
+++ focusblur-3.2.6/src/focusblurrc.c	2016-08-13 17:44:31.571902394 +0200
@@ -24,8 +24,7 @@
 #include <fcntl.h>
 #include <sys/stat.h>
 #include <unistd.h>
-#include <glib/gstdio.h>
-#include <glib/gprintf.h>
+#include <glib.h>
 #include <libgimp/gimp.h>
 
 #include "libgimp/stdplugins-intl.h"
diff -uNr focusblur-3.2.6.orig/src/focusblurstock.h focusblur-3.2.6/src/focusblurstock.h
--- focusblur-3.2.6.orig/src/focusblurstock.h	2009-01-23 14:06:54.000000000 +0100
+++ focusblur-3.2.6/src/focusblurstock.h	2016-08-13 17:43:19.791732454 +0200
@@ -22,7 +22,7 @@
 #ifndef __FOCUSBLUR_STOCK_H__
 #define __FOCUSBLUR_STOCK_H__
 
-#include <glib/gtypes.h>
+#include <glib.h>
 
 G_BEGIN_DECLS
 
diff -uNr focusblur-3.2.6.orig/src/focusblurtypes.h focusblur-3.2.6/src/focusblurtypes.h
--- focusblur-3.2.6.orig/src/focusblurtypes.h	2009-01-23 14:06:54.000000000 +0100
+++ focusblur-3.2.6/src/focusblurtypes.h	2016-08-13 17:44:12.275319051 +0200
@@ -22,7 +22,7 @@
 #ifndef __FOCUSBLUR_TYPES_H__
 #define __FOCUSBLUR_TYPES_H__
 
-#include <glib/gmacros.h>
+#include <glib.h>
 
 
 G_BEGIN_DECLS
diff -uNr focusblur-3.2.6.orig/src/interface.h focusblur-3.2.6/src/interface.h
--- focusblur-3.2.6.orig/src/interface.h	2009-01-23 14:06:54.000000000 +0100
+++ focusblur-3.2.6/src/interface.h	2016-08-13 17:44:34.526991727 +0200
@@ -22,7 +22,7 @@
 #ifndef __FOCUSBLUR_INTERFACE_H__
 #define __FOCUSBLUR_INTERFACE_H__
 
-#include <glib/gtypes.h>
+#include <glib.h>
 
 #include "focusblurtypes.h"
 
diff -uNr focusblur-3.2.6.orig/src/render.h focusblur-3.2.6/src/render.h
--- focusblur-3.2.6.orig/src/render.h	2009-01-23 14:06:54.000000000 +0100
+++ focusblur-3.2.6/src/render.h	2016-08-13 17:43:36.221229123 +0200
@@ -24,7 +24,7 @@
 
 #include "config.h"
 
-#include <glib/gtypes.h>
+#include <glib.h>
 //#include <libgimp/gimp.h>
 #include <libgimp/gimpui.h>
 
diff -uNr focusblur-3.2.6.orig/src/shine.h focusblur-3.2.6/src/shine.h
--- focusblur-3.2.6.orig/src/shine.h	2009-01-23 14:06:54.000000000 +0100
+++ focusblur-3.2.6/src/shine.h	2016-08-13 17:44:20.250560146 +0200
@@ -22,7 +22,7 @@
 #ifndef __FOCUSBLUR_SHINE_H__
 #define __FOCUSBLUR_SHINE_H__
 
-#include <glib/gtypes.h>
+#include <glib.h>
 #include <libgimp/gimptypes.h>
 
 #include "focusblurtypes.h"
diff -uNr focusblur-3.2.6.orig/src/source.h focusblur-3.2.6/src/source.h
--- focusblur-3.2.6.orig/src/source.h	2009-01-23 14:06:54.000000000 +0100
+++ focusblur-3.2.6/src/source.h	2016-08-13 17:44:24.011673846 +0200
@@ -24,7 +24,7 @@
 
 #include "config.h"
 
-#include <glib/gtypes.h>
+#include <glib.h>
 #include <libgimp/gimptypes.h>
 
 #include "focusblurtypes.h"
