diff -uNr focusblur-3.2.6.orig/libgimp/gettext.h focusblur-3.2.6/libgimp/gettext.h
--- focusblur-3.2.6.orig/libgimp/gettext.h	2007-04-19 14:06:49.000000000 +0200
+++ focusblur-3.2.6/libgimp/gettext.h	2016-08-13 17:55:28.468760114 +0200
@@ -5,6 +5,10 @@
 
 #include <glib.h>
 
+#include <libintl.h>
+#define _(String) gettext (String)
+#define gettext_noop(String) String
+#define N_(String) gettext_noop (String)
 #define INIT_I18N()     G_STMT_START{                           \
     bindtextdomain (GETTEXT_PACKAGE, gimp_locale_directory ()); \
     bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");         \
Binary files focusblur-3.2.6.orig/libgimp/.gettext.h.swp and focusblur-3.2.6/libgimp/.gettext.h.swp differ
