diff -Nuar a/data/meson.build b/data/meson.build
--- a/data/meson.build	2021-12-27 16:02:54.000000000 +0300
+++ b/data/meson.build	2023-04-13 15:50:47.302489933 +0300
@@ -6,7 +6,6 @@
 desktop_config.set('icon', application_id)
 desktop_config.set('binary', gitg_name)
 i18n.merge_file(
-  desktop,
   type: 'desktop',
   input: configure_file(
     input: desktop + '.in.in',
@@ -42,7 +41,6 @@
 appdata_config.set('app-id', application_id)
 appdata_config.set('gettext', gitg_name)
 i18n.merge_file(
-  appdata,
   type: 'xml',
   input: configure_file(
     input: appdata + '.in.in',
