diff -Nuar a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2017-12-08 04:01:31.000000000 +0300
+++ b/CMakeLists.txt	2017-12-16 13:45:06.592702705 +0300
@@ -32,7 +32,7 @@
 )
 
 ########### Find packages ###########
-set(QT_REQUIRED_VERSION 5.8.0)
+set(QT_REQUIRED_VERSION 5.7.1)
 find_package(Qt5 ${QT_REQUIRED_VERSION} CONFIG REQUIRED Core Qml)
 
 remove_definitions(-DQT_NO_CAST_FROM_ASCII -DQT_NO_CAST_TO_ASCII)
