diff -Nuar a/libpisiyap/templates/pspec.py b/libpisiyap/templates/pspec.py
--- a/libpisiyap/templates/pspec.py	2016-05-16 15:49:45.000000000 +0300
+++ b/libpisiyap/templates/pspec.py	2018-08-19 18:11:18.639496171 +0300
@@ -77,7 +77,32 @@
         </Provides>
         -->
     </Package>
-
+    <!--
+    <Package>
+        <Name>%(packageName)s-devel</Name>
+        <Summary>Development files for %(packageName)s</Summary>
+        <RuntimeDependencies>
+            <Dependency release="current">%(packageName)s</Dependency>
+        </RuntimeDependencies>
+        <Files>
+            <Path fileType="header">/usr/include</Path>
+            <Path fileType="data">/usr/lib/cmake</Path>
+            <Path fileType="library">/usr/lib/pkgconfig</Path>
+        </Files>
+    </Package>
+    -->
+    <!--
+    <Package>
+        <Name>%(packageName)s-docs</Name>
+        <Summary>Development files for %(packageName)s</Summary>
+        <RuntimeDependencies>
+            <Dependency release="current">%(packageName)s</Dependency>
+        </RuntimeDependencies>
+        <Files>
+            <Path fileType="data">/usr/share/gtk-doc</Path>
+        </Files>
+    </Package>
+    -->
     <History>
         <Update release="1">
             <Date>%(date)s</Date>
