--- a/src/wm/20_solus-project.budgie.wm.gschema.override
+++ b/src/wm/20_solus-project.budgie.wm.gschema.override
@@ -12,5 +12,5 @@
 picture-uri = 'file:///usr/share/backgrounds/budgie/default.jpg'

 [org.gnome.desktop.wm.keybindings:Budgie]
-switch-input-source = '[\'<Alt>Shift_L\', \'<Super>space\', \'XF86Keyboard\']'
-switch-input-source-backward = '[\'<Super><Shift>space\', \'<Shift>XF86Keyboard\']'
+switch-input-source = ['<Alt>Shift_L', '<Super>space', 'XF86Keyboard']
+switch-input-source-backward = ['<Super><Shift>space', '<Shift>XF86Keyboard']
