--- lept.pc.in  2016-01-29 21:41:01.000000000 +0200
+++ lept        2018-01-05 14:46:02.453185333 +0300
@@ -1,12 +1,9 @@
-prefix=@prefix@
-exec_prefix=@exec_prefix@
-libdir=@libdir@
-includedir=@includedir@/leptonica
-
-Name: leptonica
-Description: An open source C library for efficient image processing and image analysis operations
-Version: @VERSION@
-Libs: -L${libdir} -llept
-Libs.private: @ZLIB_LIBS@ @LIBPNG_LIBS@ @JPEG_LIBS@ @GIFLIB_LIBS@ @LIBTIFF_LIBS@ @LIBWEBP_LIBS@
-Cflags: -I${includedir}
-
+libdir=@libdir@
+includedir=@includedir@/leptonica
+
+Name: leptonica
+Description: An open source C library for efficient image processing and image analysis operations
+Version: @VERSION@
+Libs: -L${libdir} -llept
+Libs.private: @ZLIB_LIBS@ @LIBPNG_LIBS@ @JPEG_LIBS@ @GIFLIB_LIBS@ @LIBTIFF_LIBS@ @LIBWEBP_LIBS@
+Cflags: -I${includedir}
