From 018bcb5824a956409baef6cf68a8469a6c0e8c1b Mon Sep 17 00:00:00 2001
From: Bahar Kurt <kurtbahartr@users.noreply.github.com>
Date: Tue, 22 Oct 2024 12:10:52 +0300
Subject: [PATCH] Change captive portal detection URL to something known more
 widely

---
 src/MainWindow.vala | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/MainWindow.vala b/src/MainWindow.vala
index a125814..c410eeb 100644
--- a/src/MainWindow.vala
+++ b/src/MainWindow.vala
@@ -19,7 +19,9 @@
 */
 
 public class Captive.MainWindow : Gtk.ApplicationWindow {
-    private const string DUMMY_URL = "http://capnet.elementary.io";
+    // This was elementaryOS' own capnet server, which isn't acknowledged by networks like universities.
+    // As a quick workaround, we can use a more widely acknowledged server, like Google's.
+    private const string DUMMY_URL = "http://gstatic.com/generate_204";
 
     private Adw.TabView tabview;
     private Granite.HeaderLabel cert_subject;
-- 
2.46.2

