--- a/src/config.c	2015-03-03 01:54:36.000000000 +0300
+++ b/src/config.c	2022-05-02 08:55:22.210600991 +0300
@@ -79,16 +79,16 @@
 
     // Notifications
     .show_notification = TRUE,
-    .notification_type = 0,
+    .notification_type = 1,
 
     // Status icon
-    .stepsize = 0,
+    .stepsize = 5,
     .helper_program = NULL,
     .theme = NULL,
     .use_panel_specific_icons = FALSE,
 
     // Left mouse button action
-    .lmb_slider = FALSE,
+    .lmb_slider = TRUE,
 
     // Middle mouse button action
     .mmb_mute = FALSE,
@@ -99,9 +99,9 @@
     .use_transparent_background = FALSE,
 
     // Hotkeys
-    .hotkey_up_enabled = FALSE,
-    .hotkey_down_enabled = FALSE,
-    .hotkey_mute_enabled = FALSE,
+    .hotkey_up_enabled = TRUE,
+    .hotkey_down_enabled = TRUE,
+    .hotkey_mute_enabled = TRUE,
     .hotkey_up = NULL,
     .hotkey_down = NULL,
     .hotkey_mute = NULL
