diff -Nuar a/kaptan/kaptan.py b/kaptan/kaptan.py
--- a/kaptan/kaptan.py	2021-04-03 21:29:51.000000000 +0300
+++ b/kaptan/kaptan.py	2023-10-31 11:29:42.336651225 +0300
@@ -26,14 +26,15 @@
 class Kaptan(QtWidgets.QWizard):
     def __init__(self):
         super().__init__()
+        self.setFixedSize(850, 600)
 
         self.setWindowTitle(self.tr("Kaptan"))
         self.setWindowIcon(QIcon.fromTheme("kaptan-icon"))
         self.setMinimumSize(850, 600)
         self.setMaximumSize(950, 620)
-        x = (QtWidgets.QDesktopWidget().size().width() - self.width()) / 2
-        y = (QtWidgets.QDesktopWidget().size().height() - self.height()) / 2
-        self.move(x, y)
+        # x = (QtWidgets.QDesktopWidget().size().width() - self.width()) / 2
+        # y = (QtWidgets.QDesktopWidget().size().height() - self.height()) / 2
+        # self.move(x, y)
         self.setPixmap(QtWidgets.QWizard.LogoPixmap, QPixmap("/usr/share/kaptan/images/kaptan.png"))
 
         self.setButtonText(QtWidgets.QWizard.NextButton, self.tr("Next"))
