diff -uprN libcryptui-3.12.2.orig/configure.ac libcryptui-3.12.2/configure.ac
--- libcryptui-3.12.2.orig/configure.ac	2021-07-27 19:35:29.668962000 +0200
+++ configure.ac	2021-07-27 19:36:21.194924363 +0200
@@ -96,7 +96,7 @@ AC_ARG_ENABLE(gpg-check,
 
 if test	"$DO_CHECK" = "yes"; then
 	accepted_versions="1.2 1.4 2.0 2.1 2.2"
-	AC_PATH_PROGS(GNUPG, [gpg gpg2], no)
+	AC_PATH_PROGS(GNUPG, [gpg1 gpg2], no)
 	AC_DEFINE_UNQUOTED(GNUPG, "$GNUPG", [Path to gpg executable.])
 	ok="no"
 	if test "$GNUPG" != "no"; then
