diff --git a/share/cmake/modules/FindExtPackages.cmake b/share/cmake/modules/FindExtPackages.cmake
index 8787b872a..f37340a08 100644
--- a/share/cmake/modules/FindExtPackages.cmake
+++ b/share/cmake/modules/FindExtPackages.cmake
@@ -30,7 +30,7 @@ find_package(expat 2.2.8 REQUIRED)

 # yaml-cpp
 # https://github.com/jbeder/yaml-cpp
-find_package(yaml-cpp 0.7.0 REQUIRED)
+find_package(yaml-cpp 0.8.0 REQUIRED)

 # pystring
 # https://github.com/imageworks/pystring
diff --git a/share/cmake/modules/Findyaml-cpp.cmake b/share/cmake/modules/Findyaml-cpp.cmake
index 59558ce6..c5e051ac 100644
--- a/share/cmake/modules/Findyaml-cpp.cmake
+++ b/share/cmake/modules/Findyaml-cpp.cmake
@@ -42,10 +42,6 @@ if(NOT OCIO_INSTALL_EXT_PACKAGES STREQUAL ALL)
         find_package(yaml-cpp ${yaml-cpp_FIND_VERSION} CONFIG QUIET)
     endif()

-    if(yaml-cpp_FOUND)
-        get_target_property(yaml-cpp_LIBRARY yaml-cpp LOCATION)
-    else()
-
         # As yaml-cpp-config.cmake search fails, search an installed library
         # using yaml-cpp.pc .

@@ -111,7 +107,6 @@ if(NOT OCIO_INSTALL_EXT_PACKAGES STREQUAL ALL)
         if(PC_yaml-cpp_FOUND)
             set(yaml-cpp_VERSION "${PC_yaml-cpp_VERSION}")
         endif()
-    endif()

     # Override REQUIRED if package can be installed
     if(OCIO_INSTALL_EXT_PACKAGES STREQUAL MISSING)
