From 9af9d0f69eda99cc034743adf789a0005fbb3f1b Mon Sep 17 00:00:00 2001
From: Mike Frysinger <vapier@gentoo.org>
Date: Fri, 3 Jan 2020 17:20:06 +0100
Subject: [PATCH 11/14] Log connect errors to /var/log/ppp-connect-errors

---
 pppd/pathnames.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/pppd/pathnames.h b/pppd/pathnames.h
index 7647d49..2d7c621 100644
--- a/pppd/pathnames.h
+++ b/pppd/pathnames.h
@@ -34,7 +34,7 @@
 #define _PATH_AUTHDOWN	 _ROOT_PATH "/etc/ppp/auth-down"
 #define _PATH_AUTHFAIL	 _ROOT_PATH "/etc/ppp/auth-fail"
 #define _PATH_TTYOPT	 _ROOT_PATH "/etc/ppp/options."
-#define _PATH_CONNERRS	 _ROOT_PATH "/etc/ppp/connect-errors"
+#define _PATH_CONNERRS	 _ROOT_PATH "/var/log/ppp-connect-errors"
 #define _PATH_PEERFILES	 _ROOT_PATH "/etc/ppp/peers/"
 #define _PATH_RESOLV	 _ROOT_PATH "/etc/ppp/resolv.conf"
 
-- 
2.30.0

