diff -Nru libva-vdpau-driver-0.7.4.orig/src/vdpau_video.c libva-vdpau-driver-0.7.4/src/vdpau_video.c
--- libva-vdpau-driver-0.7.4.orig/src/vdpau_video.c	2012-10-05 18:02:58.000000000 +0300
+++ libva-vdpau-driver-0.7.4/src/vdpau_video.c	2020-01-14 14:52:16.835039658 +0300
@@ -927,8 +927,10 @@
 vdpau_CreateSurfaceFromV4L2Buf(
     VADriverContextP    ctx,
     int                 v4l2_fd,
+	/*
     struct v4l2_format *v4l2_fmt,
     struct v4l2_buffer *v4l2_buf,
+	*/
     VASurfaceID        *surface
 )
 {
diff -Nru libva-vdpau-driver-0.7.4.orig/src/vdpau_video.h libva-vdpau-driver-0.7.4/src/vdpau_video.h
--- libva-vdpau-driver-0.7.4.orig/src/vdpau_video.h	2012-10-05 18:02:58.000000000 +0300
+++ libva-vdpau-driver-0.7.4/src/vdpau_video.h	2020-01-14 14:52:33.336040027 +0300
@@ -274,8 +274,10 @@
 vdpau_CreateSurfaceFromV4L2Buf(
     VADriverContextP    ctx,
     int                 v4l2_fd,
+	/*
     struct v4l2_format *v4l2_fmt,
     struct v4l2_buffer *v4l2_buf,
+	*/
     VASurfaceID        *surface
 ) attribute_hidden;
 
