Index: shaderc-2022.1/CMakeLists.txt
===================================================================
--- shaderc-2022.1/CMakeLists.txt
+++ shaderc-2022.1/CMakeLists.txt	2022-03-08 10:53:40.587725107 +0100
@@ -124,7 +124,6 @@
 
 # Configure subdirectories.
 # We depend on these for later projects, so they should come first.
-add_subdirectory(third_party)
 
 add_subdirectory(libshaderc_util)
 add_subdirectory(libshaderc)
Index: shaderc-2022.1/glslc/CMakeLists.txt
===================================================================
--- shaderc-2022.1/glslc/CMakeLists.txt
+++ shaderc-2022.1/glslc/CMakeLists.txt	2022-03-08 10:53:48.232728595 +0100
@@ -53,7 +53,6 @@
 target_include_directories(glslc_exe PRIVATE ${CMAKE_CURRENT_BINARY_DIR}/.. ${spirv-tools_SOURCE_DIR}/include)
 set_target_properties(glslc_exe PROPERTIES OUTPUT_NAME glslc)
 target_link_libraries(glslc_exe PRIVATE glslc shaderc_util shaderc)
-add_dependencies(glslc_exe build-version)
 
 shaderc_add_tests(
   TEST_PREFIX glslc
