diff -Nuar a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2017-12-08 04:01:41.000000000 +0300
+++ b/CMakeLists.txt	2017-12-16 23:06:41.830377891 +0300
@@ -12,7 +12,7 @@
 set(KF5_VERSION "5.39.0")
 set(kmime_VERSION "5.7.0")
 set(libkleo_VERSION "5.7.0")
-set(QT_REQUIRED_VERSION "6.9.0")
+set(QT_REQUIRED_VERSION "6.9.0")
 set(GPGME_REQUIRED_VERSION "1.8.0")
 
 find_package(ECM ${KF5_VERSION} CONFIG REQUIRED)
diff -Nuar a/src/utils/systemtrayicon.cpp b/src/utils/systemtrayicon.cpp
--- a/src/utils/systemtrayicon.cpp	2017-12-06 14:46:42.000000000 +0300
+++ b/src/utils/systemtrayicon.cpp	2017-12-16 23:14:33.923343827 +0300
@@ -174,7 +174,9 @@
         case QEvent::Close:
             doMainWindowClosed(static_cast<QWidget *>(o));
         // fall through:
+#if QT_VERSION >= QT_VERSION_CHECK(5,8,0)
         Q_FALLTHROUGH();
+#endif
         case QEvent::Show:
         case QEvent::DeferredDelete:
             QMetaObject::invokeMethod(this, "slotEnableDisableActions", Qt::QueuedConnection);
@@ -185,7 +187,9 @@
         case QEvent::Close:
             doAttentionWindowClosed(static_cast<QWidget *>(o));
         // fall through:
+#if QT_VERSION >= QT_VERSION_CHECK(5,8,0)
         Q_FALLTHROUGH();
+#endif
         case QEvent::Show:
         case QEvent::DeferredDelete:
             QMetaObject::invokeMethod(this, "slotEnableDisableActions", Qt::QueuedConnection);
