diff -Nuar a/data/user-dirs.defaults b/data/user-dirs.defaults
--- a/data/user-dirs.defaults	2026-04-08 22:18:20.000000000 +0300
+++ b/data/user-dirs.defaults	2026-04-09 13:56:57.856833510 +0300
@@ -4,8 +4,8 @@
 # will be translated on a per-path-element basis into the users locale
 DESKTOP=Desktop
 DOWNLOAD=Downloads
-TEMPLATES=Templates
-PUBLICSHARE=Public
+# TEMPLATES=Templates
+# PUBLICSHARE=Public
 DOCUMENTS=Documents
 MUSIC=Music
 PICTURES=Pictures
