diff -Nuar a/src/grantleeki18nlocalizer.cpp b/src/grantleeki18nlocalizer.cpp
--- a/src/grantleeki18nlocalizer.cpp	2017-12-04 13:20:01.000000000 +0300
+++ b/src/grantleeki18nlocalizer.cpp	2017-12-16 18:18:43.000518179 +0300
@@ -67,7 +67,9 @@
                 break;
             }
             // fall-through
+#if QT_VERSION >= QT_VERSION_CHECK(5,8,0)
             Q_FALLTHROUGH();
+#endif
         default:
             qCWarning(kcalutils_LOG) << "Unknown type" << iter->typeName() << "(" << iter->type() << ")";
             break;
