diff -Nuar a/cmake/libcrc32c.cmake b/cmake/libcrc32c.cmake
--- a/cmake/libcrc32c.cmake	2022-06-29 18:11:40.492018000 +0300
+++ b/cmake/libcrc32c.cmake	2022-06-29 18:46:54.004865939 +0300
@@ -2,7 +2,7 @@
 init_target(libcrc32c)
 add_library(tg_owt::libcrc32c ALIAS libcrc32c)
 
-set(libcrc32c_loc ${third_party_loc}/crc32c/src)
+set(libcrc32c_loc ${third_party_loc}/crc32c)
 
 configure_file(
     ${libcrc32c_loc}/src/crc32c_config.h.in
diff -Nuar a/src/net/dcsctp/packet/crc32c.cc b/src/net/dcsctp/packet/crc32c.cc
--- a/src/net/dcsctp/packet/crc32c.cc	2022-06-29 18:11:40.874018000 +0300
+++ b/src/net/dcsctp/packet/crc32c.cc	2022-06-29 18:57:39.967819329 +0300
@@ -11,7 +11,7 @@
 
 #include <cstdint>
 
-#include "third_party/crc32c/src/include/crc32c/crc32c.h"
+#include "third_party/crc32c/include/crc32c/crc32c.h"
 
 namespace dcsctp {
 
