From dde9b1c1930d333d8daff25c73400e1e67f3c630 Mon Sep 17 00:00:00 2001
From: Hrvoje Senjan <hrvoje.senjan@gmail.com>
Date: Sun, 7 Sep 2014 22:55:24 +0200
Subject: [PATCH 3/3] Remove export of QT_PLUGIN_PATH

---
 startkde/startkde.cmake | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/startkde/startkde.cmake b/startkde/startkde.cmake
index 51721d1..1d67be0 100644
--- a/startkde/startkde.cmake
+++ b/startkde/startkde.cmake
@@ -220,12 +220,6 @@ for prefix in `echo $scriptpath`; do
   done
 done
 
-# Set the path for Qt plugins provided by KDE
-QT_PLUGIN_PATH=${QT_PLUGIN_PATH+$QT_PLUGIN_PATH:}`qtpaths --plugin-dir`
-# TODO: Do we really need this?
-QT_PLUGIN_PATH=$QT_PLUGIN_PATH:$kdehome/lib/kde5/plugins/
-export QT_PLUGIN_PATH
-
 # Activate the kde font directories.
 #
 # There are 4 directories that may be used for supplying fonts for KDE.
-- 
2.1.0

