diff -Nuar a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2016-09-29 14:35:53.000000000 +0300
+++ b/CMakeLists.txt	2016-10-08 15:25:48.701882533 +0300
@@ -4,7 +4,7 @@
 
 cmake_minimum_required(VERSION 2.8.12 FATAL_ERROR)
 
-set(QT_MIN_VERSION "5.6.1")
+set(QT_MIN_VERSION "5.6.0")
 set(KF5_MIN_VERSION "5.26.0")
 set(INSTALL_SDDM_THEME TRUE)
 find_package(Qt5 ${QT_MIN_VERSION} CONFIG REQUIRED COMPONENTS Widgets Quick QuickWidgets Concurrent Test Script Network)
