diff -Nuar plasma-workspace-5.7.2.orig/CMakeLists.txt plasma-workspace-5.7.2/CMakeLists.txt
--- plasma-workspace-5.7.2.orig/CMakeLists.txt	2016-07-31 19:08:10.309966945 +0300
+++ plasma-workspace-5.7.2/CMakeLists.txt	2016-07-31 19:08:34.248966264 +0300
@@ -4,7 +4,7 @@
 
 cmake_minimum_required(VERSION 2.8.12 FATAL_ERROR)
 
-set(QT_MIN_VERSION "5.6.1")
+set(QT_MIN_VERSION "5.6.0")
 set(KF5_MIN_VERSION "5.18.0")
 set(INSTALL_SDDM_THEME TRUE)
 find_package(Qt5 ${QT_MIN_VERSION} CONFIG REQUIRED COMPONENTS Widgets Quick QuickWidgets Concurrent Test Script Network)
