diff --git a/startkde/startkde.cmake b/startkde/startkde.cmake
index 0c658bd..51721d1 100644
--- a/startkde/startkde.cmake
+++ b/startkde/startkde.cmake
@@ -64,6 +64,14 @@ fi
 
 mkdir -p $configDir
 
+kdehome=$HOME/@KDE_DEFAULT_HOME@
+test -n "$KDEHOME" && kdehome=`echo "$KDEHOME"|sed "s,^~/,$HOME/,"`
+
+# see kstartupconfig source for usage
+mkdir -m 700 -p $kdehome
+mkdir -m 700 -p $kdehome/share
+mkdir -m 700 -p $kdehome/share/config
+
 #This is basically setting defaults so we can use them with kstartupconfig5
 cat >$configDir/startupconfigkeys <<EOF
 kcminputrc Mouse cursorTheme 'breeze_cursors'
