diff -Nuar a/terminal/terminal-window.c b/terminal/terminal-window.c
--- a/terminal/terminal-window.c	2020-12-23 00:50:31.000000000 +0300
+++ b/terminal/terminal-window.c	2021-10-01 21:49:37.528945853 +0300
@@ -390,7 +390,7 @@
   { "show-borders", NULL, N_ ("Show Window _Borders"), NULL, N_ ("Show/hide the window decorations"), G_CALLBACK (terminal_window_action_show_borders), TRUE, },
   { "fullscreen", "view-fullscreen", N_ ("_Fullscreen"), "F11", N_ ("Toggle fullscreen mode"), G_CALLBACK (terminal_window_action_fullscreen), FALSE, },
   { "read-only", NULL, N_ ("_Read-Only"), NULL, N_ ("Toggle read-only mode"), G_CALLBACK (terminal_window_action_readonly), FALSE, },
-  { "scroll-on-output", NULL, N_ ("Scroll on _Output"), NULL, N_ ("Toggle scroll on output"), G_CALLBACK (terminal_window_action_scroll_on_output), FALSE, },
+  { "scroll-on-output", NULL, N_ ("Scroll on _Output"), NULL, N_ ("Toggle scroll on output"), G_CALLBACK (terminal_window_action_scroll_on_output), TRUE, },
 };


@@ -1118,9 +1118,10 @@
       gtk_action_set_sensitive (action, input_enabled);

       /* update scroll on output mode */
+      input_enabled = terminal_screen_get_input_enabled (window->priv->active);
       action = terminal_window_get_action (window, "scroll-on-output");
-      gtk_toggle_action_set_active (GTK_TOGGLE_ACTION (action),
-                                    terminal_screen_get_scroll_on_output (window->priv->active));
+      gtk_toggle_action_set_active (GTK_TOGGLE_ACTION (action), !input_enabled);
+                                    //terminal_screen_get_scroll_on_output (window->priv->active));

       /* update the "Go" menu */
       action = g_object_get_qdata (G_OBJECT (window->priv->active), tabs_menu_action_quark);
