diff --git a/src/device_p.h b/src/device_p.h
index bbc39bc..1c6cdab 100644
--- a/src/device_p.h
+++ b/src/device_p.h
@@ -119,7 +119,7 @@ public:
             }
         }
         if (m_activePortIndex == static_cast<quint32>(-1)) {
-            qCWarning(PULSEAUDIOQT) << "Failed to find active port" << QString::fromUtf8(info->active_port->name);
+            qCWarning(PULSEAUDIOQT) << "Failed to find active port";
             for (const auto &port : std::as_const(m_ports)) {
                 qCWarning(PULSEAUDIOQT) << "Available port:" << port->name();
             }
