
Commit:

From b347e04452ff6382ace8fba9c81f5bcb63be17a6 Mon Sep 17 00:00:00 2001
From: Thomas Zimmermann <tzimmermann@suse.de>
Date: Sun, 2 May 2021 12:49:53 +0200
Subject: drm: Remove pdev field from struct drm_device

removed pdev from struct.

Signed-off-by: Thomas Backlund <tmb@mageia.org>

diff -Nurp NVIDIA-Linux-x86_64-390.144.orig/kernel/nvidia-drm/nvidia-drm-drv.c NVIDIA-Linux-x86_64-390.144/kernel/nvidia-drm/nvidia-drm-drv.c
--- NVIDIA-Linux-x86_64-390.144.orig/kernel/nvidia-drm/nvidia-drm-drv.c	2021-06-03 02:10:01.000000000 +0300
+++ NVIDIA-Linux-x86_64-390.144/kernel/nvidia-drm/nvidia-drm-drv.c	2021-08-31 22:28:43.141140842 +0300
@@ -35,6 +35,8 @@
 #include "nvidia-drm-gem-nvkms-memory.h"
 #include "nvidia-drm-gem-user-memory.h"
 
+#include <linux/version.h>
+
 #if defined(NV_DRM_AVAILABLE)
 
 #include "nvidia-drm-ioctl.h"
@@ -811,7 +817,9 @@ static void nv_drm_register_drm_device(c
 
     dev->dev_private = nv_dev;
     nv_dev->dev = dev;
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 14, 0)
     dev->pdev = pdev;
+#endif
 
     /* Register DRM device to DRM sub-system */
 
