diff -Nurp NVIDIA-Linux-x86_64-390.132.orig/kernel/nvidia-uvm/uvm_linux.h NVIDIA-Linux-x86_64-390.132/kernel/nvidia-uvm/uvm_linux.h
--- NVIDIA-Linux-x86_64-390.132.orig/kernel/nvidia-uvm/uvm_linux.h	2019-11-01 14:37:49.000000000 +0200
+++ NVIDIA-Linux-x86_64-390.132/kernel/nvidia-uvm/uvm_linux.h	2020-04-11 01:44:45.233557933 +0300
@@ -329,7 +329,16 @@ static inline uint64_t NV_DIV64(uint64_t
 }
 #endif
 
-#if defined(CLOCK_MONOTONIC_RAW)
+#if defined(NV_KTIME_GET_RAW_TS64_PRESENT)
+static inline NvU64 NV_GETTIME(void)
+{
+    struct timespec64 ts64 = {0};
+
+    ktime_get_raw_ts64(&ts64);
+
+    return (ts64.tv_sec * 1000000000ULL + ts64.tv_nsec);
+}
+#elif defined(CLOCK_MONOTONIC_RAW)
 /* Return a nanosecond-precise value */
 static inline NvU64 NV_GETTIME(void)
 {
@@ -345,7 +354,7 @@ static inline NvU64 NV_GETTIME(void)
  * available non-GPL symbols. */
 static inline NvU64 NV_GETTIME(void)
 {
-    struct timeval tv = {0};
+    struct nv_timeval tv = {0};
 
     nv_gettimeofday(&tv);
 
