diff -rupN --no-dereference PDAL-2.4.2-src/doc/conf.py PDAL-2.4.2-src-new/doc/conf.py
--- PDAL-2.4.2-src/doc/conf.py	2022-06-06 20:04:01.000000000 +0200
+++ PDAL-2.4.2-src-new/doc/conf.py	2022-06-07 08:49:14.420039447 +0200
@@ -58,6 +58,7 @@ extensions = ['breathe', 'sphinx.ext.aut
               'sphinx.ext.mathjax', 'sphinx.ext.intersphinx',
               'sphinxcontrib.bibtex', 'embed']

+bibtex_bibfiles = ['bibliography.bib']

 # Add any paths that contain templates here, relative to this directory.
 templates_path = ['_templates']
