diff -Nuar a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2017-12-08 04:04:19.000000000 +0300
+++ b/CMakeLists.txt	2017-12-17 00:13:48.399087354 +0300
@@ -28,7 +28,7 @@
 
 set(AKONADIMIME_LIB_VERSION "5.7.0")
 set(KDEPIM_LIB_VERSION "5.7.0")
-set(QT_REQUIRED_VERSION "6.9.0")
+set(QT_REQUIRED_VERSION "6.9.0")
 set(kmime_LIB_VERSION "5.7.0")
 set(AKONADI_VERSION "5.7.0")
 
diff -Nuar a/src/mailimporter/filters/filterlnotes.cpp b/src/mailimporter/filters/filterlnotes.cpp
--- a/src/mailimporter/filters/filterlnotes.cpp	2017-12-04 13:20:16.000000000 +0300
+++ b/src/mailimporter/filters/filterlnotes.cpp	2017-12-17 00:18:28.129067170 +0300
@@ -132,7 +132,9 @@
                 tempfile->setAutoRemove(false);
                 tempfile->open();
                 // fall through
+#if QT_VERSION >= QT_VERSION_CHECK(5,8,0)
                 Q_FALLTHROUGH();
+#endif
             // inside a message state
             case 1:
                 if (ch == 0x0c) {
