diff -Nuar lxdm-0.5.3.orig/data/lxdm.conf.in lxdm-0.5.3/data/lxdm.conf.in
--- lxdm-0.5.3.orig/data/lxdm.conf.in	2017-11-19 17:40:33.970019559 +0300
+++ lxdm-0.5.3/data/lxdm.conf.in	2017-11-19 17:41:50.406021375 +0300
@@ -7,7 +7,7 @@
 # timeout=10
 
 ## default session or desktop used when no systemwide config
-# session=/usr/bin/startlxde
+session=/usr/bin/startkde
 
 ## uncomment and set to set numlock on your keyboard
 # numlock=0
