diff -Nuar a/pam/lxdm b/pam/lxdm
--- a/pam/lxdm	2013-09-08 07:52:35.000000000 +0300
+++ b/pam/lxdm	2021-05-01 15:25:24.131962109 +0300
@@ -4,6 +4,6 @@
 account    include	system-auth
 session    optional	pam_keyinit.so force revoke
 session    include	system-auth
-session    optional	pam_console.so
+#session    optional	pam_console.so
 session    optional	pam_gnome_keyring.so auto_start
-session    optional	pam_selinux.so
+#session    optional	pam_selinux.so
