diff -Nuar a/data/default.layout b/data/default.layout
--- a/data/default.layout	2021-04-20 13:39:33.458596338 +0300
+++ b/data/default.layout	2021-04-20 13:46:07.673626151 +0300
@@ -9,11 +9,33 @@
 size=24
 
 [Object menu-bar]
-object-type=menu-bar
-toplevel-id=top
+object-type=applet
+applet-iid=BriskMenuFactory::BriskMenu
+toplevel-id=top
 position=0
 locked=true
 
+[Object file-browser]
+object-type=launcher
+launcher-location=/usr/share/applications/caja-browser.desktop
+toplevel-id=top
+position=10
+locked=true
+
+[Object terminal]
+object-type=launcher
+launcher-location=/usr/share/applications/mate-terminal.desktop
+toplevel-id=top
+position=20
+locked=true
+
+[Object web-browser]
+object-type=launcher
+launcher-location=/usr/share/applications/firefox.desktop
+toplevel-id=top
+position=30
+locked=true
+
 [Object notification-area]
 object-type=applet
 applet-iid=NotificationAreaAppletFactory::NotificationArea
