diff -Nuar a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2017-12-08 04:01:18.000000000 +0300
+++ b/CMakeLists.txt	2017-12-16 18:01:14.579593828 +0300
@@ -32,7 +32,7 @@
 set(kcontacts_LIB_VERSION "5.7.0")
 set(kcalendarcore_LIB_VERSION "5.7.0")
 set(kmime_LIB_VERSION "5.7.0")
-set(QT_REQUIRED_VERSION "6.9.0")
+set(QT_REQUIRED_VERSION "6.9.0")
 
 find_package(Qt5 ${QT_REQUIRED_VERSION} CONFIG REQUIRED Core Test)
 find_package(KF5 ${KF5_VERSION} REQUIRED COMPONENTS I18n Config Crash kcmutils)
