--- lxqt-policykit-0.11.0/src/policykitagentgui.cpp	2016-09-24 00:46:00.000000000 +0200
+++ src/policykitagentgui.cpp.new	2016-09-26 11:45:19.666066086 +0200
@@ -49,6 +49,8 @@
     if (icon.isNull())
         icon = QIcon::fromTheme(QLatin1String("dialog-question"));
     iconLabel->setPixmap(icon.pixmap(64, 64));
+    identityComboBox->hide();
+    label->hide();
 
     const uid_t current_uid = getuid();
     int current_user_index = -1;
 
