diff -Nuar a/etc/clamav-milter.conf.sample b/etc/clamav-milter.conf.sample
--- a/etc/clamav-milter.conf.sample	2023-10-24 20:10:14.000000000 +0300
+++ b/etc/clamav-milter.conf.sample	2023-11-04 11:15:20.706075909 +0300
@@ -3,7 +3,7 @@
 ##

 # Comment or remove the line below.
-Example
+# Example


 ##
diff -Nuar a/etc/clamd.conf.sample b/etc/clamd.conf.sample
--- a/etc/clamd.conf.sample	2023-10-24 20:10:14.000000000 +0300
+++ b/etc/clamd.conf.sample	2023-11-04 11:20:07.859485998 +0300
@@ -5,13 +5,14 @@


 # Comment or remove the line below.
-Example
+# Example

 # Uncomment this option to enable logging.
 # LogFile must be writable for the user running daemon.
 # A full path is required.
 # Default: disabled
 #LogFile /tmp/clamd.log
+LogFile /var/log/clamd.log

 # By default the log file is locked for writing - the lock protects against
 # running clamd multiple times (if want to run another clamd, please
@@ -19,7 +20,7 @@
 # the daemon with --config-file option).
 # This option disables log file locking.
 # Default: no
-#LogFileUnlock yes
+LogFileUnlock yes

 # Maximum size of the log file.
 # Value of 0 disables the limit.
@@ -75,6 +76,7 @@
 # also owned by root to keep other users from tampering with it.
 # Default: disabled
 #PidFile /run/clamav/clamd.pid
+PidFile /run/clamd.pid

 # Optional path to the global temporary directory.
 # Default: system specific (usually /tmp or /var/tmp).
@@ -82,7 +84,7 @@

 # Path to the database directory.
 # Default: hardcoded (depends on installation options)
-#DatabaseDirectory /var/lib/clamav
+DatabaseDirectory /var/lib/clamav

 # Only load the official signatures published by the ClamAV project.
 # Default: no
@@ -100,6 +102,7 @@
 # Default: disabled (must be specified by a user)
 #LocalSocket /run/clamav/clamd.sock
 #LocalSocket /tmp/clamd.sock
+LocalSocket /run/clamd.socket

 # Sets the group ownership on the unix socket.
 # Default: disabled (the primary group of the user running clamd)
@@ -115,7 +118,7 @@

 # TCP port address.
 # Default: no
-#TCPSocket 3310
+TCPSocket 3310

 # TCP address.
 # By default we bind to INADDR_ANY, probably not wise.
diff -Nuar a/etc/freshclam.conf.sample b/etc/freshclam.conf.sample
--- a/etc/freshclam.conf.sample	2023-10-24 20:10:14.000000000 +0300
+++ b/etc/freshclam.conf.sample	2023-11-04 11:22:10.603951481 +0300
@@ -5,16 +5,16 @@


 # Comment or remove the line below.
-Example
+# Example

 # Path to the database directory.
 # WARNING: It must match clamd.conf's directive!
 # Default: hardcoded (depends on installation options)
-#DatabaseDirectory /var/lib/clamav
+DatabaseDirectory /var/lib/clamav

 # Path to the log file (make sure it has proper permissions)
 # Default: disabled
-#UpdateLogFile /var/log/freshclam.log
+UpdateLogFile /var/log/freshclam.log

 # Maximum size of the log file.
 # Value of 0 disables the limit.
@@ -53,6 +53,7 @@
 # also owned by root to keep other users from tampering with it.
 # Default: disabled
 #PidFile /run/clamav/freshclam.pid
+PidFile /run/freshclam.pid

 # By default when started freshclam drops privileges and switches to the
 # "clamav" user. This directive allows you to change the database owner.
