diff --git a/etc/clamav-milter.conf.sample b/etc/clamav-milter.conf.sample
index 7ca0e6e..98af596 100644
--- a/etc/clamav-milter.conf.sample
+++ b/etc/clamav-milter.conf.sample
@@ -3,7 +3,7 @@
 ##
 
 # Comment or remove the line below.
-Example
+#Example
 
 
 ##
diff --git a/etc/clamd.conf.sample b/etc/clamd.conf.sample
index 2f18866..2c63980 100644
--- a/etc/clamd.conf.sample
+++ b/etc/clamd.conf.sample
@@ -5,13 +5,13 @@
 
 
 # Comment or remove the line below.
-Example
+#Example
 
 # Uncomment this option to enable logging.
 # LogFile must be writable for the user running daemon.
 # A full path is required.
 # Default: disabled
-#LogFile /tmp/clamd.log
+LogFile /var/log/clamd.log
 
 # By default the log file is locked for writing - the lock protects against
 # running clamd multiple times (if want to run another clamd, please
@@ -19,7 +19,7 @@ Example
 # the daemon with --config-file option).
 # This option disables log file locking.
 # Default: no
-#LogFileUnlock yes
+LogFileUnlock yes
 
 # Maximum size of the log file.
 # Value of 0 disables the limit.
@@ -74,7 +74,7 @@ Example
 # It is recommended that the directory where this file is stored is
 # also owned by root to keep other users from tampering with it.
 # Default: disabled
-#PidFile /var/run/clamd.pid
+PidFile /run/clamd.pid
 
 # Optional path to the global temporary directory.
 # Default: system specific (usually /tmp or /var/tmp).
@@ -82,7 +82,7 @@ Example
 
 # Path to the database directory.
 # Default: hardcoded (depends on installation options)
-#DatabaseDirectory /var/lib/clamav
+DatabaseDirectory /var/lib/clamav
 
 # Only load the official signatures published by the ClamAV project.
 # Default: no
@@ -109,7 +109,7 @@ Example
 
 # TCP port address.
 # Default: no
-#TCPSocket 3310
+TCPSocket 3310
 
 # TCP address.
 # By default we bind to INADDR_ANY, probably not wise.
@@ -117,7 +117,7 @@ Example
 # from the outside world. This option can be specified multiple
 # times if you want to listen on multiple IPs. IPv6 is now supported.
 # Default: no
-#TCPAddr localhost
+TCPAddr localhost
 
 # Maximum length the queue of pending connections may grow to.
 # Default: 200
diff --git a/etc/freshclam.conf.sample b/etc/freshclam.conf.sample
index 26c4e7a..5f791e8 100644
--- a/etc/freshclam.conf.sample
+++ b/etc/freshclam.conf.sample
@@ -5,12 +5,12 @@
 
 
 # Comment or remove the line below.
-Example
+#Example
 
 # Path to the database directory.
 # WARNING: It must match clamd.conf's directive!
 # Default: hardcoded (depends on installation options)
-#DatabaseDirectory /var/lib/clamav
+DatabaseDirectory /var/lib/clamav
 
 # Path to the log file (make sure it has proper permissions)
 # Default: disabled
@@ -51,7 +51,7 @@ Example
 # It is recommended that the directory where this file is stored is
 # also owned by root to keep other users from tampering with it.
 # Default: disabled
-#PidFile /var/run/freshclam.pid
+PidFile /run/freshclam.pid
 
 # By default when started freshclam drops privileges and switches to the
 # "clamav" user. This directive allows you to change the database owner.
