diff -Nuar a/src/mimeutils.c b/src/mimeutils.c
--- a/src/mimeutils.c	2016-05-11 23:05:33.000000000 +0300
+++ b/src/mimeutils.c	2018-04-19 23:01:05.615018503 +0300
@@ -52,9 +52,9 @@
   "x-directory"
 };
 
-/* Defined in RFC 2045/2046 and RFC 2077 */
+/* Defined in RFC 2045/2046, RFC 2077 and RFC 8081 */
 static const char *registered_discrete_media_types[] = {
-  "application", "audio", "image", "model", "text", "video"
+  "application", "audio", "font", "image", "model", "text", "video"
 };
 
 /* Defined in RFC 2045/2046 */
