diff -Nuar a/src/lib/preferences/preferences.cpp b/src/lib/preferences/preferences.cpp
--- a/src/lib/preferences/preferences.cpp	2018-02-27 16:56:58.000000000 +0300
+++ b/src/lib/preferences/preferences.cpp	2018-04-08 21:18:23.966125844 +0300
@@ -131,7 +131,7 @@
     Settings settings;
     //GENERAL URLs
     settings.beginGroup("Web-URL-Settings");
-    m_homepage = settings.value("homepage", QUrl(QSL("falkon:start"))).toUrl();
+    m_homepage = settings.value("homepage", QUrl(QSL("falkon:speeddial"))).toUrl();
     m_newTabUrl = settings.value("newTabUrl", QUrl(QSL("falkon:speeddial"))).toUrl();
     ui->homepage->setText(m_homepage.toEncoded());
     ui->newTabUrl->setText(m_newTabUrl.toEncoded());
diff -Nuar a/src/lib/app/browserwindow.cpp b/src/lib/app/browserwindow.cpp
--- a/src/lib/app/browserwindow.cpp	2018-02-27 16:56:58.000000000 +0300
+++ b/src/lib/app/browserwindow.cpp	2018-04-08 20:11:08.593417016 +0300
@@ -576,7 +576,7 @@
 
     //Url settings
     settings.beginGroup("Web-URL-Settings");
-    m_homepage = settings.value("homepage", "falkon:start").toUrl();
+    m_homepage = settings.value("homepage", "falkon:speeddial").toUrl();
     settings.endGroup();
 
     //Browser Window settings
