diff --git a/gmusicbrowser_layout.pm b/gmusicbrowser_layout.pm
index e3683ab..c556ace 100644
--- a/gmusicbrowser_layout.pm
+++ b/gmusicbrowser_layout.pm
@@ -1823,7 +1823,7 @@ sub SaveWindowOptions
 		$wstate{fullscreen}=1 if $self->{fullscreen};
 		$wstate{ontop}=1 if $self->{ontop};
 		$wstate{below}=1 if $self->{below};
-		$wstate{nodecoration}=1 unless $self->get_decorated;
+		#$wstate{nodecoration}=1 unless $self->get_decorated;
 		$wstate{skippager}=1 if $self->get_skip_pager_hint;
 		if ($self->{saved_position})
 		{	$wstate{pos}=$self->{saved_position};
@@ -1852,7 +1852,7 @@ sub SetWindowOptions
 	$self->stick if $opt->{sticky};
 	$self->set_keep_above(1) if $opt->{ontop};
 	$self->set_keep_below(1) if $opt->{below};
-	$self->set_decorated(0)  if $opt->{nodecoration};
+	#$self->set_decorated(0)  if $opt->{nodecoration};
 	$self->set_skip_pager_hint(1) if $opt->{skippager};
 	$self->set_skip_taskbar_hint(1) if $opt->{skiptaskbar};
 	$self->{opacity}=$opt->{opacity} if defined $opt->{opacity};
